{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use kernel py3-6\n",
    "#import goatools\n",
    "from goatools import obo_parser\n",
    "import re\n",
    "import json\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from goatools.go_enrichment import GOEnrichmentStudy\n",
    "import matplotlib.pyplot as plt\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def matchgff2(feature, gff_file='/home/t44p/PW_rawdata/Transciptome_GenomeAnnotation/Xele_annotated2_gff_export2.gff', obo_path=\"/home/t44p/PW_rawdata/go_obo/go.obo\", namespace=None, depth_threshold=0, goea=False):\n",
    "    \"\"\"\n",
    "    Searches a GFF (General Feature Format) file for specific features and retrieves associated Gene Ontology (GO) terms,\n",
    "    with optional filtering by GO term namespace and depth.\n",
    "\n",
    "    Parameters:\n",
    "    ----------\n",
    "    feature : list or iterable\n",
    "        An iterable of strings representing the features to search for in the GFF file.\n",
    "\n",
    "    gff_file : str, optional\n",
    "        The file path to the GFF file. Defaults to a predefined path.\n",
    "\n",
    "    obo_path : str, optional\n",
    "        The file path to the Gene Ontology .obo file. Defaults to a predefined path.\n",
    "\n",
    "    namespace : list of str or None, optional\n",
    "        List of GO term namespaces to filter term count results. Valid options are 'biological_process', \n",
    "        'molecular_function', 'cellular_component', or None. Defaults to None (no filtering).\n",
    "\n",
    "    depth_threshold : int, optional\n",
    "        Minimum depth of GO terms to include in the results. Defaults to 0 (no filtering).\n",
    "\n",
    "    goea : bool, optional\n",
    "        Whether to perform Gene Ontology Enrichment Analysis. Defaults to False.\n",
    "\n",
    "    Returns:\n",
    "    -------\n",
    "    tuple\n",
    "        A tuple containing:\n",
    "        1. A dictionary with features as keys and lists of lines from the GFF file as values.\n",
    "        2. A dictionary mapping features to their associated GO terms.\n",
    "        3. A dictionary of GO term counts.\n",
    "        4. GOEA results (if goea is True), filtered by the specified depth threshold for all namespaces\n",
    "\n",
    "    Notes:\n",
    "    -----\n",
    "    - The function uses regular expressions for precise matching of features.\n",
    "    - It extracts GO IDs from matched lines and retrieves their corresponding names and namespaces.\n",
    "    - It is possible for the rawcounts to filter for namespaces \n",
    "    - If `goea` is True, the function performs GO Enrichment Analysis and returns the results filtered by the specified depth threshold but not for namespaces. \n",
    "    - The Enrichment analysis will be performed across all namespaces, filtering parameter \"namespace\" will have no effect on the enrichment analysis\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    valid_namespaces = {'biological_process', 'molecular_function', 'cellular_component', None}\n",
    "    # Check if namespace is a list containing only valid elements\n",
    "    if isinstance(namespace, list) and not all(ns in valid_namespaces for ns in namespace):\n",
    "        raise ValueError(\"Invalid namespace provided. Valid options are 'biological_process', \"\n",
    "                         \"'molecular_function', 'cellular_component', or a list containing any of these. \"\n",
    "                         \"You can also use None for no filtering.\")\n",
    "\n",
    "    with open(gff_file, 'r') as file:\n",
    "        go_ontology = obo_parser.GODag(obo_path)\n",
    "        \n",
    "        lines_where_feat_found = {}\n",
    "        go_ids = {}\n",
    "        background_genes = []\n",
    "        go_term_count = {}\n",
    "\n",
    "         # Find the depth of each GO term\n",
    "        go_depths = {go_id: go_term.depth for go_id, go_term in go_ontology.items()}\n",
    "\n",
    "        # construct background genes\n",
    "        if goea:\n",
    "            for line in file:\n",
    "                if not line.lstrip().startswith('#'):\n",
    "                    background_genes.append(line.split('\\t')[0])\n",
    "\n",
    "        for feat in feature:\n",
    "            file.seek(0)  # reset file pointer to the beginning for each feature\n",
    "            lines_where_feat_found[feat] = []\n",
    "            go_ids[feat] = {}\n",
    "            pattern = re.compile(re.escape(feat) + r'\\t')  # exact match followed by a tab\n",
    "            for line in file:\n",
    "                if pattern.search(line):\n",
    "                    lines_where_feat_found[feat].append(line.strip())  # Store the line (as a string) if feature is found\n",
    "                    # Extract GO id\n",
    "                    match = re.search(r\"Ontology_id=([GO:\\d,]+)\", line.strip())\n",
    "                    if match:\n",
    "                        ids = match.group(1).split(',')\n",
    "                        # Map Terms to Ids and Count Occurrences\n",
    "                        for id in ids:\n",
    "                            term = go_ontology.get(id)\n",
    "                            if term is not None:\n",
    "                                go_ids[feat][id] = {'name': term.name, 'namespace': term.namespace}\n",
    "\n",
    "                                if namespace is None or term.namespace in namespace and go_ontology[id].depth >= depth_threshold:\n",
    "                                    # Count Occurrences\n",
    "                                    if id in go_term_count:\n",
    "                                        go_term_count[id] = (term.name, go_term_count[id][1] + 1, term.namespace)\n",
    "                                    else:\n",
    "                                        go_term_count[id] = (term.name, 1, term.namespace)\n",
    "                            else:\n",
    "                                go_ids[feat][id] = {'name': None, 'namespace': None}\n",
    "                                if id not in go_term_count:\n",
    "                                    go_term_count[id] = (None, 1)\n",
    "        if goea:\n",
    "            print(\"GO Enrichment Analysis >>\")\n",
    "            goea_obj = GOEnrichmentStudy(\n",
    "                background_genes,\n",
    "                go_ids,  # This needs to be a dict mapping gene IDs to a set of GO IDs\n",
    "                go_ontology,\n",
    "                propagate_counts=False,\n",
    "                alpha=0.05,  # significance level for the statistical test\n",
    "                methods=['fdr_bh']  # correction method for multiple testing\n",
    "            )\n",
    "            goea_result = goea_obj.run_study(go_ids.keys())\n",
    "\n",
    "            # filter based on depth\n",
    "            filtered_goea_results = [res for res in goea_result if res.goterm.depth >= depth_threshold]\n",
    "            return lines_where_feat_found, go_ids, go_term_count, filtered_goea_results\n",
    "\n",
    "\n",
    "        return lines_where_feat_found, go_ids, go_term_count\n",
    "    \n",
    "\n",
    "def tabulate(term_count_dict, sort=True):\n",
    "    \"\"\"\n",
    "    Prints a tabulated view of the term counts from a dictionary, such as the one returned by matchgff2.\n",
    "\n",
    "    Parameters:\n",
    "    ----------\n",
    "    term_count_dict : dict\n",
    "        A dictionary where keys are Gene Ontology (GO) IDs, and values are tuples containing the GO term name, \n",
    "        the count of occurrences, and optionally the namespace. \n",
    "        The structure is typically: {GO_ID: (GO_Term, Count, Namespace)}.\n",
    "\n",
    "    sort : bool, optional\n",
    "        Whether to sort the output based on the count of occurrences of each GO term. Defaults to True.\n",
    "\n",
    "    Description:\n",
    "    ------------\n",
    "    This function iterates through the term_count_dict and prints each GO term's count, ID, name, and namespace \n",
    "    in a tabular format. It can optionally sort these terms based on their count in descending order.\n",
    "\n",
    "    Note:\n",
    "    -----\n",
    "    The function is primarily a utility for visualizing the output of the matchgff2 function. It does not return any value \n",
    "    but prints the information directly to the console.\n",
    "\n",
    "    Example Usage:\n",
    "    --------------\n",
    "    go_term_count = {'GO:0000001': ('term1', 5, 'biological_process'), 'GO:0000002': ('term2', 3, 'cellular_component')}\n",
    "    tabulate(go_term_count, sort=True)\n",
    "    \"\"\"\n",
    "    print(f\"count\\tGO ID\\tGO Term\\tnamespace\")\n",
    "\n",
    "    # Conditionally sort the dictionary if required\n",
    "    items = sorted(term_count_dict.items(), key=lambda x: x[1][1], reverse=True) if sort else term_count_dict.items()\n",
    "\n",
    "    for goid, values in items:\n",
    "        count = values[1]\n",
    "        term = values[0] if values[0] is not None else \"N/A\"\n",
    "        namespace = values[2] if len(values) > 2 else \"N/A\"\n",
    "\n",
    "        print(f\"{count}\\t{goid}\\t{term}\\t{namespace}\")\n",
    "    #return \n",
    "\n",
    "\n",
    "def parse_obo_file(obo_path):\n",
    "    go_terms = {}\n",
    "    with open(obo_path, 'r') as file:\n",
    "        for line in file:\n",
    "            if line.startswith('id: '):\n",
    "                current_id = line.strip().split(' ')[1]\n",
    "            elif line.startswith('namespace: '):\n",
    "                current_namespace = line.strip().split(' ')[1]\n",
    "                go_terms[current_id] = current_namespace\n",
    "    return go_terms\n",
    "\n",
    "def goea_results_to_file(goea_results, obo_path, path='./lasso_models/10xKfold_lasso_output/goea/', to_csv=True, to_excel=False):\n",
    "    \"\"\"\n",
    "    Processes and saves Gene Ontology Enrichment Analysis (GOEA) results in CSV and/or Excel format.\n",
    "\n",
    "    Parameters:\n",
    "    - goea_results: dict\n",
    "      - A dictionary with prefixes for file names as keys and lists of GOEA result objects as values.\n",
    "    - obo_path: str\n",
    "      - Path to the .obo file containing GO terms.\n",
    "    - path: str, optional\n",
    "      - Directory path for saving output files. Defaults to './lasso_models/10xKfold_lasso_output/goea/'.\n",
    "    - to_csv: bool, optional\n",
    "      - If True, saves results in CSV format. Defaults to True.\n",
    "    - to_excel: bool, optional\n",
    "      - If True, saves results in Excel format. Defaults to False.\n",
    "\n",
    "    Functionality:\n",
    "    - Parses the GO terms from the .obo file.\n",
    "    - Creates a DataFrame from the GOEA results for each namespace ('BP', 'MF', 'CC').\n",
    "    - Constructs an enrichment matrix with GO IDs as rows and goea_results keys as columns.\n",
    "    - Saves individual DataFrames and the enrichment matrix in the specified format (CSV/Excel).\n",
    "\n",
    "    Note:\n",
    "    - 'openpyxl' library is required for saving to Excel files.\n",
    "    - The function handles namespaces and enrichment status ('e' for enriched) for GOEA results.\n",
    "    - Ensures all GO IDs are represented in the enrichment matrix, populating missing values with zeros.\n",
    "    - Generates file names based on the provided prefixes and namespaces.\n",
    "    - Prints file paths of saved files for user reference.\n",
    "    \"\"\"\n",
    "    # Load GO terms from the obo file\n",
    "    go_terms = parse_obo_file(obo_path)\n",
    "\n",
    "    # Mapping for suffixes\n",
    "    suffix_mapping = {'BP': 'biop', 'MF': 'molf', 'CC': 'cellc'}\n",
    "\n",
    "    for ns in ['BP', 'MF', 'CC']:\n",
    "        \n",
    "        # mapping for enrichment matrix population\n",
    "        namespace_mapping = {\n",
    "        'BP': 'biological_process',\n",
    "        'MF': 'molecular_function',\n",
    "        'CC': 'cellular_component'\n",
    "        }\n",
    "        # populating enrichment matrix\n",
    "        enrichment_matrix = {key: {go_id: 0 for go_id in go_terms if go_terms[go_id] == namespace_mapping[ns]} for key in goea_results}\n",
    "\n",
    "        for file_prefix, res in goea_results.items():\n",
    "            suffix = suffix_mapping[ns]\n",
    "\n",
    "            # Filtering relevant records\n",
    "            relevant_records = [r for r in res if r.NS == ns]\n",
    "\n",
    "            # Creating a DataFrame\n",
    "            data = {\n",
    "                \"GO_ID\": [r.goterm.id for r in relevant_records],\n",
    "                \"level\": [r.goterm.level for r in relevant_records],\n",
    "                \"depth\": [r.goterm.depth for r in relevant_records],\n",
    "                \"GO_Term\": [r.goterm.name for r in relevant_records],\n",
    "                \"study_count\": [r.study_count for r in relevant_records],\n",
    "                \"study_n\": [r.study_n for r in relevant_records],\n",
    "                \"pop_count\": [r.pop_count for r in relevant_records],\n",
    "                \"pop_n\": [r.pop_n for r in relevant_records],\n",
    "                \"enrichment\": [r.enrichment for r in relevant_records],\n",
    "                \"adj_p_fdr_bh\": [r.p_fdr_bh for r in relevant_records]\n",
    "            }\n",
    "\n",
    "            for record in relevant_records:\n",
    "                if record.enrichment == 'e':\n",
    "                    enrichment_matrix[file_prefix][record.goterm.id] = 1\n",
    "\n",
    "            df = pd.DataFrame(data)\n",
    "            # Saving to CSV\n",
    "            if to_csv:\n",
    "                file_name = f\"{path}{file_prefix}_{suffix}.csv\"\n",
    "                df.to_csv(file_name, index=False)\n",
    "                print(f\"Saved to {file_name}\")\n",
    "            if to_excel:\n",
    "                path_suffix = f\"{path}goea_{suffix}.xlsx\"\n",
    "                if os.path.exists(path_suffix):\n",
    "                    x_mode = 'a'\n",
    "                else:\n",
    "                    x_mode = 'w'\n",
    "                    \n",
    "                with pd.ExcelWriter(path_suffix, mode=x_mode, engine='openpyxl') as writer:\n",
    "                    sheet = file_prefix\n",
    "                    df.to_excel(writer, sheet_name=sheet, index=False)\n",
    "\n",
    "        # Ensure that all GO IDs are represented and convert the matrix to DataFrame\n",
    "        enrichment_df = pd.DataFrame(enrichment_matrix)\n",
    "        \n",
    "        # The DataFrame should now be fully populated with 0s and 1s. Converting to integer type should not be an issue.\n",
    "        enrichment_df = enrichment_df.fillna(0).astype(int)\n",
    "\n",
    "        # Save the DataFrame\n",
    "        enrichment_file_name = f\"{path}enrichment_matrix_{suffix}.csv\"\n",
    "        enrichment_df.to_csv(enrichment_file_name)\n",
    "        print(f\"Enrichment matrix saved to {enrichment_file_name}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GOEA for all files "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cellobiose_361_204_rt14_40\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  1%    405 of 30,522 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 405 IDs.\n",
      "100%    405 of    405 study items found in association\n",
      "100%    405 of    405 study items found in population(30522)\n",
      "Calculating 1,693 uncorrected p-values using fisher_scipy_stats\n",
      "   1,693 terms are associated with    362 of 30,522 population items\n",
      "   1,693 terms are associated with    362 of    405 study items\n",
      "  METHOD fdr_bh:\n",
      "   1,693 GO terms found significant (< 0.05=alpha) (1693 enriched +   0 purified): statsmodels fdr_bh\n",
      "     362 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "tyrosine_218_280_rt10_78\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  2%    608 of 30,522 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 608 IDs.\n",
      "100%    608 of    608 study items found in association\n",
      "100%    608 of    608 study items found in population(30522)\n",
      "Calculating 2,356 uncorrected p-values using fisher_scipy_stats\n",
      "   2,356 terms are associated with    537 of 30,522 population items\n",
      "   2,356 terms are associated with    537 of    608 study items\n",
      "  METHOD fdr_bh:\n",
      "   2,356 GO terms found significant (< 0.05=alpha) (2356 enriched +   0 purified): statsmodels fdr_bh\n",
      "     537 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "_671_2775803_7_254060113\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  3%  1,059 of 30,522 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 1059 IDs.\n",
      "100%  1,059 of  1,059 study items found in association\n",
      "100%  1,059 of  1,059 study items found in population(30522)\n",
      "Calculating 3,667 uncorrected p-values using fisher_scipy_stats\n",
      "   3,667 terms are associated with    944 of 30,522 population items\n",
      "   3,667 terms are associated with    944 of  1,059 study items\n",
      "  METHOD fdr_bh:\n",
      "   3,667 GO terms found significant (< 0.05=alpha) (3667 enriched +   0 purified): statsmodels fdr_bh\n",
      "     944 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Coumaroyl_quinic_acid_\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  2%    563 of 30,522 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 563 IDs.\n",
      "100%    563 of    563 study items found in association\n",
      "100%    563 of    563 study items found in population(30522)\n",
      "Calculating 2,339 uncorrected p-values using fisher_scipy_stats\n",
      "   2,339 terms are associated with    509 of 30,522 population items\n",
      "   2,339 terms are associated with    509 of    563 study items\n",
      "  METHOD fdr_bh:\n",
      "   2,339 GO terms found significant (< 0.05=alpha) (2339 enriched +   0 purified): statsmodels fdr_bh\n",
      "     509 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n"
     ]
    }
   ],
   "source": [
    "path_gc = \"/home/t44p/PW_rawdata/results/full_lasso/gcms/\"\n",
    "path_lc =\"/home/t44p/PW_rawdata/results/full_lasso/lcms/\"\n",
    "\n",
    "\n",
    "lasso_goea = {}\n",
    "\n",
    "\n",
    "# Iterate over each file in the directory\n",
    "for p in [path_gc, path_lc]:\n",
    "    tmp = 0\n",
    "\n",
    "    for file in os.listdir(p):\n",
    "        if file.endswith(\".json\") and not(file.startswith('gcms_dict_nXcv') or file.startswith('lcms_dict_nXcv')):\n",
    "            file_path = os.path.join(p, file)\n",
    "            with open(file_path, 'r') as json_file:\n",
    "                data = json.load(json_file)\n",
    "            \n",
    "            # Extract mean scores and fold scores\n",
    "            file_name = os.path.splitext(file)[0]\n",
    "            file_name_stripped = file_name.rsplit('_nXcv', 1)[0]\n",
    "            print(file_name_stripped)\n",
    "\n",
    "\n",
    "            # perform GOEA\n",
    "            data_matched, data_goids, data_term_count, data_goea_all = matchgff2(data['selected_features'], namespace=['molecular_function'], depth_threshold=2, goea=True)\n",
    "            lasso_goea[file_name_stripped] = data_goea_all\n",
    "            \n",
    "\n",
    "            tmp += 1\n",
    "            if tmp == 2:\n",
    "              break\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/goea/Cellobiose_361_204_rt14_40_biop.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/goea/tyrosine_218_280_rt10_78_biop.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/goea/_671_2775803_7_254060113_biop.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/goea/Coumaroyl_quinic_acid__biop.csv\n",
      "Enrichment matrix saved to /home/t44p/PW_rawdata/results/full_lasso/goea/enrichment_matrix_biop.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/goea/Cellobiose_361_204_rt14_40_molf.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/goea/tyrosine_218_280_rt10_78_molf.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/goea/_671_2775803_7_254060113_molf.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/goea/Coumaroyl_quinic_acid__molf.csv\n",
      "Enrichment matrix saved to /home/t44p/PW_rawdata/results/full_lasso/goea/enrichment_matrix_molf.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/goea/Cellobiose_361_204_rt14_40_cellc.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/goea/tyrosine_218_280_rt10_78_cellc.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/goea/_671_2775803_7_254060113_cellc.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/goea/Coumaroyl_quinic_acid__cellc.csv\n",
      "Enrichment matrix saved to /home/t44p/PW_rawdata/results/full_lasso/goea/enrichment_matrix_cellc.csv\n"
     ]
    }
   ],
   "source": [
    "goea_results_to_file(lasso_goea, path='/home/t44p/PW_rawdata/results/full_lasso/goea/', obo_path=\"/home/t44p/PW_rawdata/go_obo/go.obo\", to_excel=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GCMS GOEA by Grouping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# Define the categories\n",
    "categories = {\n",
    "    \"Sugars_Carbohydrates\": [\n",
    "        \"Cellobiose_361_204_rt14_40\", \n",
    "        \"raffinose_437_451_rt16_91\", \n",
    "        \"glucose_160_rt9_81\", \n",
    "        \"inositol_myo_305_265_rt10_71\", \n",
    "        \"sucrose_437_361_rt13_77\", \n",
    "        \"hexose_307_217_rt9_58\", \n",
    "        \"glucose_1_phosphate_217_rt9_16\", \n",
    "        \"trehalose_alpha_alpha_191_169_\", \n",
    "        \"fructose_307_217_rt9_48\", \n",
    "        \"glucose_160_319_rt9_68\", \n",
    "        \"galactinol_204_191_rt15_38\", \n",
    "        \"Ribulose_5_phosphate_357_299_1\", \n",
    "        \"glucose_6_phosphate_160_387_rt\", \n",
    "        \"Cellobiose_361_204_or_maltose_\", \n",
    "        \"_6_phospho_gluconate\"\n",
    "    ],\n",
    "    \"Amino_Acids_and_Derivatives\": [\n",
    "        \"tyrosine_218_280_rt10_78\",\n",
    "        \"serine_204_218_rt5_84\", \n",
    "        \"isoleucin_158_233_rt_5_21\", \n",
    "        \"asparagine_116_188_rt9_00\", \n",
    "        \"ornithine_142_174_rt9_34\", \n",
    "        \"alanine_3_cyano_141_100_rt6_78\",\n",
    "        \"arginine_157_256_rt9_92\", \n",
    "        \"threonine_219_291_rt6_01\", \n",
    "        \"tryptophan_202_291_rt12_94\", \n",
    "        \"alanine_116_218_rt3_38\", \n",
    "        \"leucine_158_232_rt4_97\", \n",
    "        \"valine_144_218_rt4_42\", \n",
    "        \"asparagine_188_216_rt7_84\", \n",
    "        \"methionine_176_128_rt7_76\", \n",
    "        \"asparagine_188_216_rt7_45\", \n",
    "        \"lysine_156_174_rt10_07\", \n",
    "        \"glycine_102_147_rt3_70\", \n",
    "        \"aspartic_acid_232_218_rt7_48\", \n",
    "        \"histidine_154_254_rt11_10\", \n",
    "        \"glutamine_156_245_rt9_80\", \n",
    "        \"glycine_174_248_rt_5_31\", \n",
    "        \"homoserine_218_128_rt6_64\", \n",
    "        \"proline_142_130_rt_5_53\", \n",
    "        \"alanine_beta_248_290_rt6_44\", \n",
    "        \"ornithine_142_348_rt8_03\"\n",
    "    ],\n",
    "    \"Nucleotides_and_Derivatives\": [\n",
    "        \"adenine_264_279_rt11_12\", \n",
    "        \"adenosine_5_monophosphate_169_\"\n",
    "    ],\n",
    "    \"Organic_Acids\": [\n",
    "        \"threonic_acid_292_220_rt7_49\", \n",
    "        \"citric_acid_273_375_rt9_72\", \n",
    "        \"quinic_acid_255_345_rt9_45\", \n",
    "        \"Oxalic_acid_219_147_3_94\", \n",
    "        \"dehydroascorbic_acid_316_173_r\", \n",
    "        \"erythronic_acid_292_rt7_24\", \n",
    "        \"pyroglutamic_acid_156_258_rt8_\", \n",
    "        \"nonanoic_acid_117_215_rt6_19\", \n",
    "        \"malic_acid_233_245_rt7_22\", \n",
    "        \"glyceric_acid_292_189_rt5_63\", \n",
    "        \"butyric_acid_4_amino_174_304_r\", \n",
    "        \"quinic_acid_3_caffeoyl_trans_3\", \n",
    "        \"galactonic_acid_1_4_lacton_217\"\n",
    "    ],\n",
    "    \"Amines_Amides\": [\n",
    "        \"dopamine_174_426_rt11_11\", \n",
    "        \"ethanolamine_174_rt4_63\", \n",
    "        \"guanidine_146_171_rt4_33\"\n",
    "    ],\n",
    "    \"Vitamins_and_Cofactors\": [\n",
    "        \"nicotinic_acid_180_136_rt6_32\", \n",
    "        \"pantethaine_4_03_220_235\"\n",
    "    ],\n",
    "    \"Sterols\": [\n",
    "        \"beta_Sitosterol_1TMS_129_18_44\"\n",
    "    ],\n",
    "    \"Others\": [\n",
    "        \"phosphoric_acid_314_299_rt_5_4\",\n",
    "        \"glycerol_117_205_rt4_75\",\n",
    "        \"pyridine_2_hydroxy_152_167_rt3\", \n",
    "        \"inositol_1_phosphate_myo_299_3\", \n",
    "        \"urea_189_204_rt5_56\", \n",
    "        \"glycerol_3_phosphate_357_445_r\", \n",
    "        \"Hydroxylamine_133_3_07_No_need\", \n",
    "        \"glutamic_acid_246_363_rt8_31\"\n",
    "    ]\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "lcms_categories2 = {\n",
    "    \"Amino_Acids_and_Derivatives\": [\n",
    "        \"Argininosuccinic_acid_L\", \"Aspartyphenylalanine_1_L\", \"gamma_Glutamylisoleucine_\", \n",
    "        \"gamma_Glutamylvaline_\", \"Glycyl_L_phenylalanine_L\", \"Histidine_L\", \n",
    "        \"L_gamma_glutamyl_L_isoleucine\", \"N_gamma_L_Glutamyl_L_methionin\", \"Phenylalanylglutamic_acid_\", \n",
    "        \"Prolylglycine_L\", \"Prolyphenylalanine_or_Leucymet\", \"Tyrosine_L\", \"Kynurenine_free_base_L\", \"N_Decanoylglycine_\",\n",
    "        \"Phenylalanyaspartatic_acid_L\"\n",
    "    ],\n",
    "    \"Quinic_Acid_Derivatives\": [\n",
    "        \"_3_5_dicaffeoul_quinic_acid_L\", \"_1_3_Dicaffeoylquinic_acid_\", \"_4_O_p_Coumaroylquinic_acid_\", \n",
    "        \"_4_5_Di_O_caffeoylquinic_acid_\", \"Coumaroyl_quinic_acid_\", \"Caffeoylshikimic_acid_L\",\"Quinic_acid_161_05_\", \n",
    "        \"Quinic_acid_derivative_3_56\", \"Quinic_acid_derivative_with_he\"\n",
    "    ],\n",
    "    \"Phenolic_Compounds\": [\n",
    "        \"Caffeic_acid_1_L\", \"Caffeic_acid_glucoside_L\", \"Caffeoylglucarate_L\", \n",
    "        \"Chlorogenic_acid_\", \"Conduritol_B_Epoxide_L\", \"Dehydro_Ascorbic_acid_L\", \"Ferulic_acid_4_O_glucuronide_L\", \n",
    "        \"Feruloyl_di_glucoside_L\", \"Feruloylglucose_L\", \"Flavonoid_glycoside_\", \"Galloylglucose_L\", \n",
    "        \"Gentesic_acid_glucoside_L\", \"Kaempferol_3_O_glucoside_7_O_r\", \"Kaempferol_3_O_rhamnoside_7_O_\", \n",
    "        \"Kaempferol_3_O_rutinoside_\", \"Kaempferol_7_3G_glucosylgentio\", \"Kaempferol_7_O_Glucoside_1_L\", \n",
    "        \"Kaempferol_glucose_xylose_L\", \"Kaempferol_rhamnose_glucose\", \"Naringenin_chalcone_L\", \"p_coumaroyl_di_glucoside_L\", \n",
    "        \"p_coumaroyl_glucoside_L\", \"Phenolic_glycoside_\", \"Quercetin_3_7_di_glucoside_\", \"Quercetin_3_glucoside_3\", \n",
    "        \"Quercetin_7_4_O_diglucoside_\", \"Quercetin_glc_glc_rha\", \"Quercetin_glucose_glucose_rham\", \n",
    "        \"Quercetin_glucose\", \"Quercetin_glucose_xylose\", \"Quercitin_3_7_diglucoside_6_rh\", \"Quercitin_3_O_Glucoside_2_L\", \n",
    "        \"Resveratrol_3_4_diglucoside_L\", \"Sinapoylglucose_2_L\", \"Sinapoylglucose_L\",\"Catechol_\", \"_5_Chlorogenic_acid\", \"chlorgenic_acid_L\", \n",
    "        \"hydroxyjasmonic_acid_glucoside\", \"Hydroxy_jasmonic_acid_glucosid\"\n",
    "    ],\n",
    "    \"Sugars_and_Sugar_Derivatives\": [\n",
    "        \"alpha_D_Galacturonic_acid_1_ph\", \"alpha_L_Fucose_1_Phosphate\", \"Dihydrophaseic_acid_glucoside_\", \n",
    "        \"Dihydroxyacetone_phosphate_L\", \"Fructose_1_6_biosphasphate_L\", \"Glucaric_acid_1_4_lactone_L\", \n",
    "        \"Gluconic_acid_lactone_L\", \"Gluconic_acid_L\", \"Glutamic_acid_L\", \"Inosine_L\", \"Isorhamnetin_3_O_glucoside_L\", \n",
    "        \"Isorhamnetin_3_O_rutinoside\", \"Lactobionic_acid_L\", \"Maltotriose_\", \"naringenin_7_O_glucoside_1_L\", \n",
    "        \"naringenin_7_O_glucoside_3_L\", \"Pantothenic_acid_L\", \"Phaseoloidin\", \n",
    "        \"Pyroglutamic_acid_3_L\", \"Ribulose_5_phosphate_L\", \"Tartaric_acid_L\", \"Trehalose\", \n",
    "        \"Trehalose_phenolic_acid_\", \"Vanilloloside\", \"Vanilloside_L\", \"Zeatin_glucoside_L\", \"_1_O_Feruloyl_glucose_L\"\n",
    "    ],\n",
    "    \"Nucleotides_and_Derivatives\": [\n",
    "        \"_5_Deoxy_5_Methylthioadenosine\", \"Guanosine_\", \"Nicotinamide_adenine_dinucleot\", \n",
    "        \"Oxidized_glutathione_\",\n",
    "    ],\n",
    "    \"Terpenes_and_Triterpenes\": [\n",
    "        \"Diterpene_8_2\", \"Terpene_10_26\", \"Terpene_9_53\", \"Triterpene_10_43\", \"Triterpene_10_78\", \n",
    "        \"Triterpene_8_14\", \"Triterpene_8_21\", \"Triterpene_9_79\"\n",
    "    ],\n",
    "    \"Saponins\": [\n",
    "        \"Saponin_10_182\", \"Saponin_10_183\", \"Saponin_10_264\", \"Saponin_10_47\", \"Saponin_10_56\", \n",
    "        \"Saponin_10_67780171\", \"Saponin_10_72\", \"Saponin_8_02\", \"Saponin_8_42\", \"Saponin_8_83\", \n",
    "        \"Saponin_8_8\", \"Saponin_9_18\", \"Saponin_9_88\", \"Soyasaponin_A2_L\"\n",
    "    ],\n",
    "    \"Porphyrins\": [\n",
    "        \"porphobilinogen_2_L\"\n",
    "    ],\n",
    "    \"Flavonoids\": [\n",
    "        \"_Rutin\"\n",
    "    ],\n",
    "    \"Other\": [\n",
    "        \"_10_Formyltetrahydrofolate_L\", \"_3_Deoxy_D_manno_2_octulosonic\", \"_5_hydroxy_Ferulic_acid_Glucos2\", \n",
    "        \"_5_hydroxy_ferulic_acid_glucos\", \"_6_phosphogluconic_acid_L\", \"_7_Epi_12_hydroxyjasmonic_acid\", \n",
    "        \"_D_Glycero_alpha_D_Manno_Hepto\", \"_R_2_Phenylglycin\", \"Azelaic_acid_L\", \"Azukisaponin_VI_1_L\", \n",
    "        \"CGA_hexose_\", \"Urocanic_acid_L\"\n",
    "    ],\n",
    "    \"unidentified\": [\n",
    "        \"_1021_486521_9_141280832\",\n",
    "        \"_1021_48667462_9_66538870707\",\n",
    "        \"_1063_497231_10_58247297\",\n",
    "        \"_1063_529871_9_415000766\",\n",
    "        \"_1141_528589_8_043846139\",\n",
    "        \"_1151_551493_10_40677464\",\n",
    "        \"_1165_527084_8_995631942\",\n",
    "        \"_1183_537778_8_600427106\",\n",
    "        \"_1183_538043_8_673358308\",\n",
    "        \"_1191_542641_10_3213884\",\n",
    "        \"_1209_556345_10_14653119\",\n",
    "        \"_1241_580464_8_916526509\",\n",
    "        \"_1253_57885907_8_99579782362\",\n",
    "        \"_1283_593009_9_259004534\",\n",
    "        \"_1283_593348_9_160768845\",\n",
    "        \"_1313_601235_9_01121774\",\n",
    "        \"_303_0720727_3_335904942\",\n",
    "        \"_312_0943959_2_614228811\",\n",
    "        \"_351_1294837_5_809380415\",\n",
    "        \"_366_9939681_1_170199352\",\n",
    "        \"_374_1568694_3_674168475\",\n",
    "        \"_380_1288743_6_36896135\",\n",
    "        \"_427_1824019_5_95992104\",\n",
    "        \"_429_1765684_7_873841266\",\n",
    "        \"_441_1978503_6_218123415\",\n",
    "        \"_475_1821542_6_560220101\",\n",
    "        \"_487_2029578_5_338775921\",\n",
    "        \"_496_1502126_3_792994891\",\n",
    "        \"_512_1445532_3_132648409\",\n",
    "        \"_523_1663837_5_238225973\",\n",
    "        \"_531_2448764_7_209307815\",\n",
    "        \"_567_2828405_6_133036019\",\n",
    "        \"_641_171822_6_885700966\",\n",
    "        \"_658_1575839_4_528042834\",\n",
    "        \"_671_2775803_7_254060113\",\n",
    "        \"_821_3258533_9_228249892\",\n",
    "        \"_857_418157679_9_32212355859\",\n",
    "        \"_857_418693067_9_24167206032\",\n",
    "        \"_931_4518066_8_63280459\",\n",
    "        \"_933_4699157_8_265868462\",\n",
    "        \"_947_486080952_9_72692719903\",\n",
    "        \"_963_4798371_8_624190173\",\n",
    "        \"_987_4809544_10_24318525\"\n",
    "    ]\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "lcms_categories = {\n",
    "    \"lcms\":[\n",
    "        \"_10_Formyltetrahydrofolate_L\",\n",
    "        \"_3_5_dicaffeoul_quinic_acid_L\",\n",
    "        \"_1_3_Dicaffeoylquinic_acid_\",\n",
    "        \"_1_O_Feruloyl_glucose_L\",\n",
    "        \"_4_O_p_Coumaroylquinic_acid_\",\n",
    "        \"_4_5_Di_O_caffeoylquinic_acid_\",\n",
    "        \"_5_Chlorogenic_acid\",\n",
    "        \"_3_Deoxy_D_manno_2_octulosonic\",\n",
    "        \"_5_Deoxy_5_Methylthioadenosine\",\n",
    "        \"_5_hydroxy_Ferulic_acid_Glucos2\",\n",
    "        \"_5_hydroxy_ferulic_acid_glucos\",\n",
    "        \"_6_phosphogluconic_acid_L\",\n",
    "        \"_7_Epi_12_hydroxyjasmonic_acid\",\n",
    "        \"alpha_D_Galacturonic_acid_1_ph\",\n",
    "        \"alpha_L_Fucose_1_Phosphate\",\n",
    "        \"Argininosuccinic_acid_L\",\n",
    "        \"Aspartyphenylalanine_1_L\",\n",
    "        \"Azelaic_acid_L\",\n",
    "        \"Azukisaponin_VI_1_L\",\n",
    "        \"Caffeic_acid_1_L\",\n",
    "        \"Caffeic_acid_glucoside_L\",\n",
    "        \"Caffeoylglucarate_L\",\n",
    "        \"Caffeoylshikimic_acid_L\",\n",
    "        \"Catechol_\",\n",
    "        \"CGA_hexose_\",\n",
    "        \"chlorgenic_acid_L\",\n",
    "        \"Chlorogenic_acid_\",\n",
    "        \"Conduritol_B_Epoxide_L\",\n",
    "        \"Coumaroyl_quinic_acid_\",\n",
    "        \"Dehydro_Ascorbic_acid_L\",\n",
    "        \"_D_Glycero_alpha_D_Manno_Hepto\",\n",
    "        \"Dihydrophaseic_acid_glucoside_\",\n",
    "        \"Dihydroxyacetone_phosphate_L\",\n",
    "        \"Diterpene_8_2\",\n",
    "        \"Ferulic_acid_4_O_glucuronide_L\",\n",
    "        \"Feruloyl_di_glucoside_L\",\n",
    "        \"Feruloylglucose_L\",\n",
    "        \"Flavonoid_glycoside_\",\n",
    "        \"Fructose_1_6_biosphasphate_L\",\n",
    "        \"Galloylglucose_L\",\n",
    "        \"gamma_Glutamylisoleucine_\",\n",
    "        \"gamma_Glutamylvaline_\",\n",
    "        \"Gentesic_acid_glucoside_L\",\n",
    "        \"Glucaric_acid_1_4_lactone_L\",\n",
    "        \"Gluconic_acid_lactone_L\",\n",
    "        \"Gluconic_acid_L\",\n",
    "        \"Glutamic_acid_L\",\n",
    "        \"Glycyl_L_phenylalanine_L\",\n",
    "        \"Guanosine_\",\n",
    "        \"Histidine_L\",\n",
    "        \"hydroxyjasmonic_acid_glucoside\",\n",
    "        \"Hydroxy_jasmonic_acid_glucosid\",\n",
    "        \"Inosine_L\",\n",
    "        \"Isorhamnetin_3_O_glucoside_L\",\n",
    "        \"Isorhamnetin_3_O_rutinoside\",\n",
    "        \"Kaempferol_3_O_glucoside_7_O_r\",\n",
    "        \"Kaempferol_3_O_rhamnoside_7_O_\",\n",
    "        \"Kaempferol_3_O_rutinoside_\",\n",
    "        \"Kaempferol_7_3G_glucosylgentio\",\n",
    "        \"Kaempferol_7_O_Glucoside_1_L\",\n",
    "        \"Kaempferol_glucose_xylose_L\",\n",
    "        \"Kaempferol_rhamnose_glucose\",\n",
    "        \"Kynurenine_free_base_L\",\n",
    "        \"Lactobionic_acid_L\",\n",
    "        \"L_gamma_glutamyl_L_isoleucine\",\n",
    "        \"Maltotriose_\",\n",
    "        \"naringenin_7_O_glucoside_1_L\",\n",
    "        \"naringenin_7_O_glucoside_3_L\",\n",
    "        \"Naringenin_chalcone_L\",\n",
    "        \"N_Decanoylglycine_\",\n",
    "        \"N_gamma_L_Glutamyl_L_methionin\",\n",
    "        \"Nicotinamide_adenine_dinucleot\",\n",
    "        \"Oxidized_glutathione_\",\n",
    "        \"Pantothenic_acid_L\",\n",
    "        \"p_coumaroyl_di_glucoside_L\",\n",
    "        \"p_coumaroyl_glucoside_L\",\n",
    "        \"Phaseoloidin\",\n",
    "        \"Phenolic_glycoside_\",\n",
    "        \"Phenylalanyaspartatic_acid_L\",\n",
    "        \"Phenylalanylglutamic_acid_\",\n",
    "        \"porphobilinogen_2_L\",\n",
    "        \"Prolylglycine_L\",\n",
    "        \"Prolyphenylalanine_or_Leucymet\",\n",
    "        \"Pyroglutamic_acid_3_L\",\n",
    "        \"Quercetin_3_7_di_glucoside_\",\n",
    "        \"Quercetin_3_glucoside_3\",\n",
    "        \"Quercetin_7_4_O_diglucoside_\",\n",
    "        \"Quercetin_glc_glc_rha\",\n",
    "        \"Quercetin_glucose_glucose_rham\",\n",
    "        \"Quercetin_glucose\",\n",
    "        \"Quercetin_glucose_xylose\",\n",
    "        \"Quercitin_3_7_diglucoside_6_rh\",\n",
    "        \"Quercitin_3_O_Glucoside_2_L\",\n",
    "        \"Quinic_acid_161_05_\",\n",
    "        \"Quinic_acid_derivative_3_56\",\n",
    "        \"Quinic_acid_derivative_with_he\",\n",
    "        \"_R_2_Phenylglycin\",\n",
    "        \"Resveratrol_3_4_diglucoside_L\",\n",
    "        \"Ribulose_5_phosphate_L\",\n",
    "        \"_Rutin\",\n",
    "        \"Saponin_10_182\",\n",
    "        \"Saponin_10_183\",\n",
    "        \"Saponin_10_264\",\n",
    "        \"Saponin_10_47\",\n",
    "        \"Saponin_10_56\",\n",
    "        \"Saponin_10_67780171\",\n",
    "        \"Saponin_10_72\",\n",
    "        \"Saponin_8_02\",\n",
    "        \"Saponin_8_42\",\n",
    "        \"Saponin_8_83\",\n",
    "        \"Saponin_8_8\",\n",
    "        \"Saponin_9_18\",\n",
    "        \"Saponin_9_88\",\n",
    "        \"Sinapoylglucose_2_L\",\n",
    "        \"Sinapoylglucose_L\",\n",
    "        \"Soyasaponin_A2_L\",\n",
    "        \"Tartaric_acid_L\",\n",
    "        \"Terpene_10_26\",\n",
    "        \"Terpene_9_53\",\n",
    "        \"Trehalose\",\n",
    "        \"Trehalose_phenolic_acid_\",\n",
    "        \"Triterpene_10_43\",\n",
    "        \"Triterpene_10_78\",\n",
    "        \"Triterpene_8_14\",\n",
    "        \"Triterpene_8_21\",\n",
    "        \"Triterpene_9_79\",\n",
    "        \"Tyrosine_L\",\n",
    "        \"Urocanic_acid_L\",\n",
    "        \"Vanilloloside\",\n",
    "        \"Vanilloside_L\",\n",
    "        \"Zeatin_glucoside_L\"\n",
    "    ],\n",
    "    \"unidentified\": [\n",
    "        \"_1021_486521_9_141280832\",\n",
    "        \"_1021_48667462_9_66538870707\",\n",
    "        \"_1063_497231_10_58247297\",\n",
    "        \"_1063_529871_9_415000766\",\n",
    "        \"_1141_528589_8_043846139\",\n",
    "        \"_1151_551493_10_40677464\",\n",
    "        \"_1165_527084_8_995631942\",\n",
    "        \"_1183_537778_8_600427106\",\n",
    "        \"_1183_538043_8_673358308\",\n",
    "        \"_1191_542641_10_3213884\",\n",
    "        \"_1209_556345_10_14653119\",\n",
    "        \"_1241_580464_8_916526509\",\n",
    "        \"_1253_57885907_8_99579782362\",\n",
    "        \"_1283_593009_9_259004534\",\n",
    "        \"_1283_593348_9_160768845\",\n",
    "        \"_1313_601235_9_01121774\",\n",
    "        \"_303_0720727_3_335904942\",\n",
    "        \"_312_0943959_2_614228811\",\n",
    "        \"_351_1294837_5_809380415\",\n",
    "        \"_366_9939681_1_170199352\",\n",
    "        \"_374_1568694_3_674168475\",\n",
    "        \"_380_1288743_6_36896135\",\n",
    "        \"_427_1824019_5_95992104\",\n",
    "        \"_429_1765684_7_873841266\",\n",
    "        \"_441_1978503_6_218123415\",\n",
    "        \"_475_1821542_6_560220101\",\n",
    "        \"_487_2029578_5_338775921\",\n",
    "        \"_496_1502126_3_792994891\",\n",
    "        \"_512_1445532_3_132648409\",\n",
    "        \"_523_1663837_5_238225973\",\n",
    "        \"_531_2448764_7_209307815\",\n",
    "        \"_567_2828405_6_133036019\",\n",
    "        \"_641_171822_6_885700966\",\n",
    "        \"_658_1575839_4_528042834\",\n",
    "        \"_671_2775803_7_254060113\",\n",
    "        \"_821_3258533_9_228249892\",\n",
    "        \"_857_418157679_9_32212355859\",\n",
    "        \"_857_418693067_9_24167206032\",\n",
    "        \"_931_4518066_8_63280459\",\n",
    "        \"_933_4699157_8_265868462\",\n",
    "        \"_947_486080952_9_72692719903\",\n",
    "        \"_963_4798371_8_624190173\",\n",
    "        \"_987_4809544_10_24318525\"\n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amino_Acids_and_Derivatives\n",
      "Quinic_Acid_Derivatives\n",
      "Phenolic_Compounds\n",
      "Sugars_and_Sugar_Derivatives\n",
      "Nucleotides_and_Derivatives\n",
      "Terpenes_and_Triterpenes\n",
      "Saponins\n",
      "Porphyrins\n",
      "Flavonoids\n",
      "Other\n",
      "unidentified\n",
      "11 11 174\n"
     ]
    }
   ],
   "source": [
    "c = 0\n",
    "kl = []\n",
    "cat2 = []\n",
    "for k,v in lcms_categories2.items():\n",
    "    print(k)\n",
    "    kl.append(k)\n",
    "    for m in v:\n",
    "        cat2.append(m)\n",
    "    c+=1\n",
    "\n",
    "print(c, len(kl), len(cat2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lcms\n",
      "unidentified\n",
      "2 2 174\n"
     ]
    }
   ],
   "source": [
    "c = 0\n",
    "kl = []\n",
    "cat1 = []\n",
    "for k,v in lcms_categories.items():\n",
    "    print(k)\n",
    "    kl.append(k)\n",
    "    for m in v:\n",
    "        cat1.append(m)\n",
    "    c+=1\n",
    "\n",
    "print(c, len(kl), len(cat1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(cat2).difference(cat1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# Define the categories\n",
    "categories2 = {\n",
    "    \"Sugars_Carbohydrates\": [\n",
    "        \"Cellobiose_361_204_rt14_40\",\n",
    "        \"trehalose_alpha_alpha_191_169_\"\n",
    "    ],\n",
    "    \"Others\": [\n",
    "        \"phosphoric_acid_314_299_rt_5_4\",\n",
    "        \"glycerol_117_205_rt4_75\"\n",
    "    ]\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def goea_for_category(category, metabolites, path):\n",
    "    #excel_writer = pd.ExcelWriter(f'{outpath}{category}.xlsx', engine='xlsxwriter')\n",
    "    goea_results = {}\n",
    "    print(category)\n",
    "    for m in metabolites:\n",
    "        print(os.path.join(path, f\"{m}_nXcv.json\"))\n",
    "        file_path = os.path.join(path, f\"{m}_nXcv.json\")\n",
    "        with open(file_path, 'r') as json_file:\n",
    "            data = json.load(json_file)\n",
    "        \n",
    "\n",
    "        # perform GOEA\n",
    "        data_matched, data_goids, data_term_count, data_goea_all = matchgff2(data['selected_features'], namespace=['molecular_function'], depth_threshold=2, goea=True)\n",
    "        goea_results[m] = data_goea_all\n",
    "    \n",
    "    return goea_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory created: /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Sugars_Carbohydrates/\n",
      "/home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Sugars_Carbohydrates/\n",
      "Sugars_Carbohydrates\n",
      "/home/t44p/PW_rawdata/results/full_lasso/gcms/Cellobiose_361_204_rt14_40_nXcv.json\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  1%    405 of 30,522 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 405 IDs.\n",
      "100%    405 of    405 study items found in association\n",
      "100%    405 of    405 study items found in population(30522)\n",
      "Calculating 1,693 uncorrected p-values using fisher_scipy_stats\n",
      "   1,693 terms are associated with    362 of 30,522 population items\n",
      "   1,693 terms are associated with    362 of    405 study items\n",
      "  METHOD fdr_bh:\n",
      "   1,693 GO terms found significant (< 0.05=alpha) (1693 enriched +   0 purified): statsmodels fdr_bh\n",
      "     362 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/results/full_lasso/gcms/trehalose_alpha_alpha_191_169__nXcv.json\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  2%    592 of 30,522 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 592 IDs.\n",
      "100%    592 of    592 study items found in association\n",
      "100%    592 of    592 study items found in population(30522)\n",
      "Calculating 2,385 uncorrected p-values using fisher_scipy_stats\n",
      "   2,385 terms are associated with    519 of 30,522 population items\n",
      "   2,385 terms are associated with    519 of    592 study items\n",
      "  METHOD fdr_bh:\n",
      "   2,385 GO terms found significant (< 0.05=alpha) (2385 enriched +   0 purified): statsmodels fdr_bh\n",
      "     519 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Sugars_Carbohydrates/Cellobiose_361_204_rt14_40_biop.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Sugars_Carbohydrates/trehalose_alpha_alpha_191_169__biop.csv\n",
      "Enrichment matrix saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Sugars_Carbohydrates/enrichment_matrix_biop.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Sugars_Carbohydrates/Cellobiose_361_204_rt14_40_molf.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Sugars_Carbohydrates/trehalose_alpha_alpha_191_169__molf.csv\n",
      "Enrichment matrix saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Sugars_Carbohydrates/enrichment_matrix_molf.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Sugars_Carbohydrates/Cellobiose_361_204_rt14_40_cellc.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Sugars_Carbohydrates/trehalose_alpha_alpha_191_169__cellc.csv\n",
      "Enrichment matrix saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Sugars_Carbohydrates/enrichment_matrix_cellc.csv\n",
      "Directory created: /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Others/\n",
      "/home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Others/\n",
      "Others\n",
      "/home/t44p/PW_rawdata/results/full_lasso/gcms/phosphoric_acid_314_299_rt_5_4_nXcv.json\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  1%    388 of 30,522 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 388 IDs.\n",
      "100%    388 of    388 study items found in association\n",
      "100%    388 of    388 study items found in population(30522)\n",
      "Calculating 1,672 uncorrected p-values using fisher_scipy_stats\n",
      "   1,672 terms are associated with    347 of 30,522 population items\n",
      "   1,672 terms are associated with    347 of    388 study items\n",
      "  METHOD fdr_bh:\n",
      "   1,672 GO terms found significant (< 0.05=alpha) (1672 enriched +   0 purified): statsmodels fdr_bh\n",
      "     347 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/results/full_lasso/gcms/glycerol_117_205_rt4_75_nXcv.json\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  2%    744 of 30,522 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 744 IDs.\n",
      "100%    744 of    744 study items found in association\n",
      "100%    744 of    744 study items found in population(30522)\n",
      "Calculating 2,836 uncorrected p-values using fisher_scipy_stats\n",
      "   2,836 terms are associated with    655 of 30,522 population items\n",
      "   2,836 terms are associated with    655 of    744 study items\n",
      "  METHOD fdr_bh:\n",
      "   2,836 GO terms found significant (< 0.05=alpha) (2836 enriched +   0 purified): statsmodels fdr_bh\n",
      "     655 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Others/phosphoric_acid_314_299_rt_5_4_biop.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Others/glycerol_117_205_rt4_75_biop.csv\n",
      "Enrichment matrix saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Others/enrichment_matrix_biop.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Others/phosphoric_acid_314_299_rt_5_4_molf.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Others/glycerol_117_205_rt4_75_molf.csv\n",
      "Enrichment matrix saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Others/enrichment_matrix_molf.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Others/phosphoric_acid_314_299_rt_5_4_cellc.csv\n",
      "Saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Others/glycerol_117_205_rt4_75_cellc.csv\n",
      "Enrichment matrix saved to /home/t44p/PW_rawdata/results/full_lasso/gcms/goea/Others/enrichment_matrix_cellc.csv\n"
     ]
    }
   ],
   "source": [
    "path_gc = \"/home/t44p/PW_rawdata/results/full_lasso/gcms/\"\n",
    "\n",
    "\n",
    "for category, metabolites in categories2.items():\n",
    "    # check if output directories exist otherwise createt them\n",
    "    category_path = os.path.join(path_gc, 'goea', f\"{category}/\")\n",
    "    if not os.path.exists(category_path):\n",
    "        try:\n",
    "            # If the path does not exist, create the directory\n",
    "            os.makedirs(category_path)\n",
    "            print(f\"Directory created: {category_path}\")\n",
    "        except Exception as e:\n",
    "            # If there is an error in creating the directory, raise an error\n",
    "            raise Exception(f\"Error creating directory: {category_path}. Error: {e}\")\n",
    "    else:\n",
    "        print(f\"Path already exists: {category_path}\")\n",
    "\n",
    "    print(category_path)\n",
    "\n",
    "    # do GOEA for each metabolite in the category and save them into \n",
    "    # dictionary where each key is the prefix before _nXcv.json\n",
    "    lasso_goea = goea_for_category(category, metabolites, path=path_gc)\n",
    "\n",
    "    # save the goea results to .csv and excel tables and create enrichment matrix .csv\n",
    "    goea_results_to_file(lasso_goea, path=category_path, obo_path=\"/home/t44p/PW_rawdata/go_obo/go.obo\", to_excel=True)\n",
    "\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py3-6",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
