{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use kernel py3-6\n",
    "#import goatools\n",
    "from goatools import obo_parser\n",
    "import re\n",
    "import json\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from goatools.go_enrichment import GOEnrichmentStudy\n",
    "import matplotlib.pyplot as plt\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def matchgff2(feature, gff_file='/home/t44p/PW_rawdata/Transciptome_GenomeAnnotation/Xele_annotated2_gff_export2.gff', obo_path=\"/home/t44p/PW_rawdata/go_obo/go.obo\", namespace=None, depth_threshold=0, goea=False):\n",
    "    \"\"\"\n",
    "    Searches a GFF (General Feature Format) file for specific features and retrieves associated Gene Ontology (GO) terms,\n",
    "    with optional filtering by GO term namespace and depth.\n",
    "\n",
    "    Parameters:\n",
    "    ----------\n",
    "    feature : list or iterable\n",
    "        An iterable of strings representing the features to search for in the GFF file.\n",
    "\n",
    "    gff_file : str, optional\n",
    "        The file path to the GFF file. Defaults to a predefined path.\n",
    "\n",
    "    obo_path : str, optional\n",
    "        The file path to the Gene Ontology .obo file. Defaults to a predefined path.\n",
    "\n",
    "    namespace : list of str or None, optional\n",
    "        List of GO term namespaces to filter term count results. Valid options are 'biological_process', \n",
    "        'molecular_function', 'cellular_component', or None. Defaults to None (no filtering).\n",
    "\n",
    "    depth_threshold : int, optional\n",
    "        Minimum depth of GO terms to include in the results. Defaults to 0 (no filtering).\n",
    "\n",
    "    goea : bool, optional\n",
    "        Whether to perform Gene Ontology Enrichment Analysis. Defaults to False.\n",
    "\n",
    "    Returns:\n",
    "    -------\n",
    "    tuple\n",
    "        A tuple containing:\n",
    "        1. A dictionary with features as keys and lists of lines from the GFF file as values.\n",
    "        2. A dictionary mapping features to their associated GO terms.\n",
    "        3. A dictionary of GO term counts.\n",
    "        4. GOEA results (if goea is True), filtered by the specified depth threshold for all namespaces\n",
    "\n",
    "    Notes:\n",
    "    -----\n",
    "    - The function uses regular expressions for precise matching of features.\n",
    "    - It extracts GO IDs from matched lines and retrieves their corresponding names and namespaces.\n",
    "    - It is possible for the rawcounts to filter for namespaces \n",
    "    - If `goea` is True, the function performs GO Enrichment Analysis and returns the results filtered by the specified depth threshold but not for namespaces. \n",
    "    - The Enrichment analysis will be performed across all namespaces, filtering parameter \"namespace\" will have no effect on the enrichment analysis\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    valid_namespaces = {'biological_process', 'molecular_function', 'cellular_component', None}\n",
    "    # Check if namespace is a list containing only valid elements\n",
    "    if isinstance(namespace, list) and not all(ns in valid_namespaces for ns in namespace):\n",
    "        raise ValueError(\"Invalid namespace provided. Valid options are 'biological_process', \"\n",
    "                         \"'molecular_function', 'cellular_component', or a list containing any of these. \"\n",
    "                         \"You can also use None for no filtering.\")\n",
    "\n",
    "    with open(gff_file, 'r') as file:\n",
    "        go_ontology = obo_parser.GODag(obo_path)\n",
    "        \n",
    "        lines_where_feat_found = {}\n",
    "        go_ids = {}\n",
    "        background_genes = []\n",
    "        go_term_count = {}\n",
    "\n",
    "         # Find the depth of each GO term\n",
    "        go_depths = {go_id: go_term.depth for go_id, go_term in go_ontology.items()}\n",
    "\n",
    "        # construct background genes\n",
    "        if goea:\n",
    "            for line in file:\n",
    "                if not line.lstrip().startswith('#'):\n",
    "                    background_genes.append(line.split('\\t')[0])\n",
    "\n",
    "        for feat in feature:\n",
    "            file.seek(0)  # reset file pointer to the beginning for each feature\n",
    "            lines_where_feat_found[feat] = []\n",
    "            go_ids[feat] = {}\n",
    "            pattern = re.compile(re.escape(feat) + r'\\t')  # exact match followed by a tab\n",
    "            for line in file:\n",
    "                if pattern.search(line):\n",
    "                    lines_where_feat_found[feat].append(line.strip())  # Store the line (as a string) if feature is found\n",
    "                    # Extract GO id\n",
    "                    match = re.search(r\"Ontology_id=([GO:\\d,]+)\", line.strip())\n",
    "                    if match:\n",
    "                        ids = match.group(1).split(',')\n",
    "                        # Map Terms to Ids and Count Occurrences\n",
    "                        for id in ids:\n",
    "                            term = go_ontology.get(id)\n",
    "                            if term is not None:\n",
    "                                go_ids[feat][id] = {'name': term.name, 'namespace': term.namespace}\n",
    "\n",
    "                                if namespace is None or term.namespace in namespace and go_ontology[id].depth >= depth_threshold:\n",
    "                                    # Count Occurrences\n",
    "                                    if id in go_term_count:\n",
    "                                        go_term_count[id] = (term.name, go_term_count[id][1] + 1, term.namespace)\n",
    "                                    else:\n",
    "                                        go_term_count[id] = (term.name, 1, term.namespace)\n",
    "                            else:\n",
    "                                go_ids[feat][id] = {'name': None, 'namespace': None}\n",
    "                                if id not in go_term_count:\n",
    "                                    go_term_count[id] = (None, 1)\n",
    "        if goea:\n",
    "            print(\"GO Enrichment Analysis >>\")\n",
    "            goea_obj = GOEnrichmentStudy(\n",
    "                background_genes,\n",
    "                go_ids,  # This needs to be a dict mapping gene IDs to a set of GO IDs\n",
    "                go_ontology,\n",
    "                propagate_counts=False,\n",
    "                alpha=0.05,  # significance level for the statistical test\n",
    "                methods=['fdr_bh']  # correction method for multiple testing\n",
    "            )\n",
    "            goea_result = goea_obj.run_study(go_ids.keys())\n",
    "\n",
    "            # filter based on depth\n",
    "            filtered_goea_results = [res for res in goea_result if res.goterm.depth >= depth_threshold]\n",
    "            return lines_where_feat_found, go_ids, go_term_count, filtered_goea_results\n",
    "\n",
    "\n",
    "        return lines_where_feat_found, go_ids, go_term_count\n",
    "    \n",
    "\n",
    "def tabulate(term_count_dict, sort=True):\n",
    "    \"\"\"\n",
    "    Prints a tabulated view of the term counts from a dictionary, such as the one returned by matchgff2.\n",
    "\n",
    "    Parameters:\n",
    "    ----------\n",
    "    term_count_dict : dict\n",
    "        A dictionary where keys are Gene Ontology (GO) IDs, and values are tuples containing the GO term name, \n",
    "        the count of occurrences, and optionally the namespace. \n",
    "        The structure is typically: {GO_ID: (GO_Term, Count, Namespace)}.\n",
    "\n",
    "    sort : bool, optional\n",
    "        Whether to sort the output based on the count of occurrences of each GO term. Defaults to True.\n",
    "\n",
    "    Description:\n",
    "    ------------\n",
    "    This function iterates through the term_count_dict and prints each GO term's count, ID, name, and namespace \n",
    "    in a tabular format. It can optionally sort these terms based on their count in descending order.\n",
    "\n",
    "    Note:\n",
    "    -----\n",
    "    The function is primarily a utility for visualizing the output of the matchgff2 function. It does not return any value \n",
    "    but prints the information directly to the console.\n",
    "\n",
    "    Example Usage:\n",
    "    --------------\n",
    "    go_term_count = {'GO:0000001': ('term1', 5, 'biological_process'), 'GO:0000002': ('term2', 3, 'cellular_component')}\n",
    "    tabulate(go_term_count, sort=True)\n",
    "    \"\"\"\n",
    "    print(f\"count\\tGO ID\\tGO Term\\tnamespace\")\n",
    "\n",
    "    # Conditionally sort the dictionary if required\n",
    "    items = sorted(term_count_dict.items(), key=lambda x: x[1][1], reverse=True) if sort else term_count_dict.items()\n",
    "\n",
    "    for goid, values in items:\n",
    "        count = values[1]\n",
    "        term = values[0] if values[0] is not None else \"N/A\"\n",
    "        namespace = values[2] if len(values) > 2 else \"N/A\"\n",
    "\n",
    "        print(f\"{count}\\t{goid}\\t{term}\\t{namespace}\")\n",
    "    #return \n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "def goea_results_to_file_old(goea_results, path='./lasso_models/10xKfold_lasso_output/goea/', to_excel=False):\n",
    "    \"\"\"\n",
    "    Save Gene Ontology Enrichment Analysis (GOEA) results to CSV or Excel files.\n",
    "\n",
    "    Parameters:\n",
    "    ----------\n",
    "    goea_results : dict\n",
    "        A dictionary where keys are prefixes for file names and values are lists of GOEA result objects.\n",
    "    \n",
    "    path : str, optional\n",
    "        The directory path where the output files will be saved. Defaults to './lasso_models/10xKfold_lasso_output/goea/'.\n",
    "\n",
    "    to_excel : bool, optional\n",
    "        If True, results are saved to Excel files; otherwise, results are saved to CSV files. Defaults to False.\n",
    "\n",
    "    Description:\n",
    "    ------------\n",
    "    This function iterates through the goea_results dictionary. For each key-value pair, it creates a DataFrame \n",
    "    from the GOEA results and saves it as either a CSV or an Excel file. The function supports appending \n",
    "    to existing Excel files using the 'openpyxl' engine.\n",
    "\n",
    "    Note:\n",
    "    -----\n",
    "    Ensure that the 'openpyxl' library is installed if saving to Excel files.\n",
    "\n",
    "    Example Usage:\n",
    "    --------------\n",
    "    goea_results = {\n",
    "        \"gluc_goea\": gluc_goea_biop_sig,\n",
    "        \"suc_goea\": suc_goea_biop_sig\n",
    "    }\n",
    "    goea_results_to_file(goea_results, to_excel=True)\n",
    "    \"\"\"\n",
    "    # Mapping for suffixes\n",
    "    suffix_mapping = {'BP': 'biop', 'MF': 'molf', 'CC': 'cellc'}\n",
    "\n",
    "    \n",
    "    for file_prefix, res in goea_results.items():\n",
    "        for ns in ['BP', 'MF', 'CC']:\n",
    "            suffix = suffix_mapping[ns]\n",
    "\n",
    "            # Filtering relevant records\n",
    "            relevant_records = [r for r in res if r.NS == ns]\n",
    "\n",
    "            # Creating a DataFrame\n",
    "            data = {\n",
    "                \"GO_ID\": [r.goterm.id for r in relevant_records],\n",
    "                \"level\": [r.goterm.level for r in relevant_records],\n",
    "                \"depth\": [r.goterm.depth for r in relevant_records],\n",
    "                \"GO_Term\": [r.goterm.name for r in relevant_records],\n",
    "                \"study_count\": [r.study_count for r in relevant_records],\n",
    "                \"study_n\": [r.study_n for r in relevant_records],\n",
    "                \"pop_count\": [r.pop_count for r in relevant_records],\n",
    "                \"pop_n\": [r.pop_n for r in relevant_records],\n",
    "                \"enrichment\": [r.enrichment for r in relevant_records],\n",
    "                \"adj_p_fdr_bh\": [r.p_fdr_bh for r in relevant_records]\n",
    "            }\n",
    "        \n",
    "            df = pd.DataFrame(data)\n",
    "            # Saving to CSV\n",
    "            if not to_excel:\n",
    "                file_name = f\"{path}{file_prefix}_{suffix}.csv\"\n",
    "                df.to_csv(file_name, index=False)\n",
    "                print(f\"Saved to {file_name}\")\n",
    "            else:\n",
    "                path_suffix = f\"{path}goea_{suffix}.xlsx\"\n",
    "                if os.path.exists(path_suffix):\n",
    "                    x_mode = 'a'\n",
    "                else:\n",
    "                    x_mode = 'w'\n",
    "                    \n",
    "                with pd.ExcelWriter(path_suffix, mode=x_mode, engine='openpyxl') as writer:\n",
    "                    sheet = file_prefix\n",
    "                    print(f\"to excel, {sheet} for {path_suffix}\")\n",
    "                    df.to_excel(writer, sheet_name=sheet, index=False)\n",
    "\n",
    "\n",
    "\n",
    "def parse_obo_file(obo_path):\n",
    "    go_terms = {}\n",
    "    with open(obo_path, 'r') as file:\n",
    "        for line in file:\n",
    "            if line.startswith('id: '):\n",
    "                current_id = line.strip().split(' ')[1]\n",
    "            elif line.startswith('namespace: '):\n",
    "                current_namespace = line.strip().split(' ')[1]\n",
    "                go_terms[current_id] = current_namespace\n",
    "    return go_terms\n",
    "\n",
    "def goea_results_to_file(goea_results, obo_path, path='./lasso_models/10xKfold_lasso_output/goea/', to_excel=False):\n",
    "    # Load GO terms from the obo file\n",
    "    go_terms = parse_obo_file(obo_path)\n",
    "\n",
    "    # Mapping for suffixes\n",
    "    suffix_mapping = {'BP': 'biop', 'MF': 'molf', 'CC': 'cellc'}\n",
    "\n",
    "    for ns in ['BP', 'MF', 'CC']:\n",
    "        enrichment_matrix = {key: {go_id: 0 for go_id in go_terms if go_terms[go_id] == ns} for key in goea_results}\n",
    "\n",
    "        for file_prefix, res in goea_results.items():\n",
    "            suffix = suffix_mapping[ns]\n",
    "\n",
    "            # Filtering relevant records\n",
    "            relevant_records = [r for r in res if r.NS == ns]\n",
    "\n",
    "            # Creating a DataFrame\n",
    "            data = {\n",
    "                \"GO_ID\": [r.goterm.id for r in relevant_records],\n",
    "                \"level\": [r.goterm.level for r in relevant_records],\n",
    "                \"depth\": [r.goterm.depth for r in relevant_records],\n",
    "                \"GO_Term\": [r.goterm.name for r in relevant_records],\n",
    "                \"study_count\": [r.study_count for r in relevant_records],\n",
    "                \"study_n\": [r.study_n for r in relevant_records],\n",
    "                \"pop_count\": [r.pop_count for r in relevant_records],\n",
    "                \"pop_n\": [r.pop_n for r in relevant_records],\n",
    "                \"enrichment\": [r.enrichment for r in relevant_records],\n",
    "                \"adj_p_fdr_bh\": [r.p_fdr_bh for r in relevant_records]\n",
    "            }\n",
    "\n",
    "            for record in relevant_records:\n",
    "                if record.enrichment == 'e':\n",
    "                    enrichment_matrix[file_prefix][record.goterm.id] = 1\n",
    "\n",
    "            df = pd.DataFrame(data)\n",
    "            # Saving to CSV\n",
    "            if not to_excel:\n",
    "                file_name = f\"{path}{file_prefix}_{suffix}.csv\"\n",
    "                df.to_csv(file_name, index=False)\n",
    "                print(f\"Saved to {file_name}\")\n",
    "            else:\n",
    "                path_suffix = f\"{path}goea_{suffix}.xlsx\"\n",
    "                if os.path.exists(path_suffix):\n",
    "                    x_mode = 'a'\n",
    "                else:\n",
    "                    x_mode = 'w'\n",
    "                    \n",
    "                with pd.ExcelWriter(path_suffix, mode=x_mode, engine='openpyxl') as writer:\n",
    "                    sheet = file_prefix\n",
    "                    df.to_excel(writer, sheet_name=sheet, index=False)\n",
    "\n",
    "        # Ensure that all GO IDs are represented and convert the matrix to DataFrame\n",
    "        enrichment_df = pd.DataFrame(enrichment_matrix)\n",
    "        \n",
    "        # The DataFrame should now be fully populated with 0s and 1s. Converting to integer type should not be an issue.\n",
    "        enrichment_df = enrichment_df.fillna(0).astype(int)\n",
    "\n",
    "        # Save the DataFrame\n",
    "        enrichment_file_name = f\"{path}enrichment_matrix_{suffix}.csv\"\n",
    "        enrichment_df.to_csv(enrichment_file_name)\n",
    "        print(f\"Enrichment matrix saved to {enrichment_file_name}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  1%    405 of 30,522 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 405 IDs.\n",
      "100%    405 of    405 study items found in association\n",
      "100%    405 of    405 study items found in population(30522)\n",
      "Calculating 1,693 uncorrected p-values using fisher_scipy_stats\n",
      "   1,693 terms are associated with    362 of 30,522 population items\n",
      "   1,693 terms are associated with    362 of    405 study items\n",
      "  METHOD fdr_bh:\n",
      "   1,693 GO terms found significant (< 0.05=alpha) (1693 enriched +   0 purified): statsmodels fdr_bh\n",
      "     362 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n"
     ]
    }
   ],
   "source": [
    "with open(\"/home/t44p/PW_rawdata/results/full_lasso/gcms/Cellobiose_361_204_rt14_40_nXcv.json\", 'r') as file:\n",
    "    data = json.load(file)\n",
    "\n",
    "data_matched, data_goids, data_term_count, data_goea_all = matchgff2(data['selected_features'], namespace=['molecular_function'], depth_threshold=2, goea=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "405"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data['selected_features'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cellobiose_361_204_rt14_40\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  1%    405 of 30,522 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 405 IDs.\n",
      "100%    405 of    405 study items found in association\n",
      "100%    405 of    405 study items found in population(30522)\n",
      "Calculating 1,693 uncorrected p-values using fisher_scipy_stats\n",
      "   1,693 terms are associated with    362 of 30,522 population items\n",
      "   1,693 terms are associated with    362 of    405 study items\n",
      "  METHOD fdr_bh:\n",
      "   1,693 GO terms found significant (< 0.05=alpha) (1693 enriched +   0 purified): statsmodels fdr_bh\n",
      "     362 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "tyrosine_218_280_rt10_78\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  2%    608 of 30,522 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 608 IDs.\n",
      "100%    608 of    608 study items found in association\n",
      "100%    608 of    608 study items found in population(30522)\n",
      "Calculating 2,356 uncorrected p-values using fisher_scipy_stats\n",
      "   2,356 terms are associated with    537 of 30,522 population items\n",
      "   2,356 terms are associated with    537 of    608 study items\n",
      "  METHOD fdr_bh:\n",
      "   2,356 GO terms found significant (< 0.05=alpha) (2356 enriched +   0 purified): statsmodels fdr_bh\n",
      "     537 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "threonic_acid_292_220_rt7_49\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  3%    814 of 30,522 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 814 IDs.\n",
      "100%    814 of    814 study items found in association\n",
      "100%    814 of    814 study items found in population(30522)\n",
      "Calculating 2,991 uncorrected p-values using fisher_scipy_stats\n",
      "   2,991 terms are associated with    723 of 30,522 population items\n",
      "   2,991 terms are associated with    723 of    814 study items\n",
      "  METHOD fdr_bh:\n",
      "   2,991 GO terms found significant (< 0.05=alpha) (2991 enriched +   0 purified): statsmodels fdr_bh\n",
      "     723 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "_671_2775803_7_254060113\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  3%  1,059 of 30,522 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 1059 IDs.\n",
      "100%  1,059 of  1,059 study items found in association\n",
      "100%  1,059 of  1,059 study items found in population(30522)\n",
      "Calculating 3,667 uncorrected p-values using fisher_scipy_stats\n",
      "   3,667 terms are associated with    944 of 30,522 population items\n",
      "   3,667 terms are associated with    944 of  1,059 study items\n",
      "  METHOD fdr_bh:\n",
      "   3,667 GO terms found significant (< 0.05=alpha) (3667 enriched +   0 purified): statsmodels fdr_bh\n",
      "     944 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Coumaroyl_quinic_acid_\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  2%    563 of 30,522 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 563 IDs.\n",
      "100%    563 of    563 study items found in association\n",
      "100%    563 of    563 study items found in population(30522)\n",
      "Calculating 2,339 uncorrected p-values using fisher_scipy_stats\n",
      "   2,339 terms are associated with    509 of 30,522 population items\n",
      "   2,339 terms are associated with    509 of    563 study items\n",
      "  METHOD fdr_bh:\n",
      "   2,339 GO terms found significant (< 0.05=alpha) (2339 enriched +   0 purified): statsmodels fdr_bh\n",
      "     509 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Quercetin_3_glucoside_3\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  3%    996 of 30,522 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 996 IDs.\n",
      "100%    996 of    996 study items found in association\n",
      "100%    996 of    996 study items found in population(30522)\n",
      "Calculating 3,630 uncorrected p-values using fisher_scipy_stats\n",
      "   3,630 terms are associated with    890 of 30,522 population items\n",
      "   3,630 terms are associated with    890 of    996 study items\n",
      "  METHOD fdr_bh:\n",
      "   3,630 GO terms found significant (< 0.05=alpha) (3630 enriched +   0 purified): statsmodels fdr_bh\n",
      "     890 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n"
     ]
    }
   ],
   "source": [
    "path_gc = \"/home/t44p/PW_rawdata/results/full_lasso/gcms/\"\n",
    "path_lc =\"/home/t44p/PW_rawdata/results/full_lasso/lcms/\"\n",
    "\n",
    "\n",
    "lasso_goea = {}\n",
    "\n",
    "\n",
    "# Iterate over each file in the directory\n",
    "for p in [path_gc, path_lc]:\n",
    "    tmp = 0\n",
    "\n",
    "    for file in os.listdir(p):\n",
    "        if file.endswith(\".json\") and not(file.startswith('gcms_dict_nXcv') or file.startswith('lcms_dict_nXcv')):\n",
    "            file_path = os.path.join(p, file)\n",
    "            with open(file_path, 'r') as json_file:\n",
    "                data = json.load(json_file)\n",
    "            \n",
    "            # Extract mean scores and fold scores\n",
    "            file_name = os.path.splitext(file)[0]\n",
    "            file_name_stripped = file_name.rsplit('_nXcv', 1)[0]\n",
    "            print(file_name_stripped)\n",
    "\n",
    "\n",
    "            # perform GOEA\n",
    "            data_matched, data_goids, data_term_count, data_goea_all = matchgff2(data['selected_features'], namespace=['molecular_function'], depth_threshold=2, goea=True)\n",
    "            lasso_goea[file_name_stripped] = data_goea_all\n",
    "            \n",
    "\n",
    "            tmp += 1\n",
    "            if tmp == 3:\n",
    "              break\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enrichment matrix saved to /home/t44p/PW_rawdata/results/full_lasso/goea/enrichment_matrix_biop.csv\n",
      "Enrichment matrix saved to /home/t44p/PW_rawdata/results/full_lasso/goea/enrichment_matrix_molf.csv\n",
      "Enrichment matrix saved to /home/t44p/PW_rawdata/results/full_lasso/goea/enrichment_matrix_cellc.csv\n"
     ]
    }
   ],
   "source": [
    "goea_results_to_file(lasso_goea, path='/home/t44p/PW_rawdata/results/full_lasso/goea/', obo_path=\"/home/t44p/PW_rawdata/go_obo/go.obo\", to_excel=True)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py3-6",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
