{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use kernel py3-6\n",
    "#import goatools\n",
    "from goatools import obo_parser\n",
    "import re\n",
    "import json\n",
    "import numpy as np\n",
    "from goatools.go_enrichment import GOEnrichmentStudy\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Function to match non-zero Features in GFF and Map GO-Term to GO-ID using GOATools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def matchgff2(feature, gff_file='/home/t44p/PW_rawdata/Transciptome_GenomeAnnotation/Xele_annotated2_gff_export2.gff', obo_path=\"/home/t44p/PW_rawdata/go_obo/go.obo\", namespace=None, depth_threshold=0, goea=False):\n",
    "    \"\"\"\n",
    "    Searches a GFF (General Feature Format) file for specific features and retrieves associated Gene Ontology (GO) terms.\n",
    "\n",
    "    Parameters:\n",
    "    ----------\n",
    "    feature : list or iterable\n",
    "        An iterable of strings representing the features to search for in the GFF file.\n",
    "\n",
    "    gff_file : str, optional\n",
    "        The file path to the GFF file. Defaults to '/home/t44p/PW_rawdata/Transciptome_GenomeAnnotation/Xele_annotated2_gff_export2.gff'.\n",
    "\n",
    "    obo_path : str, optional\n",
    "        The file path to the Gene Ontology .obo file. Defaults to \"/home/t44p/PW_rawdata/go_obo/combined.obo\".\n",
    "\n",
    "    Returns:\n",
    "    -------\n",
    "    tuple of dict\n",
    "        A tuple containing two dictionaries:\n",
    "        1. A dictionary where keys are the features and values are lists of lines (as strings) from the GFF file where these features are found.\n",
    "        2. A dictionary where keys are the features and values are dictionaries. Each inner dictionary has GO IDs as keys and a dictionary with 'name' and 'namespace' of the GO term as values.\n",
    "\n",
    "    Example:\n",
    "    --------\n",
    "    #>>> features = ['gene1', 'gene2']\n",
    "    #>>> matchgff2(features)\n",
    "    ({'gene1': ['line content from GFF file'], 'gene2': ['line content from GFF file']},\n",
    "     {'gene1': {'GO:0000001': {'name': 'term name', 'namespace': 'biological_process'}}, ... })\n",
    "\n",
    "    Notes:\n",
    "    -----\n",
    "    The function uses regular expressions for precise matching of features. It expects the feature followed by a tab character in the GFF file.\n",
    "    Additionally, it extracts GO IDs from matched lines and retrieves their corresponding names and namespaces from the provided Gene Ontology .obo file.\n",
    "    \"\"\"\n",
    "\n",
    "    valid_namespaces = {'biological_process', 'molecular_function', 'cellular_component', None}\n",
    "    # Check if namespace is a list containing only valid elements\n",
    "    if isinstance(namespace, list) and not all(ns in valid_namespaces for ns in namespace):\n",
    "        raise ValueError(\"Invalid namespace provided. Valid options are 'biological_process', \"\n",
    "                         \"'molecular_function', 'cellular_component', or a list containing any of these. \"\n",
    "                         \"You can also use None for no filtering.\")\n",
    "\n",
    "    with open(gff_file, 'r') as file:\n",
    "        go_ontology = obo_parser.GODag(obo_path)\n",
    "        \n",
    "        lines_where_feat_found = {}\n",
    "        go_ids = {}\n",
    "        background_genes = []\n",
    "        go_term_count = {}\n",
    "\n",
    "         # Find the depth of each GO term\n",
    "        go_depths = {go_id: go_term.depth for go_id, go_term in go_ontology.items()}\n",
    "\n",
    "\n",
    "        for feat in feature:\n",
    "            file.seek(0)  # reset file pointer to the beginning for each feature\n",
    "            lines_where_feat_found[feat] = []\n",
    "            go_ids[feat] = {}\n",
    "            pattern = re.compile(re.escape(feat) + r'\\t')  # exact match followed by a tab\n",
    "            for line in file:\n",
    "                if goea:\n",
    "                    if not line.lstrip().startswith('#'):\n",
    "                        background_genes.append(line.split('\\t')[0])\n",
    "\n",
    "                if pattern.search(line):\n",
    "                    lines_where_feat_found[feat].append(line.strip())  # Store the line (as a string) if feature is found\n",
    "                    # Extract GO id\n",
    "                    match = re.search(r\"Ontology_id=([GO:\\d,]+)\", line.strip())\n",
    "                    if match:\n",
    "                        ids = match.group(1).split(',')\n",
    "                        # Map Terms to Ids and Count Occurrences\n",
    "                        for id in ids:\n",
    "                            term = go_ontology.get(id)\n",
    "                            if term is not None:\n",
    "                                go_ids[feat][id] = {'name': term.name, 'namespace': term.namespace}\n",
    "\n",
    "                                if namespace is None or term.namespace in namespace and go_ontology[id].depth >= depth_threshold:\n",
    "                                    # Count Occurrences\n",
    "                                    if id in go_term_count:\n",
    "                                        go_term_count[id] = (term.name, go_term_count[id][1] + 1, term.namespace)\n",
    "                                    else:\n",
    "                                        go_term_count[id] = (term.name, 1, term.namespace)\n",
    "                            else:\n",
    "                                go_ids[feat][id] = {'name': None, 'namespace': None}\n",
    "                                if id not in go_term_count:\n",
    "                                    go_term_count[id] = (None, 1)\n",
    "        if goea:\n",
    "            print(\"GO Enrichment Analysis >>\")\n",
    "            goea_obj = GOEnrichmentStudy(\n",
    "                background_genes,\n",
    "                go_ids,  # This needs to be a dict mapping gene IDs to a set of GO IDs\n",
    "                go_ontology,\n",
    "                propagate_counts=False,\n",
    "                alpha=0.05,  # significance level for the statistical test\n",
    "                methods=['fdr_bh']  # correction method for multiple testing\n",
    "            )\n",
    "            goea_result = goea_obj.run_study(go_ids.keys())\n",
    "            return lines_where_feat_found, go_ids, go_term_count, goea_result\n",
    "\n",
    "\n",
    "        return lines_where_feat_found, go_ids, go_term_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def plot_go_terms(term_count_dict, top_n=3, figsize=(25, 15), title=\"GO Term Occurrences\"):\n",
    "    print(\"THIS FUNCTION IS DEPRECATED; DO NOT USE IT\")\n",
    "    \"\"\"\n",
    "    Plots a bar chart of GO term occurrences.\n",
    "\n",
    "    Parameters:\n",
    "    term_count_dict (dict): Dictionary of GO terms with their counts.\n",
    "    top_n (int): Number of top features to highlight.\n",
    "    figsize (tuple): Figure size for the plot.\n",
    "    title (str): Title of the plot.\n",
    "    \"\"\"\n",
    "    # Extract terms and counts\n",
    "    go_terms = list(term_count_dict.keys())\n",
    "    counts = [term_count_dict[term][1] for term in go_terms]\n",
    "\n",
    "    # Sort the terms by count and identify top features\n",
    "    sorted_terms = sorted(term_count_dict.items(), key=lambda x: x[1][1], reverse=True)\n",
    "    top_terms = sorted_terms[:top_n]\n",
    "\n",
    "    # Create lists for the plot\n",
    "    terms_to_plot = [term for term, _ in sorted_terms]\n",
    "    counts_to_plot = [count for _, (_, count) in sorted_terms]\n",
    "\n",
    "    # Colors - highlight top features in red, others in blue\n",
    "    colors = ['red' if term in [t[0] for t in top_terms] else 'blue' for term in terms_to_plot]\n",
    "\n",
    "    # Create the bar plot\n",
    "    plt.figure(figsize=figsize)\n",
    "    plt.bar(terms_to_plot, counts_to_plot, color=colors)\n",
    "    plt.xlabel('GO Terms')\n",
    "    plt.ylabel('Counts')\n",
    "    plt.xticks(rotation=90)\n",
    "    plt.title(title)\n",
    "    plt.tight_layout()\n",
    "\n",
    "    plt.show()\n",
    "    print(f\"top {top_n} most abundant GO Terms\")\n",
    "    for t in top_terms:\n",
    "        print(t)\n",
    "\n",
    "# Example usage:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"def tabulate(term_count_dict, sort=True):\n",
    "    if sort:\n",
    "        print(f\"count\\tGO ID\\tGO Term\\tnamespace\")\n",
    "        term_count_dict = sorted(term_count_dict.items(), key=lambda x: x[1][1], reverse=True)\n",
    "        for goid, values in term_count_dict:\n",
    "            count, term, namespace = values\n",
    "            print(f\"{count}\\t{goid}\\t{term}\\t{namespace}\")\n",
    "    else:\n",
    "        print(f\"count\\tGO ID\\tGO Term\\tnamespace\")\n",
    "        for goid in term_count_dict.keys():\n",
    "            if term_count_dict[goid][0] is not None:\n",
    "                print(f\"{term_count_dict[goid][1]}\\t{goid}\\t{term_count_dict[goid][0]}\\t{term_count_dict[goid][2]}\")\n",
    "            else:\n",
    "                print(f\"{term_count_dict[goid][1]}\\t{goid}\")\"\"\"\n",
    "def tabulate(term_count_dict, sort=True):\n",
    "    print(f\"count\\tGO ID\\tGO Term\\tnamespace\")\n",
    "\n",
    "    # Conditionally sort the dictionary if required\n",
    "    items = sorted(term_count_dict.items(), key=lambda x: x[1][1], reverse=True) if sort else term_count_dict.items()\n",
    "\n",
    "    for goid, values in items:\n",
    "        count = values[1]\n",
    "        term = values[0] if values[0] is not None else \"N/A\"\n",
    "        namespace = values[2] if len(values) > 2 else \"N/A\"\n",
    "\n",
    "        print(f\"{count}\\t{goid}\\t{term}\\t{namespace}\")\n",
    "    #return \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Glucose "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    676 of 793,572 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 26 IDs.\n",
      "100%     26 of     26 study items found in association\n",
      "100%     26 of     26 study items found in population(793572)\n",
      "Calculating 236 uncorrected p-values using fisher_scipy_stats\n",
      "     236 terms are associated with     20 of 793,572 population items\n",
      "     236 terms are associated with     20 of     26 study items\n",
      "  METHOD fdr_bh:\n",
      "     236 GO terms found significant (< 0.05=alpha) (236 enriched +   0 purified): statsmodels fdr_bh\n",
      "      20 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    676 of 793,572 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 26 IDs.\n",
      "100%     26 of     26 study items found in association\n",
      "100%     26 of     26 study items found in population(793572)\n",
      "Calculating 236 uncorrected p-values using fisher_scipy_stats\n",
      "     236 terms are associated with     20 of 793,572 population items\n",
      "     236 terms are associated with     20 of     26 study items\n",
      "  METHOD fdr_bh:\n",
      "     236 GO terms found significant (< 0.05=alpha) (236 enriched +   0 purified): statsmodels fdr_bh\n",
      "      20 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    676 of 793,572 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 26 IDs.\n",
      "100%     26 of     26 study items found in association\n",
      "100%     26 of     26 study items found in population(793572)\n",
      "Calculating 236 uncorrected p-values using fisher_scipy_stats\n",
      "     236 terms are associated with     20 of 793,572 population items\n",
      "     236 terms are associated with     20 of     26 study items\n",
      "  METHOD fdr_bh:\n",
      "     236 GO terms found significant (< 0.05=alpha) (236 enriched +   0 purified): statsmodels fdr_bh\n",
      "      20 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n"
     ]
    }
   ],
   "source": [
    "with open(\"./lasso_models/10xKfold_lasso_output/glucose_nXcv.json\", 'r') as file:\n",
    "    glucose_nXcv = json.load(file)\n",
    "#glucose_nXcv_matched, glucose_goids, glucose_term_count, glucose_goea = matchgff2(glucose_nXcv['common_features'], goea=True)\n",
    "glucose_nXcv_matched, gluc_goids_biop, gluc_term_count_biop, gluc_goea_biop = matchgff2(glucose_nXcv['common_features'], namespace=['biological_process'], depth_threshold=2, goea=True)\n",
    "glucose_nXcv_matched, gluc_goids_biop, gluc_term_count_molf, gluc_goea_molf = matchgff2(glucose_nXcv['common_features'], namespace=['molecular_function'], depth_threshold=2, goea=True)\n",
    "glucose_nXcv_matched, gluc_goids_biop, gluc_term_count_cellc, gluc_goea_cellc = matchgff2(glucose_nXcv['common_features'], namespace=['cellular_component'], depth_threshold=2, goea=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# <<< GLUCOSE >>>\n",
      "# 10x3FoldCV mean score: 0.37766355282118325 \n",
      "# count common features: 26\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "2\tGO:0005975\tcarbohydrate metabolic process\tbiological_process\n",
      "2\tGO:0051240\tpositive regulation of multicellular organismal process\tbiological_process\n",
      "2\tGO:0048766\troot hair initiation\tbiological_process\n",
      "2\tGO:0055062\tphosphate ion homeostasis\tbiological_process\n",
      "1\tGO:0006048\tUDP-N-acetylglucosamine biosynthetic process\tbiological_process\n",
      "1\tGO:0006281\tDNA repair\tbiological_process\n",
      "1\tGO:0006487\tprotein N-linked glycosylation\tbiological_process\n",
      "1\tGO:0006493\tprotein O-linked glycosylation\tbiological_process\n",
      "1\tGO:0007283\tspermatogenesis\tbiological_process\n",
      "1\tGO:0019255\tglucose 1-phosphate metabolic process\tbiological_process\n",
      "1\tGO:0030097\themopoiesis\tbiological_process\n",
      "1\tGO:0034221\tfungal-type cell wall chitin biosynthetic process\tbiological_process\n",
      "1\tGO:0006869\tlipid transport\tbiological_process\n",
      "1\tGO:0006897\tendocytosis\tbiological_process\n",
      "1\tGO:0009306\tprotein secretion\tbiological_process\n",
      "1\tGO:0016043\tcellular component organization\tbiological_process\n",
      "1\tGO:1903530\tregulation of secretion by cell\tbiological_process\n",
      "1\tGO:0005992\ttrehalose biosynthetic process\tbiological_process\n",
      "1\tGO:0005993\ttrehalose catabolic process\tbiological_process\n",
      "1\tGO:0009793\tembryo development ending in seed dormancy\tbiological_process\n",
      "1\tGO:0009832\tplant-type cell wall biogenesis\tbiological_process\n",
      "1\tGO:0010182\tsugar mediated signaling pathway\tbiological_process\n",
      "1\tGO:0030437\tascospore formation\tbiological_process\n",
      "1\tGO:0034605\tcellular response to heat\tbiological_process\n",
      "1\tGO:0036168\tfilamentous growth of a population of unicellular organisms in response to heat\tbiological_process\n",
      "1\tGO:0036180\tfilamentous growth of a population of unicellular organisms in response to biotic stimulus\tbiological_process\n",
      "1\tGO:0048364\troot development\tbiological_process\n",
      "1\tGO:0048574\tlong-day photoperiodism, flowering\tbiological_process\n",
      "1\tGO:0051301\tcell division\tbiological_process\n",
      "1\tGO:0070301\tcellular response to hydrogen peroxide\tbiological_process\n",
      "1\tGO:0070413\ttrehalose metabolism in response to stress\tbiological_process\n",
      "1\tGO:0080186\tdevelopmental vegetative growth\tbiological_process\n",
      "1\tGO:2000070\tregulation of response to water deprivation\tbiological_process\n",
      "1\tGO:1901990\tregulation of mitotic cell cycle phase transition\tbiological_process\n",
      "1\tGO:0008286\tinsulin receptor signaling pathway\tbiological_process\n",
      "1\tGO:0045597\tpositive regulation of cell differentiation\tbiological_process\n",
      "1\tGO:0045944\tpositive regulation of transcription by RNA polymerase II\tbiological_process\n",
      "1\tGO:0006337\tnucleosome disassembly\tbiological_process\n",
      "1\tGO:0042127\tregulation of cell population proliferation\tbiological_process\n",
      "1\tGO:0009887\tanimal organ morphogenesis\tbiological_process\n",
      "1\tGO:0000122\tnegative regulation of transcription by RNA polymerase II\tbiological_process\n",
      "1\tGO:2000781\tpositive regulation of double-strand break repair\tbiological_process\n",
      "1\tGO:0032435\tnegative regulation of proteasomal ubiquitin-dependent protein catabolic process\tbiological_process\n",
      "1\tGO:0046658\tN/A\tN/A\n",
      "1\tGO:0009617\tresponse to bacterium\tbiological_process\n",
      "1\tGO:0009627\tsystemic acquired resistance\tbiological_process\n",
      "1\tGO:0009737\tresponse to abscisic acid\tbiological_process\n",
      "1\tGO:0009845\tseed germination\tbiological_process\n",
      "1\tGO:0009932\tcell tip growth\tbiological_process\n",
      "1\tGO:0010067\tprocambium histogenesis\tbiological_process\n",
      "1\tGO:0010305\tleaf vascular tissue pattern formation\tbiological_process\n",
      "1\tGO:0010588\tcotyledon vascular tissue pattern formation\tbiological_process\n",
      "1\tGO:0016050\tvesicle organization\tbiological_process\n",
      "1\tGO:0032957\tinositol trisphosphate metabolic process\tbiological_process\n",
      "1\tGO:0043933\tprotein-containing complex organization\tbiological_process\n",
      "1\tGO:0046855\tN/A\tN/A\n",
      "1\tGO:0046856\tphosphatidylinositol dephosphorylation\tbiological_process\n",
      "1\tGO:0048016\tN/A\tN/A\n",
      "1\tGO:0048488\tsynaptic vesicle endocytosis\tbiological_process\n",
      "1\tGO:0048522\tpositive regulation of cellular process\tbiological_process\n",
      "1\tGO:0065008\tregulation of biological quality\tbiological_process\n",
      "1\tGO:0071472\tcellular response to salt stress\tbiological_process\n",
      "1\tGO:0071702\torganic substance transport\tbiological_process\n",
      "1\tGO:0071705\tnitrogen compound transport\tbiological_process\n",
      "1\tGO:2000369\tregulation of clathrin-dependent endocytosis\tbiological_process\n",
      "1\tGO:2000377\tregulation of reactive oxygen species metabolic process\tbiological_process\n",
      "1\tGO:0009057\tmacromolecule catabolic process\tbiological_process\n",
      "1\tGO:0007623\tcircadian rhythm\tbiological_process\n",
      "1\tGO:0009638\tphototropism\tbiological_process\n",
      "1\tGO:0009734\tauxin-activated signaling pathway\tbiological_process\n",
      "1\tGO:0009860\tpollen tube growth\tbiological_process\n",
      "1\tGO:0009908\tflower development\tbiological_process\n",
      "1\tGO:0009958\tpositive gravitropism\tbiological_process\n",
      "1\tGO:0010362\tnegative regulation of anion channel activity by blue light\tbiological_process\n",
      "1\tGO:0010540\tbasipetal auxin transport\tbiological_process\n",
      "1\tGO:0040008\tregulation of growth\tbiological_process\n",
      "1\tGO:0046777\tprotein autophosphorylation\tbiological_process\n",
      "1\tGO:0048767\troot hair elongation\tbiological_process\n",
      "1\tGO:0048825\tcotyledon development\tbiological_process\n",
      "1\tGO:0062075\tpollen aperture formation\tbiological_process\n",
      "1\tGO:2000114\tregulation of establishment of cell polarity\tbiological_process\n",
      "1\tGO:0010584\tpollen exine formation\tbiological_process\n",
      "1\tGO:0010930\tnegative regulation of auxin mediated signaling pathway\tbiological_process\n",
      "1\tGO:0015757\tgalactose transmembrane transport\tbiological_process\n",
      "1\tGO:0051260\tprotein homooligomerization\tbiological_process\n",
      "1\tGO:0052543\tcallose deposition in cell wall\tbiological_process\n",
      "1\tGO:0060918\tauxin transport\tbiological_process\n",
      "1\tGO:0008643\tcarbohydrate transport\tbiological_process\n",
      "1\tGO:0015714\tphosphoenolpyruvate transport\tbiological_process\n",
      "1\tGO:0015783\tGDP-fucose transmembrane transport\tbiological_process\n",
      "1\tGO:0015786\tUDP-glucose transmembrane transport\tbiological_process\n",
      "1\tGO:0048359\tmucilage metabolic process involved in seed coat development\tbiological_process\n",
      "1\tGO:0072334\tUDP-galactose transmembrane transport\tbiological_process\n",
      "1\tGO:0080040\tpositive regulation of cellular response to phosphate starvation\tbiological_process\n",
      "1\tGO:0070417\tcellular response to cold\tbiological_process\n",
      "1\tGO:0006094\tgluconeogenesis\tbiological_process\n",
      "1\tGO:0006096\tglycolytic process\tbiological_process\n",
      "1\tGO:0009408\tresponse to heat\tbiological_process\n",
      "1\tGO:0009416\tresponse to light stimulus\tbiological_process\n",
      "1\tGO:0010154\tfruit development\tbiological_process\n",
      "1\tGO:0034059\tresponse to anoxia\tbiological_process\n",
      "1\tGO:0045893\tpositive regulation of DNA-templated transcription\tbiological_process\n",
      "1\tGO:0046686\tresponse to cadmium ion\tbiological_process\n",
      "1\tGO:0016036\tcellular response to phosphate starvation\tbiological_process\n",
      "1\tGO:0016311\tdephosphorylation\tbiological_process\n",
      "1\tGO:0050829\tdefense response to Gram-negative bacterium\tbiological_process\n",
      "1\tGO:0050832\tdefense response to fungus\tbiological_process\n",
      "1\tGO:0051174\tregulation of phosphorus metabolic process\tbiological_process\n",
      "1\tGO:0009620\tresponse to fungus\tbiological_process\n",
      "1\tGO:0002088\tlens development in camera-type eye\tbiological_process\n",
      "1\tGO:0006749\tglutathione metabolic process\tbiological_process\n",
      "1\tGO:0006909\tphagocytosis\tbiological_process\n",
      "1\tGO:0007040\tlysosome organization\tbiological_process\n",
      "1\tGO:0007420\tbrain development\tbiological_process\n",
      "1\tGO:0007616\tlong-term memory\tbiological_process\n",
      "1\tGO:0007625\tgrooming behavior\tbiological_process\n",
      "1\tGO:0007628\tadult walking behavior\tbiological_process\n",
      "1\tGO:0008542\tvisual learning\tbiological_process\n",
      "1\tGO:0015811\tL-cystine transport\tbiological_process\n",
      "1\tGO:0031326\tregulation of cellular biosynthetic process\tbiological_process\n",
      "1\tGO:0046034\tATP metabolic process\tbiological_process\n",
      "1\tGO:0055085\ttransmembrane transport\tbiological_process\n",
      "1\tGO:1903427\tnegative regulation of reactive oxygen species biosynthetic process\tbiological_process\n",
      "1\tGO:1904263\tpositive regulation of TORC1 signaling\tbiological_process\n",
      "1\tGO:0031305\tN/A\tN/A\n",
      "1\tGO:0006979\tresponse to oxidative stress\tbiological_process\n",
      "1\tGO:0007615\tanesthesia-resistant memory\tbiological_process\n",
      "1\tGO:0008016\tregulation of heart contraction\tbiological_process\n",
      "1\tGO:0019722\tcalcium-mediated signaling\tbiological_process\n",
      "1\tGO:0032024\tpositive regulation of insulin secretion\tbiological_process\n",
      "1\tGO:0036444\tcalcium import into the mitochondrion\tbiological_process\n",
      "1\tGO:0042593\tglucose homeostasis\tbiological_process\n",
      "1\tGO:0043009\tchordate embryonic development\tbiological_process\n",
      "1\tGO:0051259\tprotein complex oligomerization\tbiological_process\n",
      "1\tGO:0051561\tpositive regulation of mitochondrial calcium ion concentration\tbiological_process\n",
      "1\tGO:0060027\tconvergent extension involved in gastrulation\tbiological_process\n",
      "1\tGO:0070509\tcalcium ion import\tbiological_process\n",
      "1\tGO:0090023\tpositive regulation of neutrophil chemotaxis\tbiological_process\n",
      "1\tGO:0090141\tpositive regulation of mitochondrial fission\tbiological_process\n",
      "1\tGO:0090527\tN/A\tN/A\n",
      "1\tGO:0001666\tresponse to hypoxia\tbiological_process\n",
      "1\tGO:0002239\tresponse to oomycetes\tbiological_process\n",
      "1\tGO:0009741\tresponse to brassinosteroid\tbiological_process\n",
      "# <<< GLUCOSE >>>\n",
      "# 10x3FoldCV mean score: 0.37766355282118325 \n",
      "# count common features: 26\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "7\tGO:0005515\tprotein binding\tmolecular_function\n",
      "1\tGO:0000287\tmagnesium ion binding\tmolecular_function\n",
      "1\tGO:0004610\tphosphoacetylglucosamine mutase activity\tmolecular_function\n",
      "1\tGO:0004614\tphosphoglucomutase activity\tmolecular_function\n",
      "1\tGO:0005543\tphospholipid binding\tmolecular_function\n",
      "1\tGO:0046872\tmetal ion binding\tmolecular_function\n",
      "1\tGO:0003825\talpha,alpha-trehalose-phosphate synthase (UDP-forming) activity\tmolecular_function\n",
      "1\tGO:0004805\ttrehalose-phosphatase activity\tmolecular_function\n",
      "1\tGO:0102986\ttrehalose synthase activity\tmolecular_function\n",
      "1\tGO:0003682\tchromatin binding\tmolecular_function\n",
      "1\tGO:0046658\tN/A\tN/A\n",
      "1\tGO:0008810\tcellulase activity\tmolecular_function\n",
      "1\tGO:0042973\tglucan endo-1,3-beta-D-glucosidase activity\tmolecular_function\n",
      "1\tGO:0004439\tphosphatidylinositol-4,5-bisphosphate 5-phosphatase activity\tmolecular_function\n",
      "1\tGO:0004445\tinositol-polyphosphate 5-phosphatase activity\tmolecular_function\n",
      "1\tGO:0034485\tphosphatidylinositol-3,4,5-trisphosphate 5-phosphatase activity\tmolecular_function\n",
      "1\tGO:0046855\tN/A\tN/A\n",
      "1\tGO:0048016\tN/A\tN/A\n",
      "1\tGO:0004806\ttriglyceride lipase activity\tmolecular_function\n",
      "1\tGO:0033878\thormone-sensitive lipase activity\tmolecular_function\n",
      "1\tGO:0047372\tacylglycerol lipase activity\tmolecular_function\n",
      "1\tGO:0004674\tprotein serine/threonine kinase activity\tmolecular_function\n",
      "1\tGO:0004712\tprotein serine/threonine/tyrosine kinase activity\tmolecular_function\n",
      "1\tGO:0005524\tATP binding\tmolecular_function\n",
      "1\tGO:0005546\tphosphatidylinositol-4,5-bisphosphate binding\tmolecular_function\n",
      "1\tGO:0005547\tphosphatidylinositol-3,4,5-trisphosphate binding\tmolecular_function\n",
      "1\tGO:0009882\tblue light photoreceptor activity\tmolecular_function\n",
      "1\tGO:0010314\tphosphatidylinositol-5-phosphate binding\tmolecular_function\n",
      "1\tGO:0019901\tprotein kinase binding\tmolecular_function\n",
      "1\tGO:0032266\tphosphatidylinositol-3-phosphate binding\tmolecular_function\n",
      "1\tGO:0070273\tphosphatidylinositol-4-phosphate binding\tmolecular_function\n",
      "1\tGO:0070300\tphosphatidic acid binding\tmolecular_function\n",
      "1\tGO:0080025\tphosphatidylinositol-3,5-bisphosphate binding\tmolecular_function\n",
      "1\tGO:0106310\tprotein serine kinase activity\tmolecular_function\n",
      "1\tGO:0008080\tN-acetyltransferase activity\tmolecular_function\n",
      "1\tGO:0051119\tsugar transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0005457\tGDP-fucose transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0005459\tUDP-galactose transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0005460\tUDP-glucose transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0015315\torganophosphate:inorganic phosphate antiporter activity\tmolecular_function\n",
      "1\tGO:0071917\ttriose-phosphate transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0003677\tDNA binding\tmolecular_function\n",
      "1\tGO:0004365\tglyceraldehyde-3-phosphate dehydrogenase (NAD+) (phosphorylating) activity\tmolecular_function\n",
      "1\tGO:0042301\tphosphate ion binding\tmolecular_function\n",
      "1\tGO:0050661\tNADP binding\tmolecular_function\n",
      "1\tGO:0070403\tNAD+ binding\tmolecular_function\n",
      "1\tGO:0003729\tmRNA binding\tmolecular_function\n",
      "1\tGO:0003993\tacid phosphatase activity\tmolecular_function\n",
      "1\tGO:0004721\tphosphoprotein phosphatase activity\tmolecular_function\n",
      "1\tGO:0008199\tferric iron binding\tmolecular_function\n",
      "1\tGO:0008270\tzinc ion binding\tmolecular_function\n",
      "1\tGO:0030145\tmanganese ion binding\tmolecular_function\n",
      "1\tGO:0003723\tRNA binding\tmolecular_function\n",
      "1\tGO:0015184\tL-cystine transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0015295\tsolute:proton symporter activity\tmolecular_function\n",
      "1\tGO:0034639\tL-amino acid efflux transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0031305\tN/A\tN/A\n",
      "1\tGO:0005262\tcalcium channel activity\tmolecular_function\n",
      "1\tGO:0015292\tuniporter activity\tmolecular_function\n",
      "1\tGO:0019855\tcalcium channel inhibitor activity\tmolecular_function\n",
      "1\tGO:0042802\tidentical protein binding\tmolecular_function\n",
      "1\tGO:0090527\tN/A\tN/A\n",
      "# <<< GLUCOSE >>>\n",
      "# 10x3FoldCV mean score: 0.37766355282118325 \n",
      "# count common features: 26\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "5\tGO:0005634\tnucleus\tcellular_component\n",
      "3\tGO:0005737\tcytoplasm\tcellular_component\n",
      "3\tGO:0005886\tplasma membrane\tcellular_component\n",
      "3\tGO:0016021\tmembrane\tcellular_component\n",
      "3\tGO:0005576\textracellular region\tcellular_component\n",
      "3\tGO:0005794\tGolgi apparatus\tcellular_component\n",
      "2\tGO:0045202\tsynapse\tcellular_component\n",
      "2\tGO:0005773\tvacuole\tcellular_component\n",
      "2\tGO:0005654\tnucleoplasm\tcellular_component\n",
      "2\tGO:0005829\tcytosol\tcellular_component\n",
      "1\tGO:0000139\tGolgi membrane\tcellular_component\n",
      "1\tGO:0010008\tendosome membrane\tcellular_component\n",
      "1\tGO:0098827\tendoplasmic reticulum subcompartment\tcellular_component\n",
      "1\tGO:0005946\talpha,alpha-trehalose-phosphate synthase complex (UDP-forming)\tcellular_component\n",
      "1\tGO:0016514\tSWI/SNF complex\tcellular_component\n",
      "1\tGO:0001741\tXY body\tcellular_component\n",
      "1\tGO:0071564\tnpBAF complex\tcellular_component\n",
      "1\tGO:0071565\tnBAF complex\tcellular_component\n",
      "1\tGO:0043231\tintracellular membrane-bounded organelle\tcellular_component\n",
      "1\tGO:0046658\tN/A\tN/A\n",
      "1\tGO:0009507\tchloroplast\tcellular_component\n",
      "1\tGO:0005768\tendosome\tcellular_component\n",
      "1\tGO:0005856\tcytoskeleton\tcellular_component\n",
      "1\tGO:0043005\tneuron projection\tcellular_component\n",
      "1\tGO:0046855\tN/A\tN/A\n",
      "1\tGO:0048016\tN/A\tN/A\n",
      "1\tGO:0005747\tmitochondrial respiratory chain complex I\tcellular_component\n",
      "1\tGO:0005811\tlipid droplet\tcellular_component\n",
      "1\tGO:0090406\tpollen tube\tcellular_component\n",
      "1\tGO:0009925\tbasal plasma membrane\tcellular_component\n",
      "1\tGO:0009986\tcell surface\tcellular_component\n",
      "1\tGO:0005789\tendoplasmic reticulum membrane\tcellular_component\n",
      "1\tGO:0005887\tplasma membrane\tcellular_component\n",
      "1\tGO:0005783\tendoplasmic reticulum\tcellular_component\n",
      "1\tGO:0016020\tmembrane\tcellular_component\n",
      "1\tGO:0005740\tmitochondrial envelope\tcellular_component\n",
      "1\tGO:0005615\textracellular space\tcellular_component\n",
      "1\tGO:0005765\tlysosomal membrane\tcellular_component\n",
      "1\tGO:0005770\tlate endosome\tcellular_component\n",
      "1\tGO:0042470\tmelanosome\tcellular_component\n",
      "1\tGO:0009941\tchloroplast envelope\tcellular_component\n",
      "1\tGO:0031305\tN/A\tN/A\n",
      "1\tGO:1990246\tuniplex complex\tcellular_component\n",
      "1\tGO:0090527\tN/A\tN/A\n",
      "1\tGO:0009505\tplant-type cell wall\tcellular_component\n"
     ]
    }
   ],
   "source": [
    "print(\"# <<< GLUCOSE >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(glucose_nXcv['mean_scores'])} \\n# count common features: {len(glucose_nXcv['common_features'])}\")\n",
    "tabulate(gluc_term_count_biop)\n",
    "print(\"# <<< GLUCOSE >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(glucose_nXcv['mean_scores'])} \\n# count common features: {len(glucose_nXcv['common_features'])}\")\n",
    "tabulate(gluc_term_count_molf)\n",
    "print(\"# <<< GLUCOSE >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(glucose_nXcv['mean_scores'])} \\n# count common features: {len(glucose_nXcv['common_features'])}\")\n",
    "tabulate(gluc_term_count_cellc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot_go_terms(gluc_term_count_biop, top_n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'PRINT EACH GFF ENTRY AND GO-TERMS CORRESPONDING TO NO-ZERO FEATURES'"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"PRINT EACH GFF ENTRY AND GO-TERMS CORRESPONDING TO NO-ZERO FEATURES\"\"\"\n",
    "#for ele in glucose_nXcv_matched:\n",
    "#    print(f\"{glucose_nXcv_matched[ele]}\")\n",
    "#    #print(glucose_goids[ele], \"\\n\")\n",
    "#    for id in glucose_goids[ele]:\n",
    "#        print(id, \": \", glucose_goids[ele][id])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sucrose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%     64 of 244,176 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 8 IDs.\n",
      "100%      8 of      8 study items found in association\n",
      "100%      8 of      8 study items found in population(244176)\n",
      "Calculating 70 uncorrected p-values using fisher_scipy_stats\n",
      "      70 terms are associated with      7 of 244,176 population items\n",
      "      70 terms are associated with      7 of      8 study items\n",
      "  METHOD fdr_bh:\n",
      "      70 GO terms found significant (< 0.05=alpha) ( 70 enriched +   0 purified): statsmodels fdr_bh\n",
      "       7 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%     64 of 244,176 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 8 IDs.\n",
      "100%      8 of      8 study items found in association\n",
      "100%      8 of      8 study items found in population(244176)\n",
      "Calculating 70 uncorrected p-values using fisher_scipy_stats\n",
      "      70 terms are associated with      7 of 244,176 population items\n",
      "      70 terms are associated with      7 of      8 study items\n",
      "  METHOD fdr_bh:\n",
      "      70 GO terms found significant (< 0.05=alpha) ( 70 enriched +   0 purified): statsmodels fdr_bh\n",
      "       7 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%     64 of 244,176 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 8 IDs.\n",
      "100%      8 of      8 study items found in association\n",
      "100%      8 of      8 study items found in population(244176)\n",
      "Calculating 70 uncorrected p-values using fisher_scipy_stats\n",
      "      70 terms are associated with      7 of 244,176 population items\n",
      "      70 terms are associated with      7 of      8 study items\n",
      "  METHOD fdr_bh:\n",
      "      70 GO terms found significant (< 0.05=alpha) ( 70 enriched +   0 purified): statsmodels fdr_bh\n",
      "       7 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n"
     ]
    }
   ],
   "source": [
    "with open(\"./lasso_models/10xKfold_lasso_output/sucrose_nXcv.json\", 'r') as file:\n",
    "    sucrose_nXcv = json.load(file)\n",
    "#sucrose_nXcv_matched, sucrose_goids, sucrose_term_count,sucrose_goea = matchgff2(sucrose_nXcv['common_features'], goea=True)\n",
    "sucrose_nXcv_matched, suc_goids_biop, suc_term_count_biop, suc_goea_biop = matchgff2(sucrose_nXcv['common_features'], namespace=['biological_process'], depth_threshold=2, goea=True)\n",
    "sucrose_nXcv_matched, suc_goids_molf, suc_term_count_molf, suc_goea_molf = matchgff2(sucrose_nXcv['common_features'], namespace=['molecular_function'], depth_threshold=2, goea=True)\n",
    "sucrose_nXcv_matched, suc_goids_cellc, suc_term_count_cellc, suc_goea_cellc = matchgff2(sucrose_nXcv['common_features'], namespace=['cellular_component'], depth_threshold=2, goea=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# <<< SUCROSE >>>\n",
      "# 10x3FoldCV mean score: 0.3188916665994363 \n",
      "# count common features: 8\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "1\tGO:0006807\tnitrogen compound metabolic process\tbiological_process\n",
      "1\tGO:0006974\tDNA damage response\tbiological_process\n",
      "1\tGO:0043170\tmacromolecule metabolic process\tbiological_process\n",
      "1\tGO:0044238\tprimary metabolic process\tbiological_process\n",
      "1\tGO:0050794\tregulation of cellular process\tbiological_process\n",
      "1\tGO:0006622\tprotein targeting to lysosome\tbiological_process\n",
      "1\tGO:0009306\tprotein secretion\tbiological_process\n",
      "1\tGO:0010324\tmembrane invagination\tbiological_process\n",
      "1\tGO:0010642\tnegative regulation of platelet-derived growth factor receptor signaling pathway\tbiological_process\n",
      "1\tGO:0016525\tnegative regulation of angiogenesis\tbiological_process\n",
      "1\tGO:0030948\tnegative regulation of vascular endothelial growth factor receptor signaling pathway\tbiological_process\n",
      "1\tGO:0043162\tubiquitin-dependent protein catabolic process via the multivesicular body sorting pathway\tbiological_process\n",
      "1\tGO:0043405\tregulation of MAP kinase activity\tbiological_process\n",
      "1\tGO:0045053\tprotein retention in Golgi apparatus\tbiological_process\n",
      "1\tGO:0045324\tlate endosome to vacuole transport\tbiological_process\n",
      "1\tGO:0046426\tnegative regulation of receptor signaling pathway via JAK-STAT\tbiological_process\n",
      "1\tGO:0071985\tmultivesicular body sorting pathway\tbiological_process\n",
      "1\tGO:0072657\tprotein localization to membrane\tbiological_process\n",
      "1\tGO:0140504\tmicrolipophagy\tbiological_process\n",
      "1\tGO:1903319\tpositive regulation of protein maturation\tbiological_process\n",
      "1\tGO:1903543\tpositive regulation of exosomal secretion\tbiological_process\n",
      "1\tGO:1904669\tATP export\tbiological_process\n",
      "1\tGO:0006508\tproteolysis\tbiological_process\n",
      "1\tGO:0009567\tdouble fertilization forming a zygote and endosperm\tbiological_process\n",
      "1\tGO:0009609\tresponse to symbiotic bacterium\tbiological_process\n",
      "1\tGO:0009610\tresponse to symbiotic fungus\tbiological_process\n",
      "1\tGO:0009682\tinduced systemic resistance\tbiological_process\n",
      "1\tGO:0009733\tresponse to auxin\tbiological_process\n",
      "1\tGO:0009860\tpollen tube growth\tbiological_process\n",
      "1\tGO:0010102\tlateral root morphogenesis\tbiological_process\n",
      "1\tGO:0005975\tcarbohydrate metabolic process\tbiological_process\n",
      "1\tGO:0006119\toxidative phosphorylation\tbiological_process\n",
      "1\tGO:0022904\trespiratory electron transport chain\tbiological_process\n",
      "1\tGO:0009620\tresponse to fungus\tbiological_process\n",
      "1\tGO:0009693\tethylene biosynthetic process\tbiological_process\n",
      "1\tGO:0009727\tdetection of ethylene stimulus\tbiological_process\n",
      "1\tGO:0009835\tfruit ripening\tbiological_process\n",
      "1\tGO:0010030\tpositive regulation of seed germination\tbiological_process\n",
      "1\tGO:0071398\tcellular response to fatty acid\tbiological_process\n",
      "1\tGO:0071732\tcellular response to nitric oxide\tbiological_process\n",
      "# <<< SUCROSE >>>\n",
      "# 10x3FoldCV mean score: 0.3188916665994363 \n",
      "# count common features: 8\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "1\tGO:0019904\tprotein domain specific binding\tmolecular_function\n",
      "1\tGO:0032266\tphosphatidylinositol-3-phosphate binding\tmolecular_function\n",
      "1\tGO:0042802\tidentical protein binding\tmolecular_function\n",
      "1\tGO:0043130\tubiquitin binding\tmolecular_function\n",
      "1\tGO:0044389\tubiquitin-like protein ligase binding\tmolecular_function\n",
      "1\tGO:0046982\tprotein heterodimerization activity\tmolecular_function\n",
      "1\tGO:0140036\tubiquitin-dependent protein binding\tmolecular_function\n",
      "1\tGO:0004252\tserine-type endopeptidase activity\tmolecular_function\n",
      "1\tGO:0016791\tphosphatase activity\tmolecular_function\n",
      "1\tGO:0009815\t1-aminocyclopropane-1-carboxylate oxidase activity\tmolecular_function\n",
      "1\tGO:0031418\tL-ascorbic acid binding\tmolecular_function\n",
      "1\tGO:0046872\tmetal ion binding\tmolecular_function\n",
      "1\tGO:0051213\tdioxygenase activity\tmolecular_function\n",
      "# <<< SUCROSE >>>\n",
      "# 10x3FoldCV mean score: 0.3188916665994363 \n",
      "# count common features: 8\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "1\tGO:0005634\tnucleus\tcellular_component\n",
      "1\tGO:0005764\tlysosome\tcellular_component\n",
      "1\tGO:0005769\tearly endosome\tcellular_component\n",
      "1\tGO:0005774\tvacuolar membrane\tcellular_component\n",
      "1\tGO:0005829\tcytosol\tcellular_component\n",
      "1\tGO:0010008\tendosome membrane\tcellular_component\n",
      "1\tGO:0033565\tESCRT-0 complex\tcellular_component\n",
      "1\tGO:0005576\textracellular region\tcellular_component\n",
      "1\tGO:0005618\tcell wall\tcellular_component\n",
      "1\tGO:0031012\textracellular matrix\tcellular_component\n",
      "1\tGO:0005737\tcytoplasm\tcellular_component\n",
      "1\tGO:0043231\tintracellular membrane-bounded organelle\tcellular_component\n",
      "1\tGO:0005747\tmitochondrial respiratory chain complex I\tcellular_component\n",
      "1\tGO:0005783\tendoplasmic reticulum\tcellular_component\n",
      "1\tGO:0005794\tGolgi apparatus\tcellular_component\n"
     ]
    }
   ],
   "source": [
    "#plot_go_terms(sucrose_term_count, top_n=10)\n",
    "print(\"# <<< SUCROSE >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(sucrose_nXcv['mean_scores'])} \\n# count common features: {len(sucrose_nXcv['common_features'])}\")\n",
    "tabulate(suc_term_count_biop)\n",
    "print(\"# <<< SUCROSE >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(sucrose_nXcv['mean_scores'])} \\n# count common features: {len(sucrose_nXcv['common_features'])}\")\n",
    "tabulate(suc_term_count_molf)\n",
    "print(\"# <<< SUCROSE >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(sucrose_nXcv['mean_scores'])} \\n# count common features: {len(sucrose_nXcv['common_features'])}\")\n",
    "tabulate(suc_term_count_cellc)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'PRINT EACH GFF ENTRY AND GO-TERMS CORRESPONDING TO NO-ZERO FEATURES'"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"PRINT EACH GFF ENTRY AND GO-TERMS CORRESPONDING TO NO-ZERO FEATURES\"\"\"\n",
    "\n",
    "#for ele in sucrose_nXcv_matched:\n",
    "#    print(f\"{sucrose_nXcv_matched[ele]}\")\n",
    "#    #print(glucose_goids[ele], \"\\n\")\n",
    "#    for id in sucrose_goids[ele]:\n",
    "#        print(id, \": \", sucrose_goids[ele][id])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### citric acid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    324 of 549,396 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 18 IDs.\n",
      "100%     18 of     18 study items found in association\n",
      "100%     18 of     18 study items found in population(549396)\n",
      "Calculating 190 uncorrected p-values using fisher_scipy_stats\n",
      "     190 terms are associated with     13 of 549,396 population items\n",
      "     190 terms are associated with     13 of     18 study items\n",
      "  METHOD fdr_bh:\n",
      "     190 GO terms found significant (< 0.05=alpha) (190 enriched +   0 purified): statsmodels fdr_bh\n",
      "      13 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    324 of 549,396 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 18 IDs.\n",
      "100%     18 of     18 study items found in association\n",
      "100%     18 of     18 study items found in population(549396)\n",
      "Calculating 190 uncorrected p-values using fisher_scipy_stats\n",
      "     190 terms are associated with     13 of 549,396 population items\n",
      "     190 terms are associated with     13 of     18 study items\n",
      "  METHOD fdr_bh:\n",
      "     190 GO terms found significant (< 0.05=alpha) (190 enriched +   0 purified): statsmodels fdr_bh\n",
      "      13 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    324 of 549,396 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 18 IDs.\n",
      "100%     18 of     18 study items found in association\n",
      "100%     18 of     18 study items found in population(549396)\n",
      "Calculating 190 uncorrected p-values using fisher_scipy_stats\n",
      "     190 terms are associated with     13 of 549,396 population items\n",
      "     190 terms are associated with     13 of     18 study items\n",
      "  METHOD fdr_bh:\n",
      "     190 GO terms found significant (< 0.05=alpha) (190 enriched +   0 purified): statsmodels fdr_bh\n",
      "      13 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n"
     ]
    }
   ],
   "source": [
    "with open(\"./lasso_models/10xKfold_lasso_output/citricAcid_nXcv.json\", 'r') as file:\n",
    "    citricAcid_nXcv = json.load(file)\n",
    "#citricAcid_nXcv_matched, citricAcid_goids, citricAcid_term_count, citricAcid_goea = matchgff2(citricAcid_nXcv['common_features'], goea=True)\n",
    "citricAcid_nXcv_matched, citA_goids_biop, citA_term_count_biop, citA_goea_biop = matchgff2(citricAcid_nXcv['common_features'], namespace=['biological_process'], depth_threshold=2, goea=True)\n",
    "citricAcid_nXcv_matched, citA_goids_molf, citA_term_count_molf, citA_goea_molf = matchgff2(citricAcid_nXcv['common_features'], namespace=['molecular_function'], depth_threshold=2, goea=True)\n",
    "citricAcid_nXcv_matched, citA_goids_cellc, citA_term_count_cellc, citA_goea_cellc = matchgff2(citricAcid_nXcv['common_features'], namespace=['cellular_component'], depth_threshold=2, goea=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# <<< CITRIC ACID >>>\n",
      "# 10x3FoldCV mean score: 0.22358901669508696 \n",
      "# count common features: 18\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "2\tGO:0009686\tgibberellin biosynthetic process\tbiological_process\n",
      "2\tGO:0009826\tunidimensional cell growth\tbiological_process\n",
      "2\tGO:0048235\tpollen sperm cell differentiation\tbiological_process\n",
      "2\tGO:0055085\ttransmembrane transport\tbiological_process\n",
      "1\tGO:0009750\tresponse to fructose\tbiological_process\n",
      "1\tGO:0090333\tregulation of stomatal closure\tbiological_process\n",
      "1\tGO:0010182\tsugar mediated signaling pathway\tbiological_process\n",
      "1\tGO:0007254\tJNK cascade\tbiological_process\n",
      "1\tGO:0042733\tembryonic digit morphogenesis\tbiological_process\n",
      "1\tGO:0001666\tresponse to hypoxia\tbiological_process\n",
      "1\tGO:2000035\tregulation of stem cell division\tbiological_process\n",
      "1\tGO:1904291\tpositive regulation of mitotic DNA damage checkpoint\tbiological_process\n",
      "1\tGO:0036289\tpeptidyl-serine autophosphorylation\tbiological_process\n",
      "1\tGO:0038066\tp38MAPK cascade\tbiological_process\n",
      "1\tGO:1900424\tregulation of defense response to bacterium\tbiological_process\n",
      "1\tGO:0071480\tcellular response to gamma radiation\tbiological_process\n",
      "1\tGO:0097696\treceptor signaling pathway via STAT\tbiological_process\n",
      "1\tGO:0071244\tcellular response to carbon dioxide\tbiological_process\n",
      "1\tGO:0010105\tnegative regulation of ethylene-activated signaling pathway\tbiological_process\n",
      "1\tGO:0050765\tnegative regulation of phagocytosis\tbiological_process\n",
      "1\tGO:1900150\tregulation of defense response to fungus\tbiological_process\n",
      "1\tGO:0009651\tresponse to salt stress\tbiological_process\n",
      "1\tGO:0009658\tchloroplast organization\tbiological_process\n",
      "1\tGO:0009414\tresponse to water deprivation\tbiological_process\n",
      "1\tGO:0006935\tchemotaxis\tbiological_process\n",
      "1\tGO:0001845\tphagolysosome assembly\tbiological_process\n",
      "1\tGO:0050829\tdefense response to Gram-negative bacterium\tbiological_process\n",
      "1\tGO:2000031\tregulation of salicylic acid mediated signaling pathway\tbiological_process\n",
      "1\tGO:2000030\tregulation of response to red or far red light\tbiological_process\n",
      "1\tGO:0009744\tresponse to sucrose\tbiological_process\n",
      "1\tGO:0009788\tnegative regulation of abscisic acid-activated signaling pathway\tbiological_process\n",
      "1\tGO:0002229\tdefense response to oomycetes\tbiological_process\n",
      "1\tGO:0048510\tregulation of timing of transition from vegetative to reproductive phase\tbiological_process\n",
      "1\tGO:0051593\tresponse to folic acid\tbiological_process\n",
      "1\tGO:0043065\tpositive regulation of apoptotic process\tbiological_process\n",
      "1\tGO:2000146\tnegative regulation of cell motility\tbiological_process\n",
      "1\tGO:2000069\tregulation of post-embryonic root development\tbiological_process\n",
      "1\tGO:0030036\tactin cytoskeleton organization\tbiological_process\n",
      "1\tGO:0030435\tsporulation resulting in formation of a cellular spore\tbiological_process\n",
      "1\tGO:0000963\tmitochondrial RNA processing\tbiological_process\n",
      "1\tGO:0010214\tseed coat development\tbiological_process\n",
      "1\tGO:0016554\tcytidine to uridine editing\tbiological_process\n",
      "1\tGO:0031425\tchloroplast RNA processing\tbiological_process\n",
      "1\tGO:0031426\tpolycistronic mRNA processing\tbiological_process\n",
      "1\tGO:0080156\tmitochondrial mRNA modification\tbiological_process\n",
      "1\tGO:0140547\tacquisition of seed longevity\tbiological_process\n",
      "1\tGO:1900865\tchloroplast RNA modification\tbiological_process\n",
      "1\tGO:0000012\tsingle strand break repair\tbiological_process\n",
      "1\tGO:0006260\tDNA replication\tbiological_process\n",
      "1\tGO:0006302\tdouble-strand break repair\tbiological_process\n",
      "1\tGO:0006357\tregulation of transcription by RNA polymerase II\tbiological_process\n",
      "1\tGO:0006783\theme biosynthetic process\tbiological_process\n",
      "1\tGO:0006790\tsulfur compound metabolic process\tbiological_process\n",
      "1\tGO:0009150\tpurine ribonucleotide metabolic process\tbiological_process\n",
      "1\tGO:0009585\tred, far-red light phototransduction\tbiological_process\n",
      "1\tGO:0009704\tde-etiolation\tbiological_process\n",
      "1\tGO:0009733\tresponse to auxin\tbiological_process\n",
      "1\tGO:0009740\tgibberellic acid mediated signaling pathway\tbiological_process\n",
      "1\tGO:0009959\tnegative gravitropism\tbiological_process\n",
      "1\tGO:0010100\tnegative regulation of photomorphogenesis\tbiological_process\n",
      "1\tGO:0010161\tred light signaling pathway\tbiological_process\n",
      "1\tGO:0010187\tnegative regulation of seed germination\tbiological_process\n",
      "1\tGO:0010197\tpolar nucleus fusion\tbiological_process\n",
      "1\tGO:0015995\tchlorophyll biosynthetic process\tbiological_process\n",
      "1\tGO:0031539\tpositive regulation of anthocyanin metabolic process\tbiological_process\n",
      "1\tGO:0042548\tregulation of photosynthesis, light reaction\tbiological_process\n",
      "1\tGO:0045893\tpositive regulation of DNA-templated transcription\tbiological_process\n",
      "1\tGO:0048586\tregulation of long-day photoperiodism, flowering\tbiological_process\n",
      "1\tGO:0048587\tregulation of short-day photoperiodism, flowering\tbiological_process\n",
      "1\tGO:0048766\troot hair initiation\tbiological_process\n",
      "1\tGO:0048767\troot hair elongation\tbiological_process\n",
      "1\tGO:0080113\tregulation of seed growth\tbiological_process\n",
      "1\tGO:0090229\tnegative regulation of red or far-red light signaling pathway\tbiological_process\n",
      "1\tGO:0090305\tN/A\tN/A\n",
      "1\tGO:1905421\tregulation of plant organ morphogenesis\tbiological_process\n",
      "1\tGO:1990785\tresponse to water-immersion restraint stress\tbiological_process\n",
      "1\tGO:2000032\tregulation of secondary shoot formation\tbiological_process\n",
      "1\tGO:0006534\tcysteine metabolic process\tbiological_process\n",
      "1\tGO:0010269\tresponse to selenium ion\tbiological_process\n",
      "1\tGO:0006833\twater transport\tbiological_process\n",
      "1\tGO:0007033\tvacuole organization\tbiological_process\n",
      "1\tGO:0015840\turea transport\tbiological_process\n",
      "1\tGO:0048366\tleaf development\tbiological_process\n",
      "1\tGO:0016579\tprotein deubiquitination\tbiological_process\n",
      "1\tGO:0031647\tregulation of protein stability\tbiological_process\n",
      "1\tGO:0010506\tregulation of autophagy\tbiological_process\n",
      "1\tGO:0006613\tcotranslational protein targeting to membrane\tbiological_process\n",
      "1\tGO:0007409\taxonogenesis\tbiological_process\n",
      "1\tGO:0045047\tprotein targeting to ER\tbiological_process\n",
      "1\tGO:0050708\tregulation of protein secretion\tbiological_process\n",
      "1\tGO:0016573\thistone acetylation\tbiological_process\n",
      "1\tGO:0000226\tmicrotubule cytoskeleton organization\tbiological_process\n",
      "1\tGO:0009561\tmegagametogenesis\tbiological_process\n",
      "1\tGO:0055046\tmicrogametogenesis\tbiological_process\n",
      "1\tGO:0008202\tsteroid metabolic process\tbiological_process\n",
      "1\tGO:0048316\tseed development\tbiological_process\n",
      "1\tGO:1901564\torganonitrogen compound metabolic process\tbiological_process\n",
      "1\tGO:0048519\tnegative regulation of biological process\tbiological_process\n",
      "1\tGO:0051641\tcellular localization\tbiological_process\n",
      "1\tGO:0097006\tregulation of plasma lipoprotein particle levels\tbiological_process\n",
      "1\tGO:0034204\tlipid translocation\tbiological_process\n",
      "1\tGO:0010874\tregulation of cholesterol efflux\tbiological_process\n",
      "1\tGO:0033993\tresponse to lipid\tbiological_process\n",
      "1\tGO:0140352\texport from cell\tbiological_process\n",
      "1\tGO:0031334\tpositive regulation of protein-containing complex assembly\tbiological_process\n",
      "1\tGO:2001140\tpositive regulation of phospholipid transport\tbiological_process\n",
      "1\tGO:0046903\tsecretion\tbiological_process\n",
      "1\tGO:0071705\tnitrogen compound transport\tbiological_process\n",
      "1\tGO:0071827\tplasma lipoprotein particle organization\tbiological_process\n",
      "1\tGO:0007165\tsignal transduction\tbiological_process\n",
      "1\tGO:0055088\tlipid homeostasis\tbiological_process\n",
      "1\tGO:0006650\tglycerophospholipid metabolic process\tbiological_process\n",
      "1\tGO:0006996\torganelle organization\tbiological_process\n",
      "1\tGO:0009719\tresponse to endogenous stimulus\tbiological_process\n",
      "1\tGO:0042908\txenobiotic transport\tbiological_process\n",
      "1\tGO:0032376\tpositive regulation of cholesterol transport\tbiological_process\n",
      "1\tGO:0019216\tregulation of lipid metabolic process\tbiological_process\n",
      "1\tGO:0006123\tmitochondrial electron transport, cytochrome c to oxygen\tbiological_process\n",
      "1\tGO:0035694\tmitochondrial protein catabolic process\tbiological_process\n",
      "# <<< CITRIC ACID >>>\n",
      "# 10x3FoldCV mean score: 0.22358901669508696 \n",
      "# count common features: 18\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "3\tGO:0005524\tATP binding\tmolecular_function\n",
      "3\tGO:0005515\tprotein binding\tmolecular_function\n",
      "2\tGO:0042802\tidentical protein binding\tmolecular_function\n",
      "1\tGO:0004706\tJUN kinase kinase kinase activity\tmolecular_function\n",
      "1\tGO:0000287\tmagnesium ion binding\tmolecular_function\n",
      "1\tGO:0004713\tprotein tyrosine kinase activity\tmolecular_function\n",
      "1\tGO:0004712\tprotein serine/threonine/tyrosine kinase activity\tmolecular_function\n",
      "1\tGO:0097677\tSTAT family protein binding\tmolecular_function\n",
      "1\tGO:0003729\tmRNA binding\tmolecular_function\n",
      "1\tGO:0004519\tendonuclease activity\tmolecular_function\n",
      "1\tGO:0008270\tzinc ion binding\tmolecular_function\n",
      "1\tGO:0000978\tRNA polymerase II cis-regulatory region sequence-specific DNA binding\tmolecular_function\n",
      "1\tGO:0000981\tDNA-binding transcription factor activity, RNA polymerase II-specific\tmolecular_function\n",
      "1\tGO:0003697\tsingle-stranded DNA binding\tmolecular_function\n",
      "1\tGO:0003725\tdouble-stranded RNA binding\tmolecular_function\n",
      "1\tGO:0010313\tphytochrome binding\tmolecular_function\n",
      "1\tGO:0030983\tmismatched DNA binding\tmolecular_function\n",
      "1\tGO:0033699\tDNA 5'-adenosine monophosphate hydrolase activity\tmolecular_function\n",
      "1\tGO:0046872\tmetal ion binding\tmolecular_function\n",
      "1\tGO:0046983\tprotein dimerization activity\tmolecular_function\n",
      "1\tGO:0047627\tadenylylsulfatase activity\tmolecular_function\n",
      "1\tGO:1990165\tsingle-strand break-containing DNA binding\tmolecular_function\n",
      "1\tGO:0090305\tN/A\tN/A\n",
      "1\tGO:0009000\tselenocysteine lyase activity\tmolecular_function\n",
      "1\tGO:0030170\tpyridoxal phosphate binding\tmolecular_function\n",
      "1\tGO:0031071\tcysteine desulfurase activity\tmolecular_function\n",
      "1\tGO:0015200\tmethylammonium transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0015204\turea transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0015250\twater channel activity\tmolecular_function\n",
      "1\tGO:0008234\tcysteine-type peptidase activity\tmolecular_function\n",
      "1\tGO:0003924\tGTPase activity\tmolecular_function\n",
      "1\tGO:0005047\tsignal recognition particle binding\tmolecular_function\n",
      "1\tGO:0097159\torganic cyclic compound binding\tmolecular_function\n",
      "1\tGO:1901363\theterocyclic compound binding\tmolecular_function\n",
      "1\tGO:0051011\tmicrotubule minus-end binding\tmolecular_function\n",
      "1\tGO:0120013\tlipid transfer activity\tmolecular_function\n",
      "1\tGO:0008525\tphosphatidylcholine transporter activity\tmolecular_function\n",
      "1\tGO:0140359\tABC-type transporter activity\tmolecular_function\n",
      "1\tGO:0015245\tfatty acid transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0140328\tfloppase activity\tmolecular_function\n",
      "# <<< CITRIC ACID >>>\n",
      "# 10x3FoldCV mean score: 0.22358901669508696 \n",
      "# count common features: 18\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "2\tGO:0005789\tendoplasmic reticulum membrane\tcellular_component\n",
      "2\tGO:0005886\tplasma membrane\tcellular_component\n",
      "2\tGO:0005634\tnucleus\tcellular_component\n",
      "2\tGO:0005739\tmitochondrion\tcellular_component\n",
      "2\tGO:0016021\tmembrane\tcellular_component\n",
      "2\tGO:0016020\tmembrane\tcellular_component\n",
      "1\tGO:0032010\tphagolysosome\tcellular_component\n",
      "1\tGO:0005829\tcytosol\tcellular_component\n",
      "1\tGO:0012510\ttrans-Golgi network transport vesicle membrane\tcellular_component\n",
      "1\tGO:0005769\tearly endosome\tcellular_component\n",
      "1\tGO:0005802\ttrans-Golgi network\tcellular_component\n",
      "1\tGO:0005938\tcell cortex\tcellular_component\n",
      "1\tGO:0030670\tphagocytic vesicle membrane\tcellular_component\n",
      "1\tGO:0009507\tchloroplast\tcellular_component\n",
      "1\tGO:0090305\tN/A\tN/A\n",
      "1\tGO:0009570\tchloroplast stroma\tcellular_component\n",
      "1\tGO:0032586\tprotein storage vacuole membrane\tcellular_component\n",
      "1\tGO:0042807\tcentral vacuole\tcellular_component\n",
      "1\tGO:0090406\tpollen tube\tcellular_component\n",
      "1\tGO:0005737\tcytoplasm\tcellular_component\n",
      "1\tGO:0005654\tnucleoplasm\tcellular_component\n",
      "1\tGO:0035267\tNuA4 histone acetyltransferase complex\tcellular_component\n",
      "1\tGO:0005819\tspindle\tcellular_component\n",
      "1\tGO:0005874\tmicrotubule\tcellular_component\n",
      "1\tGO:0009524\tphragmoplast\tcellular_component\n",
      "1\tGO:0005770\tlate endosome\tcellular_component\n",
      "1\tGO:0005794\tGolgi apparatus\tcellular_component\n",
      "1\tGO:0097208\talveolar lamellar body\tcellular_component\n",
      "1\tGO:0005783\tendoplasmic reticulum\tcellular_component\n",
      "1\tGO:0005764\tlysosome\tcellular_component\n",
      "1\tGO:0097232\tlamellar body membrane\tcellular_component\n",
      "1\tGO:0031966\tmitochondrial membrane\tcellular_component\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#plot_go_terms(citricAcid_term_count, top_n=10)\n",
    "print(\"# <<< CITRIC ACID >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(citricAcid_nXcv['mean_scores'])} \\n# count common features: {len(citricAcid_nXcv['common_features'])}\")\n",
    "tabulate(citA_term_count_biop)\n",
    "print(\"# <<< CITRIC ACID >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(citricAcid_nXcv['mean_scores'])} \\n# count common features: {len(citricAcid_nXcv['common_features'])}\")\n",
    "tabulate(citA_term_count_molf)\n",
    "print(\"# <<< CITRIC ACID >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(citricAcid_nXcv['mean_scores'])} \\n# count common features: {len(citricAcid_nXcv['common_features'])}\")\n",
    "tabulate(citA_term_count_cellc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'PRINT EACH GFF ENTRY AND GO-TERMS CORRESPONDING TO NO-ZERO FEATURES'"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"PRINT EACH GFF ENTRY AND GO-TERMS CORRESPONDING TO NO-ZERO FEATURES\"\"\"\n",
    "\n",
    "#for ele in citricAcid_nXcv_matched:\n",
    "#    print(f\"{citricAcid_nXcv_matched[ele]}\")\n",
    "#    #print(glucose_goids[ele], \"\\n\")\n",
    "#    for id in citricAcid_goids[ele]:\n",
    "#        print(id, \": \", citricAcid_goids[ele][id])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%     81 of 274,698 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 9 IDs.\n",
      "100%      9 of      9 study items found in association\n",
      "100%      9 of      9 study items found in population(274698)\n",
      "Calculating 136 uncorrected p-values using fisher_scipy_stats\n",
      "     136 terms are associated with      8 of 274,698 population items\n",
      "     136 terms are associated with      8 of      9 study items\n",
      "  METHOD fdr_bh:\n",
      "     136 GO terms found significant (< 0.05=alpha) (136 enriched +   0 purified): statsmodels fdr_bh\n",
      "       8 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%     81 of 274,698 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 9 IDs.\n",
      "100%      9 of      9 study items found in association\n",
      "100%      9 of      9 study items found in population(274698)\n",
      "Calculating 136 uncorrected p-values using fisher_scipy_stats\n",
      "     136 terms are associated with      8 of 274,698 population items\n",
      "     136 terms are associated with      8 of      9 study items\n",
      "  METHOD fdr_bh:\n",
      "     136 GO terms found significant (< 0.05=alpha) (136 enriched +   0 purified): statsmodels fdr_bh\n",
      "       8 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%     81 of 274,698 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 9 IDs.\n",
      "100%      9 of      9 study items found in association\n",
      "100%      9 of      9 study items found in population(274698)\n",
      "Calculating 136 uncorrected p-values using fisher_scipy_stats\n",
      "     136 terms are associated with      8 of 274,698 population items\n",
      "     136 terms are associated with      8 of      9 study items\n",
      "  METHOD fdr_bh:\n",
      "     136 GO terms found significant (< 0.05=alpha) (136 enriched +   0 purified): statsmodels fdr_bh\n",
      "       8 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n"
     ]
    }
   ],
   "source": [
    "with open(\"./lasso_models/10xKfold_lasso_output/isoleucine_nXcv.json\", 'r') as file:\n",
    "    isoleucine_nXcv = json.load(file)\n",
    "#isoleucine_nXcv_matched, isoleucine_goids, isoleucine_term_count, isoleucine_goea = matchgff2(isoleucine_nXcv['common_features'], goea=True) #['biological_process', 'molecular_function', 'cellular_component']\n",
    "isoleucine_nXcv_matched, ile_goids_biop, ile_term_count_biop, ile_goea_biop = matchgff2(isoleucine_nXcv['common_features'], namespace=['biological_process', 'molecular_function'], depth_threshold=2, goea=True)\n",
    "isoleucine_nXcv_matched, ile_goids_molf, ile_term_count_molf, ile_goea_molf = matchgff2(isoleucine_nXcv['common_features'], namespace=['molecular_function'], depth_threshold=2, goea=True)\n",
    "isoleucine_nXcv_matched, ile_goids_cellc, ile_term_count_cellc, ile_goea_cellc = matchgff2(isoleucine_nXcv['common_features'], namespace=['cellular_component'], depth_threshold=2, goea=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# <<< ISOLEUCINE >>>\n",
      "# 10x3FoldCV mean score: 0.07529144560758311 \n",
      "# count common features: 9\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "2\tGO:0000166\tnucleotide binding\tmolecular_function\n",
      "2\tGO:0050832\tdefense response to fungus\tbiological_process\n",
      "2\tGO:0042803\tprotein homodimerization activity\tmolecular_function\n",
      "2\tGO:0004674\tprotein serine/threonine kinase activity\tmolecular_function\n",
      "2\tGO:0005515\tprotein binding\tmolecular_function\n",
      "2\tGO:0045893\tpositive regulation of DNA-templated transcription\tbiological_process\n",
      "2\tGO:0009737\tresponse to abscisic acid\tbiological_process\n",
      "1\tGO:0042910\txenobiotic transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0032370\tpositive regulation of lipid transport\tbiological_process\n",
      "1\tGO:0055076\tN/A\tN/A\n",
      "1\tGO:0009791\tpost-embryonic development\tbiological_process\n",
      "1\tGO:0051173\tpositive regulation of nitrogen compound metabolic process\tbiological_process\n",
      "1\tGO:0006820\tmonoatomic anion transport\tbiological_process\n",
      "1\tGO:1901700\tresponse to oxygen-containing compound\tbiological_process\n",
      "1\tGO:0009733\tresponse to auxin\tbiological_process\n",
      "1\tGO:0005524\tATP binding\tmolecular_function\n",
      "1\tGO:0005548\tphospholipid transporter activity\tmolecular_function\n",
      "1\tGO:0009605\tresponse to external stimulus\tbiological_process\n",
      "1\tGO:0071310\tcellular response to organic substance\tbiological_process\n",
      "1\tGO:0034204\tlipid translocation\tbiological_process\n",
      "1\tGO:0140115\texport across plasma membrane\tbiological_process\n",
      "1\tGO:0033993\tresponse to lipid\tbiological_process\n",
      "1\tGO:0042626\tATPase-coupled transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0042802\tidentical protein binding\tmolecular_function\n",
      "1\tGO:0014070\tresponse to organic cyclic compound\tbiological_process\n",
      "1\tGO:0060918\tauxin transport\tbiological_process\n",
      "1\tGO:0009639\tresponse to red or far red light\tbiological_process\n",
      "1\tGO:0031328\tpositive regulation of cellular biosynthetic process\tbiological_process\n",
      "1\tGO:0048731\tsystem development\tbiological_process\n",
      "1\tGO:0140328\tfloppase activity\tmolecular_function\n",
      "1\tGO:0015914\tphospholipid transport\tbiological_process\n",
      "1\tGO:0042908\txenobiotic transport\tbiological_process\n",
      "1\tGO:0032491\tdetection of molecule of fungal origin\tbiological_process\n",
      "1\tGO:0008061\tchitin binding\tmolecular_function\n",
      "1\tGO:0090333\tregulation of stomatal closure\tbiological_process\n",
      "1\tGO:0046777\tprotein autophosphorylation\tbiological_process\n",
      "1\tGO:0048364\troot development\tbiological_process\n",
      "1\tGO:0002752\tcell surface pattern recognition receptor signaling pathway\tbiological_process\n",
      "1\tGO:0019722\tcalcium-mediated signaling\tbiological_process\n",
      "1\tGO:0071219\tcellular response to molecule of bacterial origin\tbiological_process\n",
      "1\tGO:0015026\tcoreceptor activity\tmolecular_function\n",
      "1\tGO:0009862\tsystemic acquired resistance, salicylic acid mediated signaling pathway\tbiological_process\n",
      "1\tGO:2001080\tchitosan binding\tmolecular_function\n",
      "1\tGO:1900425\tnegative regulation of defense response to bacterium\tbiological_process\n",
      "1\tGO:0009845\tseed germination\tbiological_process\n",
      "1\tGO:1900426\tpositive regulation of defense response to bacterium\tbiological_process\n",
      "1\tGO:0007639\thomeostasis of number of meristem cells\tbiological_process\n",
      "1\tGO:0004842\tubiquitin-protein transferase activity\tmolecular_function\n",
      "1\tGO:0009738\tabscisic acid-activated signaling pathway\tbiological_process\n",
      "1\tGO:0002229\tdefense response to oomycetes\tbiological_process\n",
      "1\tGO:0048653\tanther development\tbiological_process\n",
      "1\tGO:0140426\tpathogen-associated molecular pattern receptor signaling pathway\tbiological_process\n",
      "1\tGO:0019199\ttransmembrane receptor protein kinase activity\tmolecular_function\n",
      "1\tGO:0010726\tpositive regulation of hydrogen peroxide metabolic process\tbiological_process\n",
      "1\tGO:0032499\tdetection of peptidoglycan\tbiological_process\n",
      "1\tGO:0071323\tcellular response to chitin\tbiological_process\n",
      "1\tGO:0010942\tN/A\tN/A\n",
      "1\tGO:0016567\tprotein ubiquitination\tbiological_process\n",
      "1\tGO:0005375\tcopper ion transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0005381\tiron ion transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0005385\tzinc ion transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0006826\tiron ion transport\tbiological_process\n",
      "1\tGO:0009617\tresponse to bacterium\tbiological_process\n",
      "1\tGO:0010043\tresponse to zinc ion\tbiological_process\n",
      "1\tGO:0015675\tnickel cation transport\tbiological_process\n",
      "1\tGO:0071577\tzinc ion transmembrane transport\tbiological_process\n",
      "1\tGO:0030176\tN/A\tN/A\n",
      "1\tGO:0004252\tserine-type endopeptidase activity\tmolecular_function\n",
      "1\tGO:0051787\tmisfolded protein binding\tmolecular_function\n",
      "1\tGO:1990381\tubiquitin-specific protease binding\tmolecular_function\n",
      "1\tGO:0007283\tspermatogenesis\tbiological_process\n",
      "1\tGO:0030154\tcell differentiation\tbiological_process\n",
      "1\tGO:0030433\tubiquitin-dependent ERAD pathway\tbiological_process\n",
      "1\tGO:0030968\tendoplasmic reticulum unfolded protein response\tbiological_process\n",
      "1\tGO:0033619\tmembrane protein proteolysis\tbiological_process\n",
      "1\tGO:0051247\tpositive regulation of protein metabolic process\tbiological_process\n",
      "1\tGO:1902074\tresponse to salt\tbiological_process\n",
      "1\tGO:0009651\tresponse to salt stress\tbiological_process\n",
      "1\tGO:0090058\tmetaxylem development\tbiological_process\n",
      "1\tGO:0090059\tprotoxylem development\tbiological_process\n",
      "1\tGO:0043068\tpositive regulation of programmed cell death\tbiological_process\n",
      "1\tGO:0000976\ttranscription cis-regulatory region binding\tmolecular_function\n",
      "1\tGO:1900057\tpositive regulation of leaf senescence\tbiological_process\n",
      "1\tGO:0051091\tpositive regulation of DNA-binding transcription factor activity\tbiological_process\n",
      "1\tGO:1901348\tpositive regulation of secondary cell wall biogenesis\tbiological_process\n",
      "1\tGO:0009735\tresponse to cytokinin\tbiological_process\n",
      "1\tGO:0048759\txylem vessel member cell differentiation\tbiological_process\n",
      "1\tGO:0009734\tauxin-activated signaling pathway\tbiological_process\n",
      "1\tGO:1903648\tpositive regulation of chlorophyll catabolic process\tbiological_process\n",
      "1\tGO:1990110\tcallus formation\tbiological_process\n",
      "1\tGO:0003700\tDNA-binding transcription factor activity\tmolecular_function\n",
      "1\tGO:0045491\txylan metabolic process\tbiological_process\n",
      "1\tGO:0046982\tprotein heterodimerization activity\tmolecular_function\n",
      "1\tGO:0051782\tnegative regulation of cell division\tbiological_process\n",
      "1\tGO:0071555\tcell wall organization\tbiological_process\n",
      "1\tGO:1904250\tpositive regulation of age-related resistance\tbiological_process\n",
      "1\tGO:0010199\torgan boundary specification between lateral organs and the meristem\tbiological_process\n",
      "1\tGO:0048829\troot cap development\tbiological_process\n",
      "1\tGO:0090400\tstress-induced premature senescence\tbiological_process\n",
      "1\tGO:0080187\tfloral organ senescence\tbiological_process\n",
      "1\tGO:0010072\tprimary shoot apical meristem specification\tbiological_process\n",
      "1\tGO:0009741\tresponse to brassinosteroid\tbiological_process\n",
      "1\tGO:0042542\tresponse to hydrogen peroxide\tbiological_process\n",
      "1\tGO:0009723\tresponse to ethylene\tbiological_process\n",
      "1\tGO:0048527\tlateral root development\tbiological_process\n",
      "1\tGO:0010628\tpositive regulation of gene expression\tbiological_process\n",
      "1\tGO:0090709\tregulation of timing of plant organ formation\tbiological_process\n",
      "1\tGO:0010029\tregulation of seed germination\tbiological_process\n",
      "1\tGO:0010981\tregulation of cell wall macromolecule metabolic process\tbiological_process\n",
      "1\tGO:0010223\tsecondary shoot formation\tbiological_process\n",
      "1\tGO:0045739\tpositive regulation of DNA repair\tbiological_process\n",
      "1\tGO:0043967\thistone H4 acetylation\tbiological_process\n",
      "1\tGO:1900095\tN/A\tN/A\n",
      "1\tGO:0051053\tnegative regulation of DNA metabolic process\tbiological_process\n",
      "1\tGO:0003727\tsingle-stranded RNA binding\tmolecular_function\n",
      "1\tGO:0045995\tregulation of embryonic development\tbiological_process\n",
      "1\tGO:1904507\tpositive regulation of telomere maintenance in response to DNA damage\tbiological_process\n",
      "1\tGO:0002237\tresponse to molecule of bacterial origin\tbiological_process\n",
      "1\tGO:0006468\tprotein phosphorylation\tbiological_process\n",
      "1\tGO:0009625\tresponse to insect\tbiological_process\n",
      "1\tGO:0042742\tdefense response to bacterium\tbiological_process\n",
      "1\tGO:0050826\tresponse to freezing\tbiological_process\n",
      "# <<< ISOLEUCINE >>>\n",
      "# 10x3FoldCV mean score: 0.07529144560758311 \n",
      "# count common features: 9\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "2\tGO:0000166\tnucleotide binding\tmolecular_function\n",
      "2\tGO:0042803\tprotein homodimerization activity\tmolecular_function\n",
      "2\tGO:0004674\tprotein serine/threonine kinase activity\tmolecular_function\n",
      "2\tGO:0005515\tprotein binding\tmolecular_function\n",
      "1\tGO:0042910\txenobiotic transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0055076\tN/A\tN/A\n",
      "1\tGO:0005524\tATP binding\tmolecular_function\n",
      "1\tGO:0005548\tphospholipid transporter activity\tmolecular_function\n",
      "1\tGO:0042626\tATPase-coupled transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0042802\tidentical protein binding\tmolecular_function\n",
      "1\tGO:0140328\tfloppase activity\tmolecular_function\n",
      "1\tGO:0008061\tchitin binding\tmolecular_function\n",
      "1\tGO:0015026\tcoreceptor activity\tmolecular_function\n",
      "1\tGO:2001080\tchitosan binding\tmolecular_function\n",
      "1\tGO:0004842\tubiquitin-protein transferase activity\tmolecular_function\n",
      "1\tGO:0019199\ttransmembrane receptor protein kinase activity\tmolecular_function\n",
      "1\tGO:0010942\tN/A\tN/A\n",
      "1\tGO:0005375\tcopper ion transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0005381\tiron ion transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0005385\tzinc ion transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0030176\tN/A\tN/A\n",
      "1\tGO:0004252\tserine-type endopeptidase activity\tmolecular_function\n",
      "1\tGO:0051787\tmisfolded protein binding\tmolecular_function\n",
      "1\tGO:1990381\tubiquitin-specific protease binding\tmolecular_function\n",
      "1\tGO:0000976\ttranscription cis-regulatory region binding\tmolecular_function\n",
      "1\tGO:0003700\tDNA-binding transcription factor activity\tmolecular_function\n",
      "1\tGO:0046982\tprotein heterodimerization activity\tmolecular_function\n",
      "1\tGO:1900095\tN/A\tN/A\n",
      "1\tGO:0003727\tsingle-stranded RNA binding\tmolecular_function\n",
      "# <<< ISOLEUCINE >>>\n",
      "# 10x3FoldCV mean score: 0.07529144560758311 \n",
      "# count common features: 9\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "4\tGO:0005886\tplasma membrane\tcellular_component\n",
      "2\tGO:0016021\tmembrane\tcellular_component\n",
      "2\tGO:0005737\tcytoplasm\tcellular_component\n",
      "1\tGO:0055076\tN/A\tN/A\n",
      "1\tGO:0009706\tchloroplast inner membrane\tcellular_component\n",
      "1\tGO:0005911\tcell-cell junction\tcellular_component\n",
      "1\tGO:0031966\tmitochondrial membrane\tcellular_component\n",
      "1\tGO:0012505\tendomembrane system\tcellular_component\n",
      "1\tGO:0010942\tN/A\tN/A\n",
      "1\tGO:0005768\tendosome\tcellular_component\n",
      "1\tGO:0005773\tvacuole\tcellular_component\n",
      "1\tGO:0005802\ttrans-Golgi network\tcellular_component\n",
      "1\tGO:0009507\tchloroplast\tcellular_component\n",
      "1\tGO:0009579\tthylakoid\tcellular_component\n",
      "1\tGO:0000839\tHrd1p ubiquitin ligase ERAD-L complex\tcellular_component\n",
      "1\tGO:0030176\tN/A\tN/A\n",
      "1\tGO:0009531\tsecondary cell wall\tcellular_component\n",
      "1\tGO:0005634\tnucleus\tcellular_component\n",
      "1\tGO:1900095\tN/A\tN/A\n",
      "1\tGO:0043232\tintracellular non-membrane-bounded organelle\tcellular_component\n",
      "1\tGO:0000123\thistone acetyltransferase complex\tcellular_component\n"
     ]
    }
   ],
   "source": [
    "# plot_go_terms(isoleucine_term_count, top_n=10)\n",
    "print(\"# <<< ISOLEUCINE >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(isoleucine_nXcv['mean_scores'])} \\n# count common features: {len(isoleucine_nXcv['common_features'])}\")\n",
    "tabulate(ile_term_count_biop)\n",
    "print(\"# <<< ISOLEUCINE >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(isoleucine_nXcv['mean_scores'])} \\n# count common features: {len(isoleucine_nXcv['common_features'])}\")\n",
    "tabulate(ile_term_count_molf)\n",
    "print(\"# <<< ISOLEUCINE >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(isoleucine_nXcv['mean_scores'])} \\n# count common features: {len(isoleucine_nXcv['common_features'])}\")\n",
    "tabulate(ile_term_count_cellc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    225 of 457,830 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 15 IDs.\n",
      "100%     15 of     15 study items found in association\n",
      "100%     15 of     15 study items found in population(457830)\n",
      "Calculating 119 uncorrected p-values using fisher_scipy_stats\n",
      "     119 terms are associated with     11 of 457,830 population items\n",
      "     119 terms are associated with     11 of     15 study items\n",
      "  METHOD fdr_bh:\n",
      "     119 GO terms found significant (< 0.05=alpha) (119 enriched +   0 purified): statsmodels fdr_bh\n",
      "      11 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    225 of 457,830 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 15 IDs.\n",
      "100%     15 of     15 study items found in association\n",
      "100%     15 of     15 study items found in population(457830)\n",
      "Calculating 119 uncorrected p-values using fisher_scipy_stats\n",
      "     119 terms are associated with     11 of 457,830 population items\n",
      "     119 terms are associated with     11 of     15 study items\n",
      "  METHOD fdr_bh:\n",
      "     119 GO terms found significant (< 0.05=alpha) (119 enriched +   0 purified): statsmodels fdr_bh\n",
      "      11 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    225 of 457,830 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 15 IDs.\n",
      "100%     15 of     15 study items found in association\n",
      "100%     15 of     15 study items found in population(457830)\n",
      "Calculating 119 uncorrected p-values using fisher_scipy_stats\n",
      "     119 terms are associated with     11 of 457,830 population items\n",
      "     119 terms are associated with     11 of     15 study items\n",
      "  METHOD fdr_bh:\n",
      "     119 GO terms found significant (< 0.05=alpha) (119 enriched +   0 purified): statsmodels fdr_bh\n",
      "      11 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n"
     ]
    }
   ],
   "source": [
    "with open(\"./lasso_models/10xKfold_lasso_output/leucine_nXcv.json\", 'r') as file:\n",
    "    leucine_nXcv = json.load(file)\n",
    "#leucine_nXcv_matched, leucine_goids, leucine_term_count, leucine_goea = matchgff2(leucine_nXcv['common_features'], goea=True)\n",
    "leucine_nXcv_matched, leu_goids_biop, leu_term_count_biop, leu_goea_biop = matchgff2(leucine_nXcv['common_features'], namespace=['biological_process', 'molecular_function'], depth_threshold=2, goea=True)\n",
    "leucine_nXcv_matched, leu_goids_molf, leu_term_count_molf, leu_goea_molf = matchgff2(leucine_nXcv['common_features'], namespace=['molecular_function'], depth_threshold=2, goea=True)\n",
    "leucine_nXcv_matched, leu_goids_cellc, leu_term_count_cellc, leu_goea_cellc = matchgff2(leucine_nXcv['common_features'], namespace=['cellular_component'], depth_threshold=2, goea=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# <<< LEUCINE >>>\n",
      "# 10x3FoldCV mean score: 0.39988802939102847 \n",
      "# count common features: 15\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "6\tGO:0005515\tprotein binding\tmolecular_function\n",
      "3\tGO:0004674\tprotein serine/threonine kinase activity\tmolecular_function\n",
      "2\tGO:0046777\tprotein autophosphorylation\tbiological_process\n",
      "2\tGO:0048364\troot development\tbiological_process\n",
      "2\tGO:0000166\tnucleotide binding\tmolecular_function\n",
      "2\tGO:0019722\tcalcium-mediated signaling\tbiological_process\n",
      "2\tGO:0009845\tseed germination\tbiological_process\n",
      "2\tGO:0009738\tabscisic acid-activated signaling pathway\tbiological_process\n",
      "2\tGO:0002237\tresponse to molecule of bacterial origin\tbiological_process\n",
      "2\tGO:0009625\tresponse to insect\tbiological_process\n",
      "2\tGO:0042742\tdefense response to bacterium\tbiological_process\n",
      "2\tGO:0050826\tresponse to freezing\tbiological_process\n",
      "1\tGO:0016018\tcyclosporin A binding\tmolecular_function\n",
      "1\tGO:0009742\tbrassinosteroid mediated signaling pathway\tbiological_process\n",
      "1\tGO:0000413\tprotein peptidyl-prolyl isomerization\tbiological_process\n",
      "1\tGO:0009414\tresponse to water deprivation\tbiological_process\n",
      "1\tGO:0003755\tpeptidyl-prolyl cis-trans isomerase activity\tmolecular_function\n",
      "1\tGO:0061083\tregulation of protein refolding\tbiological_process\n",
      "1\tGO:0007623\tcircadian rhythm\tbiological_process\n",
      "1\tGO:0071482\tcellular response to light stimulus\tbiological_process\n",
      "1\tGO:0032491\tdetection of molecule of fungal origin\tbiological_process\n",
      "1\tGO:0008061\tchitin binding\tmolecular_function\n",
      "1\tGO:0090333\tregulation of stomatal closure\tbiological_process\n",
      "1\tGO:0002752\tcell surface pattern recognition receptor signaling pathway\tbiological_process\n",
      "1\tGO:0071219\tcellular response to molecule of bacterial origin\tbiological_process\n",
      "1\tGO:0050832\tdefense response to fungus\tbiological_process\n",
      "1\tGO:0015026\tcoreceptor activity\tmolecular_function\n",
      "1\tGO:0042803\tprotein homodimerization activity\tmolecular_function\n",
      "1\tGO:0009862\tsystemic acquired resistance, salicylic acid mediated signaling pathway\tbiological_process\n",
      "1\tGO:2001080\tchitosan binding\tmolecular_function\n",
      "1\tGO:1900425\tnegative regulation of defense response to bacterium\tbiological_process\n",
      "1\tGO:1900426\tpositive regulation of defense response to bacterium\tbiological_process\n",
      "1\tGO:0007639\thomeostasis of number of meristem cells\tbiological_process\n",
      "1\tGO:0004842\tubiquitin-protein transferase activity\tmolecular_function\n",
      "1\tGO:0002229\tdefense response to oomycetes\tbiological_process\n",
      "1\tGO:0048653\tanther development\tbiological_process\n",
      "1\tGO:0140426\tpathogen-associated molecular pattern receptor signaling pathway\tbiological_process\n",
      "1\tGO:0019199\ttransmembrane receptor protein kinase activity\tmolecular_function\n",
      "1\tGO:0010726\tpositive regulation of hydrogen peroxide metabolic process\tbiological_process\n",
      "1\tGO:0032499\tdetection of peptidoglycan\tbiological_process\n",
      "1\tGO:0071323\tcellular response to chitin\tbiological_process\n",
      "1\tGO:0010942\tN/A\tN/A\n",
      "1\tGO:0016567\tprotein ubiquitination\tbiological_process\n",
      "1\tGO:0003677\tDNA binding\tmolecular_function\n",
      "1\tGO:0008094\tATP-dependent activity, acting on DNA\tmolecular_function\n",
      "1\tGO:0016787\thydrolase activity\tmolecular_function\n",
      "1\tGO:0001701\tin utero embryonic development\tbiological_process\n",
      "1\tGO:0006312\tmitotic recombination\tbiological_process\n",
      "1\tGO:0007131\treciprocal meiotic recombination\tbiological_process\n",
      "1\tGO:0022607\tcellular component assembly\tbiological_process\n",
      "1\tGO:0045003\tdouble-strand break repair via synthesis-dependent strand annealing\tbiological_process\n",
      "1\tGO:0051128\tregulation of cellular component organization\tbiological_process\n",
      "1\tGO:0090068\tpositive regulation of cell cycle process\tbiological_process\n",
      "1\tGO:0090737\ttelomere maintenance via telomere trimming\tbiological_process\n",
      "1\tGO:1901990\tregulation of mitotic cell cycle phase transition\tbiological_process\n",
      "1\tGO:0019901\tprotein kinase binding\tmolecular_function\n",
      "1\tGO:0002221\tpattern recognition receptor signaling pathway\tbiological_process\n",
      "1\tGO:0045087\tinnate immune response\tbiological_process\n",
      "1\tGO:1900459\tpositive regulation of brassinosteroid mediated signaling pathway\tbiological_process\n",
      "1\tGO:0000978\tRNA polymerase II cis-regulatory region sequence-specific DNA binding\tmolecular_function\n",
      "1\tGO:0000981\tDNA-binding transcription factor activity, RNA polymerase II-specific\tmolecular_function\n",
      "1\tGO:0006357\tregulation of transcription by RNA polymerase II\tbiological_process\n",
      "1\tGO:0009416\tresponse to light stimulus\tbiological_process\n",
      "1\tGO:0009722\tdetection of cytokinin stimulus\tbiological_process\n",
      "1\tGO:0009727\tdetection of ethylene stimulus\tbiological_process\n",
      "1\tGO:0010089\txylem development\tbiological_process\n",
      "1\tGO:0010192\tmucilage biosynthetic process\tbiological_process\n",
      "1\tGO:0010214\tseed coat development\tbiological_process\n",
      "1\tGO:0045892\tnegative regulation of DNA-templated transcription\tbiological_process\n",
      "1\tGO:0048437\tfloral organ development\tbiological_process\n",
      "1\tGO:0048438\tfloral whorl development\tbiological_process\n",
      "1\tGO:0048827\tphyllome development\tbiological_process\n",
      "1\tGO:0071345\tcellular response to cytokine stimulus\tbiological_process\n",
      "1\tGO:0071369\tcellular response to ethylene stimulus\tbiological_process\n",
      "1\tGO:0080001\tmucilage extrusion from seed coat\tbiological_process\n",
      "1\tGO:0080112\tseed growth\tbiological_process\n",
      "1\tGO:2000652\tregulation of secondary cell wall biogenesis\tbiological_process\n",
      "1\tGO:0006810\ttransport\tbiological_process\n",
      "1\tGO:0019058\tviral life cycle\tbiological_process\n",
      "1\tGO:0030154\tcell differentiation\tbiological_process\n",
      "1\tGO:0031323\tregulation of cellular metabolic process\tbiological_process\n",
      "1\tGO:0048523\tnegative regulation of cellular process\tbiological_process\n",
      "1\tGO:0051171\tregulation of nitrogen compound metabolic process\tbiological_process\n",
      "1\tGO:0060255\tregulation of macromolecule metabolic process\tbiological_process\n",
      "1\tGO:0080090\tregulation of primary metabolic process\tbiological_process\n",
      "1\tGO:0003729\tmRNA binding\tmolecular_function\n",
      "1\tGO:0003735\tstructural constituent of ribosome\tmolecular_function\n",
      "1\tGO:0002181\tcytoplasmic translation\tbiological_process\n",
      "1\tGO:0016236\tmacroautophagy\tbiological_process\n",
      "1\tGO:0042273\tribosomal large subunit biogenesis\tbiological_process\n",
      "1\tGO:0006468\tprotein phosphorylation\tbiological_process\n",
      "1\tGO:0009737\tresponse to abscisic acid\tbiological_process\n",
      "1\tGO:0008168\tmethyltransferase activity\tmolecular_function\n",
      "1\tGO:0000154\trRNA modification\tbiological_process\n",
      "1\tGO:0001510\tRNA methylation\tbiological_process\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "6\tGO:0005515\tprotein binding\tmolecular_function\n",
      "3\tGO:0004674\tprotein serine/threonine kinase activity\tmolecular_function\n",
      "2\tGO:0000166\tnucleotide binding\tmolecular_function\n",
      "1\tGO:0016018\tcyclosporin A binding\tmolecular_function\n",
      "1\tGO:0003755\tpeptidyl-prolyl cis-trans isomerase activity\tmolecular_function\n",
      "1\tGO:0008061\tchitin binding\tmolecular_function\n",
      "1\tGO:0015026\tcoreceptor activity\tmolecular_function\n",
      "1\tGO:0042803\tprotein homodimerization activity\tmolecular_function\n",
      "1\tGO:2001080\tchitosan binding\tmolecular_function\n",
      "1\tGO:0004842\tubiquitin-protein transferase activity\tmolecular_function\n",
      "1\tGO:0019199\ttransmembrane receptor protein kinase activity\tmolecular_function\n",
      "1\tGO:0010942\tN/A\tN/A\n",
      "1\tGO:0003677\tDNA binding\tmolecular_function\n",
      "1\tGO:0008094\tATP-dependent activity, acting on DNA\tmolecular_function\n",
      "1\tGO:0016787\thydrolase activity\tmolecular_function\n",
      "1\tGO:0019901\tprotein kinase binding\tmolecular_function\n",
      "1\tGO:0000978\tRNA polymerase II cis-regulatory region sequence-specific DNA binding\tmolecular_function\n",
      "1\tGO:0000981\tDNA-binding transcription factor activity, RNA polymerase II-specific\tmolecular_function\n",
      "1\tGO:0003729\tmRNA binding\tmolecular_function\n",
      "1\tGO:0003735\tstructural constituent of ribosome\tmolecular_function\n",
      "1\tGO:0008168\tmethyltransferase activity\tmolecular_function\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "3\tGO:0005886\tplasma membrane\tcellular_component\n",
      "2\tGO:0005829\tcytosol\tcellular_component\n",
      "1\tGO:0005783\tendoplasmic reticulum\tcellular_component\n",
      "1\tGO:0005771\tmultivesicular body\tcellular_component\n",
      "1\tGO:0005576\textracellular region\tcellular_component\n",
      "1\tGO:0005795\tGolgi stack\tcellular_component\n",
      "1\tGO:0016020\tmembrane\tcellular_component\n",
      "1\tGO:0009570\tchloroplast stroma\tcellular_component\n",
      "1\tGO:0009941\tchloroplast envelope\tcellular_component\n",
      "1\tGO:0098807\tchloroplast thylakoid membrane protein complex\tcellular_component\n",
      "1\tGO:0010942\tN/A\tN/A\n",
      "1\tGO:0005654\tnucleoplasm\tcellular_component\n",
      "1\tGO:0005694\tchromosome\tcellular_component\n",
      "1\tGO:0005739\tmitochondrion\tcellular_component\n",
      "1\tGO:0033061\tDNA recombinase mediator complex\tcellular_component\n",
      "1\tGO:0140513\tnuclear protein-containing complex\tcellular_component\n",
      "1\tGO:0005777\tperoxisome\tcellular_component\n",
      "1\tGO:0005634\tnucleus\tcellular_component\n",
      "1\tGO:0005770\tlate endosome\tcellular_component\n",
      "1\tGO:0010008\tendosome membrane\tcellular_component\n",
      "1\tGO:0043232\tintracellular non-membrane-bounded organelle\tcellular_component\n",
      "1\tGO:0014069\tpostsynaptic density\tcellular_component\n",
      "1\tGO:0022625\tcytosolic large ribosomal subunit\tcellular_component\n",
      "1\tGO:0042788\tpolysomal ribosome\tcellular_component\n",
      "1\tGO:0098556\tcytoplasmic side of rough endoplasmic reticulum membrane\tcellular_component\n"
     ]
    }
   ],
   "source": [
    "# plot_go_terms(leucine_term_count, top_n=10)\n",
    "print(\"# <<< LEUCINE >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(leucine_nXcv['mean_scores'])} \\n# count common features: {len(leucine_nXcv['common_features'])}\")\n",
    "tabulate(leu_term_count_biop)\n",
    "print(\"# <<< LEUCINE >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(leucine_nXcv['mean_scores'])} \\n# count common features: {len(leucine_nXcv['common_features'])}\")\n",
    "tabulate(leu_term_count_molf)\n",
    "tabulate(leu_term_count_cellc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    169 of 396,786 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 13 IDs.\n",
      "100%     13 of     13 study items found in association\n",
      "100%     13 of     13 study items found in population(396786)\n",
      "Calculating 186 uncorrected p-values using fisher_scipy_stats\n",
      "     186 terms are associated with     12 of 396,786 population items\n",
      "     186 terms are associated with     12 of     13 study items\n",
      "  METHOD fdr_bh:\n",
      "     186 GO terms found significant (< 0.05=alpha) (186 enriched +   0 purified): statsmodels fdr_bh\n",
      "      12 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    169 of 396,786 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 13 IDs.\n",
      "100%     13 of     13 study items found in association\n",
      "100%     13 of     13 study items found in population(396786)\n",
      "Calculating 186 uncorrected p-values using fisher_scipy_stats\n",
      "     186 terms are associated with     12 of 396,786 population items\n",
      "     186 terms are associated with     12 of     13 study items\n",
      "  METHOD fdr_bh:\n",
      "     186 GO terms found significant (< 0.05=alpha) (186 enriched +   0 purified): statsmodels fdr_bh\n",
      "      12 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    169 of 396,786 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 13 IDs.\n",
      "100%     13 of     13 study items found in association\n",
      "100%     13 of     13 study items found in population(396786)\n",
      "Calculating 186 uncorrected p-values using fisher_scipy_stats\n",
      "     186 terms are associated with     12 of 396,786 population items\n",
      "     186 terms are associated with     12 of     13 study items\n",
      "  METHOD fdr_bh:\n",
      "     186 GO terms found significant (< 0.05=alpha) (186 enriched +   0 purified): statsmodels fdr_bh\n",
      "      12 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n"
     ]
    }
   ],
   "source": [
    "with open(\"./lasso_models/10xKfold_lasso_output/valine_nXcv.json\", 'r') as file:\n",
    "    valine_nXcv = json.load(file)\n",
    "#valine_nXcv_matched, valine_goids, valine_term_count, valine_goea = matchgff2(valine_nXcv['common_features'], goea=True)\n",
    "valine_nXcv_matched, val_goids_biop, val_term_count_biop, val_goea_biop = matchgff2(valine_nXcv['common_features'], namespace=['biological_process', 'molecular_function'], depth_threshold=2, goea=True)\n",
    "valine_nXcv_matched, val_goids_molf, val_term_count_molf, val_goea_molf = matchgff2(valine_nXcv['common_features'], namespace=['molecular_function'], depth_threshold=2, goea=True)\n",
    "valine_nXcv_matched, val_goids_cellc, val_term_count_cellc, val_goea_cellc = matchgff2(valine_nXcv['common_features'], namespace=['cellular_component'], depth_threshold=2, goea=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# <<< VALINE >>>\n",
      "# 10x3FoldCV mean score: 0.4303056564518039 \n",
      "# count common features: 13\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "2\tGO:0046777\tprotein autophosphorylation\tbiological_process\n",
      "2\tGO:0000166\tnucleotide binding\tmolecular_function\n",
      "2\tGO:0016567\tprotein ubiquitination\tbiological_process\n",
      "2\tGO:0051787\tmisfolded protein binding\tmolecular_function\n",
      "2\tGO:0051015\tactin filament binding\tmolecular_function\n",
      "1\tGO:0032491\tdetection of molecule of fungal origin\tbiological_process\n",
      "1\tGO:0008061\tchitin binding\tmolecular_function\n",
      "1\tGO:0090333\tregulation of stomatal closure\tbiological_process\n",
      "1\tGO:0048364\troot development\tbiological_process\n",
      "1\tGO:0002752\tcell surface pattern recognition receptor signaling pathway\tbiological_process\n",
      "1\tGO:0019722\tcalcium-mediated signaling\tbiological_process\n",
      "1\tGO:0071219\tcellular response to molecule of bacterial origin\tbiological_process\n",
      "1\tGO:0050832\tdefense response to fungus\tbiological_process\n",
      "1\tGO:0015026\tcoreceptor activity\tmolecular_function\n",
      "1\tGO:0042803\tprotein homodimerization activity\tmolecular_function\n",
      "1\tGO:0009862\tsystemic acquired resistance, salicylic acid mediated signaling pathway\tbiological_process\n",
      "1\tGO:2001080\tchitosan binding\tmolecular_function\n",
      "1\tGO:0004674\tprotein serine/threonine kinase activity\tmolecular_function\n",
      "1\tGO:1900425\tnegative regulation of defense response to bacterium\tbiological_process\n",
      "1\tGO:0009845\tseed germination\tbiological_process\n",
      "1\tGO:1900426\tpositive regulation of defense response to bacterium\tbiological_process\n",
      "1\tGO:0007639\thomeostasis of number of meristem cells\tbiological_process\n",
      "1\tGO:0004842\tubiquitin-protein transferase activity\tmolecular_function\n",
      "1\tGO:0009738\tabscisic acid-activated signaling pathway\tbiological_process\n",
      "1\tGO:0002229\tdefense response to oomycetes\tbiological_process\n",
      "1\tGO:0048653\tanther development\tbiological_process\n",
      "1\tGO:0140426\tpathogen-associated molecular pattern receptor signaling pathway\tbiological_process\n",
      "1\tGO:0019199\ttransmembrane receptor protein kinase activity\tmolecular_function\n",
      "1\tGO:0010726\tpositive regulation of hydrogen peroxide metabolic process\tbiological_process\n",
      "1\tGO:0032499\tdetection of peptidoglycan\tbiological_process\n",
      "1\tGO:0071323\tcellular response to chitin\tbiological_process\n",
      "1\tGO:0010942\tN/A\tN/A\n",
      "1\tGO:0006457\tprotein folding\tbiological_process\n",
      "1\tGO:0005515\tprotein binding\tmolecular_function\n",
      "1\tGO:0016210\tnaringenin-chalcone synthase activity\tmolecular_function\n",
      "1\tGO:0090439\ttetraketide alpha-pyrone synthase activity\tmolecular_function\n",
      "1\tGO:0009813\tflavonoid biosynthetic process\tbiological_process\n",
      "1\tGO:0030639\tpolyketide biosynthetic process\tbiological_process\n",
      "1\tGO:0070141\tresponse to UV-A\tbiological_process\n",
      "1\tGO:0080110\tsporopollenin biosynthetic process\tbiological_process\n",
      "1\tGO:0030176\tN/A\tN/A\n",
      "1\tGO:0004252\tserine-type endopeptidase activity\tmolecular_function\n",
      "1\tGO:1990381\tubiquitin-specific protease binding\tmolecular_function\n",
      "1\tGO:0007283\tspermatogenesis\tbiological_process\n",
      "1\tGO:0030154\tcell differentiation\tbiological_process\n",
      "1\tGO:0030433\tubiquitin-dependent ERAD pathway\tbiological_process\n",
      "1\tGO:0030968\tendoplasmic reticulum unfolded protein response\tbiological_process\n",
      "1\tGO:0033619\tmembrane protein proteolysis\tbiological_process\n",
      "1\tGO:0051247\tpositive regulation of protein metabolic process\tbiological_process\n",
      "1\tGO:0016887\tATP hydrolysis activity\tmolecular_function\n",
      "1\tGO:0032570\tresponse to progesterone\tbiological_process\n",
      "1\tGO:0061738\tlate endosomal microautophagy\tbiological_process\n",
      "1\tGO:0001786\tphosphatidylserine binding\tmolecular_function\n",
      "1\tGO:0003723\tRNA binding\tmolecular_function\n",
      "1\tGO:0045892\tnegative regulation of DNA-templated transcription\tbiological_process\n",
      "1\tGO:0001822\tkidney development\tbiological_process\n",
      "1\tGO:0051087\tprotein-folding chaperone binding\tmolecular_function\n",
      "1\tGO:0034620\tcellular response to unfolded protein\tbiological_process\n",
      "1\tGO:0031072\theat shock protein binding\tmolecular_function\n",
      "1\tGO:0044849\testrous cycle\tbiological_process\n",
      "1\tGO:0021549\tcerebellum development\tbiological_process\n",
      "1\tGO:0031625\tubiquitin protein ligase binding\tmolecular_function\n",
      "1\tGO:0043085\tpositive regulation of catalytic activity\tbiological_process\n",
      "1\tGO:0006606\tprotein import into nucleus\tbiological_process\n",
      "1\tGO:0010667\tnegative regulation of cardiac muscle cell apoptotic process\tbiological_process\n",
      "1\tGO:1904593\tprostaglandin binding\tmolecular_function\n",
      "1\tGO:1904592\tpositive regulation of protein refolding\tbiological_process\n",
      "1\tGO:0042277\tpeptide binding\tmolecular_function\n",
      "1\tGO:0009410\tresponse to xenobiotic stimulus\tbiological_process\n",
      "1\tGO:0061635\tregulation of protein complex stability\tbiological_process\n",
      "1\tGO:0016191\tsynaptic vesicle uncoating\tbiological_process\n",
      "1\tGO:0034605\tcellular response to heat\tbiological_process\n",
      "1\tGO:0048026\tpositive regulation of mRNA splicing, via spliceosome\tbiological_process\n",
      "1\tGO:0071276\tcellular response to cadmium ion\tbiological_process\n",
      "1\tGO:0061740\tprotein targeting to lysosome involved in chaperone-mediated autophagy\tbiological_process\n",
      "1\tGO:1904764\tchaperone-mediated autophagy translocation complex disassembly\tbiological_process\n",
      "1\tGO:0043254\tregulation of protein-containing complex assembly\tbiological_process\n",
      "1\tGO:0043531\tADP binding\tmolecular_function\n",
      "1\tGO:0044743\tprotein transmembrane import into intracellular organelle\tbiological_process\n",
      "1\tGO:0071383\tcellular response to steroid hormone stimulus\tbiological_process\n",
      "1\tGO:0007519\tskeletal muscle tissue development\tbiological_process\n",
      "1\tGO:1990832\tslow axonal transport\tbiological_process\n",
      "1\tGO:1990833\tclathrin-uncoating ATPase activity\tmolecular_function\n",
      "1\tGO:1990834\tresponse to odorant\tbiological_process\n",
      "1\tGO:0031686\tA1 adenosine receptor binding\tmolecular_function\n",
      "1\tGO:0055131\tC3HC4-type RING finger domain binding\tmolecular_function\n",
      "1\tGO:1903206\tN/A\tN/A\n",
      "1\tGO:0030900\tforebrain development\tbiological_process\n",
      "1\tGO:1902904\tnegative regulation of supramolecular fiber organization\tbiological_process\n",
      "1\tGO:0007608\tsensory perception of smell\tbiological_process\n",
      "1\tGO:0051726\tregulation of cell cycle\tbiological_process\n",
      "1\tGO:0097214\tpositive regulation of lysosomal membrane permeability\tbiological_process\n",
      "1\tGO:0046034\tATP metabolic process\tbiological_process\n",
      "1\tGO:0032355\tresponse to estradiol\tbiological_process\n",
      "1\tGO:0001916\tpositive regulation of T cell mediated cytotoxicity\tbiological_process\n",
      "1\tGO:0014823\tresponse to activity\tbiological_process\n",
      "1\tGO:0140545\tATP-dependent protein disaggregase activity\tmolecular_function\n",
      "1\tGO:0050766\tpositive regulation of phagocytosis\tbiological_process\n",
      "1\tGO:0042594\tresponse to starvation\tbiological_process\n",
      "1\tGO:0002020\tprotease binding\tmolecular_function\n",
      "1\tGO:0045862\tpositive regulation of proteolysis\tbiological_process\n",
      "1\tGO:0009615\tresponse to virus\tbiological_process\n",
      "1\tGO:0098880\tmaintenance of postsynaptic specialization structure\tbiological_process\n",
      "1\tGO:0099175\tregulation of postsynapse organization\tbiological_process\n",
      "1\tGO:0005524\tATP binding\tmolecular_function\n",
      "1\tGO:0045471\tresponse to ethanol\tbiological_process\n",
      "1\tGO:0000082\tG1/S transition of mitotic cell cycle\tbiological_process\n",
      "1\tGO:0044829\tpositive regulation by host of viral genome replication\tbiological_process\n",
      "1\tGO:0051085\tchaperone cofactor-dependent protein refolding\tbiological_process\n",
      "1\tGO:0051082\tunfolded protein binding\tmolecular_function\n",
      "1\tGO:0007568\tN/A\tN/A\n",
      "1\tGO:0009617\tresponse to bacterium\tbiological_process\n",
      "1\tGO:0031647\tregulation of protein stability\tbiological_process\n",
      "1\tGO:0030674\tprotein-macromolecule adaptor activity\tmolecular_function\n",
      "1\tGO:0010045\tresponse to nickel cation\tbiological_process\n",
      "1\tGO:0016413\tO-acetyltransferase activity\tmolecular_function\n",
      "1\tGO:0008312\t7S RNA binding\tmolecular_function\n",
      "1\tGO:0043022\tribosome binding\tmolecular_function\n",
      "1\tGO:0006617\tSRP-dependent cotranslational protein targeting to membrane, signal sequence recognition\tbiological_process\n",
      "1\tGO:0004222\tmetalloendopeptidase activity\tmolecular_function\n",
      "1\tGO:0046872\tmetal ion binding\tmolecular_function\n",
      "1\tGO:0006508\tproteolysis\tbiological_process\n",
      "1\tGO:0006518\tpeptide metabolic process\tbiological_process\n",
      "1\tGO:0050794\tregulation of cellular process\tbiological_process\n",
      "1\tGO:0015095\tmagnesium ion transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0015693\tmagnesium ion transport\tbiological_process\n",
      "1\tGO:0003729\tmRNA binding\tmolecular_function\n",
      "1\tGO:0000146\tmicrofilament motor activity\tmolecular_function\n",
      "1\tGO:0005547\tphosphatidylinositol-3,4,5-trisphosphate binding\tmolecular_function\n",
      "1\tGO:0032027\tmyosin light chain binding\tmolecular_function\n",
      "1\tGO:0006897\tendocytosis\tbiological_process\n",
      "1\tGO:0030036\tactin cytoskeleton organization\tbiological_process\n",
      "1\tGO:0030048\tactin filament-based movement\tbiological_process\n",
      "1\tGO:0043327\tchemotaxis to cAMP\tbiological_process\n",
      "1\tGO:0046907\tintracellular transport\tbiological_process\n",
      "1\tGO:0065008\tregulation of biological quality\tbiological_process\n",
      "1\tGO:0120036\tplasma membrane bounded cell projection organization\tbiological_process\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "2\tGO:0000166\tnucleotide binding\tmolecular_function\n",
      "2\tGO:0051787\tmisfolded protein binding\tmolecular_function\n",
      "2\tGO:0051015\tactin filament binding\tmolecular_function\n",
      "1\tGO:0008061\tchitin binding\tmolecular_function\n",
      "1\tGO:0015026\tcoreceptor activity\tmolecular_function\n",
      "1\tGO:0042803\tprotein homodimerization activity\tmolecular_function\n",
      "1\tGO:2001080\tchitosan binding\tmolecular_function\n",
      "1\tGO:0004674\tprotein serine/threonine kinase activity\tmolecular_function\n",
      "1\tGO:0004842\tubiquitin-protein transferase activity\tmolecular_function\n",
      "1\tGO:0019199\ttransmembrane receptor protein kinase activity\tmolecular_function\n",
      "1\tGO:0010942\tN/A\tN/A\n",
      "1\tGO:0005515\tprotein binding\tmolecular_function\n",
      "1\tGO:0016210\tnaringenin-chalcone synthase activity\tmolecular_function\n",
      "1\tGO:0090439\ttetraketide alpha-pyrone synthase activity\tmolecular_function\n",
      "1\tGO:0030176\tN/A\tN/A\n",
      "1\tGO:0004252\tserine-type endopeptidase activity\tmolecular_function\n",
      "1\tGO:1990381\tubiquitin-specific protease binding\tmolecular_function\n",
      "1\tGO:0016887\tATP hydrolysis activity\tmolecular_function\n",
      "1\tGO:0001786\tphosphatidylserine binding\tmolecular_function\n",
      "1\tGO:0003723\tRNA binding\tmolecular_function\n",
      "1\tGO:0051087\tprotein-folding chaperone binding\tmolecular_function\n",
      "1\tGO:0031072\theat shock protein binding\tmolecular_function\n",
      "1\tGO:0031625\tubiquitin protein ligase binding\tmolecular_function\n",
      "1\tGO:1904593\tprostaglandin binding\tmolecular_function\n",
      "1\tGO:0042277\tpeptide binding\tmolecular_function\n",
      "1\tGO:0043531\tADP binding\tmolecular_function\n",
      "1\tGO:1990833\tclathrin-uncoating ATPase activity\tmolecular_function\n",
      "1\tGO:0031686\tA1 adenosine receptor binding\tmolecular_function\n",
      "1\tGO:0055131\tC3HC4-type RING finger domain binding\tmolecular_function\n",
      "1\tGO:1903206\tN/A\tN/A\n",
      "1\tGO:0140545\tATP-dependent protein disaggregase activity\tmolecular_function\n",
      "1\tGO:0002020\tprotease binding\tmolecular_function\n",
      "1\tGO:0005524\tATP binding\tmolecular_function\n",
      "1\tGO:0051082\tunfolded protein binding\tmolecular_function\n",
      "1\tGO:0007568\tN/A\tN/A\n",
      "1\tGO:0030674\tprotein-macromolecule adaptor activity\tmolecular_function\n",
      "1\tGO:0016413\tO-acetyltransferase activity\tmolecular_function\n",
      "1\tGO:0008312\t7S RNA binding\tmolecular_function\n",
      "1\tGO:0043022\tribosome binding\tmolecular_function\n",
      "1\tGO:0004222\tmetalloendopeptidase activity\tmolecular_function\n",
      "1\tGO:0046872\tmetal ion binding\tmolecular_function\n",
      "1\tGO:0015095\tmagnesium ion transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0003729\tmRNA binding\tmolecular_function\n",
      "1\tGO:0000146\tmicrofilament motor activity\tmolecular_function\n",
      "1\tGO:0005547\tphosphatidylinositol-3,4,5-trisphosphate binding\tmolecular_function\n",
      "1\tGO:0032027\tmyosin light chain binding\tmolecular_function\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "2\tGO:0005886\tplasma membrane\tcellular_component\n",
      "2\tGO:0005783\tendoplasmic reticulum\tcellular_component\n",
      "2\tGO:0005730\tnucleolus\tcellular_component\n",
      "2\tGO:0016020\tmembrane\tcellular_component\n",
      "2\tGO:0005829\tcytosol\tcellular_component\n",
      "1\tGO:0010942\tN/A\tN/A\n",
      "1\tGO:0016021\tmembrane\tcellular_component\n",
      "1\tGO:0000839\tHrd1p ubiquitin ligase ERAD-L complex\tcellular_component\n",
      "1\tGO:0030176\tN/A\tN/A\n",
      "1\tGO:0014069\tpostsynaptic density\tcellular_component\n",
      "1\tGO:0000974\tPrp19 complex\tcellular_component\n",
      "1\tGO:0005874\tmicrotubule\tcellular_component\n",
      "1\tGO:1990124\tmessenger ribonucleoprotein complex\tcellular_component\n",
      "1\tGO:0043195\tterminal bouton\tcellular_component\n",
      "1\tGO:0043198\tdendritic shaft\tcellular_component\n",
      "1\tGO:0043197\tdendritic spine\tcellular_component\n",
      "1\tGO:0098978\tglutamatergic synapse\tcellular_component\n",
      "1\tGO:0061202\tclathrin-sculpted gamma-aminobutyric acid transport vesicle membrane\tcellular_component\n",
      "1\tGO:0005882\tintermediate filament\tcellular_component\n",
      "1\tGO:0005765\tlysosomal membrane\tcellular_component\n",
      "1\tGO:0031906\tlate endosome lumen\tcellular_component\n",
      "1\tGO:0016363\tnuclear matrix\tcellular_component\n",
      "1\tGO:0070062\textracellular exosome\tcellular_component\n",
      "1\tGO:0099523\tpresynaptic cytosol\tcellular_component\n",
      "1\tGO:0099524\tpostsynaptic cytosol\tcellular_component\n",
      "1\tGO:0098690\tglycinergic synapse\tcellular_component\n",
      "1\tGO:0098684\tphotoreceptor ribbon synapse\tcellular_component\n",
      "1\tGO:0031969\tchloroplast membrane\tcellular_component\n",
      "1\tGO:1990836\tlysosomal matrix\tcellular_component\n",
      "1\tGO:1990718\taxonemal central pair projection\tcellular_component\n",
      "1\tGO:0008021\tsynaptic vesicle\tcellular_component\n",
      "1\tGO:1903206\tN/A\tN/A\n",
      "1\tGO:0030864\tcortical actin cytoskeleton\tcellular_component\n",
      "1\tGO:0005681\tspliceosomal complex\tcellular_component\n",
      "1\tGO:0042470\tmelanosome\tcellular_component\n",
      "1\tGO:0099634\tpostsynaptic specialization membrane\tcellular_component\n",
      "1\tGO:0001917\tphotoreceptor inner segment\tcellular_component\n",
      "1\tGO:0043204\tperikaryon\tcellular_component\n",
      "1\tGO:0005776\tautophagosome\tcellular_component\n",
      "1\tGO:0007568\tN/A\tN/A\n",
      "1\tGO:0009986\tcell surface\tcellular_component\n",
      "1\tGO:0048471\tperinuclear region of cytoplasm\tcellular_component\n",
      "1\tGO:0045121\tmembrane raft\tcellular_component\n",
      "1\tGO:0101031\tprotein folding chaperone complex\tcellular_component\n",
      "1\tGO:0005794\tGolgi apparatus\tcellular_component\n",
      "1\tGO:0005786\tsignal recognition particle, endoplasmic reticulum targeting\tcellular_component\n",
      "1\tGO:0016604\tnuclear body\tcellular_component\n",
      "1\tGO:0005739\tmitochondrion\tcellular_component\n",
      "1\tGO:0009507\tchloroplast\tcellular_component\n",
      "1\tGO:0005737\tcytoplasm\tcellular_component\n",
      "1\tGO:0005911\tcell-cell junction\tcellular_component\n",
      "1\tGO:0045160\tmyosin I complex\tcellular_component\n",
      "1\tGO:0045335\tphagocytic vesicle\tcellular_component\n",
      "1\tGO:0061851\tleading edge of lamellipodium\tcellular_component\n",
      "1\tGO:0070687\tmacropinocytic cup cytoskeleton\tcellular_component\n",
      "1\tGO:0098858\tactin-based cell projection\tcellular_component\n"
     ]
    }
   ],
   "source": [
    "#plot_go_terms(valine_term_count, top_n=10)\n",
    "print(\"# <<< VALINE >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(valine_nXcv['mean_scores'])} \\n# count common features: {len(valine_nXcv['common_features'])}\")\n",
    "tabulate(val_term_count_biop)\n",
    "tabulate(val_term_count_molf)\n",
    "tabulate(val_term_count_cellc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    961 of 946,182 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 31 IDs.\n",
      "100%     31 of     31 study items found in association\n",
      "100%     31 of     31 study items found in population(946182)\n",
      "Calculating 223 uncorrected p-values using fisher_scipy_stats\n",
      "     223 terms are associated with     25 of 946,182 population items\n",
      "     223 terms are associated with     25 of     31 study items\n",
      "  METHOD fdr_bh:\n",
      "     223 GO terms found significant (< 0.05=alpha) (223 enriched +   0 purified): statsmodels fdr_bh\n",
      "      25 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    961 of 946,182 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 31 IDs.\n",
      "100%     31 of     31 study items found in association\n",
      "100%     31 of     31 study items found in population(946182)\n",
      "Calculating 223 uncorrected p-values using fisher_scipy_stats\n",
      "     223 terms are associated with     25 of 946,182 population items\n",
      "     223 terms are associated with     25 of     31 study items\n",
      "  METHOD fdr_bh:\n",
      "     223 GO terms found significant (< 0.05=alpha) (223 enriched +   0 purified): statsmodels fdr_bh\n",
      "      25 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    961 of 946,182 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 31 IDs.\n",
      "100%     31 of     31 study items found in association\n",
      "100%     31 of     31 study items found in population(946182)\n",
      "Calculating 223 uncorrected p-values using fisher_scipy_stats\n",
      "     223 terms are associated with     25 of 946,182 population items\n",
      "     223 terms are associated with     25 of     31 study items\n",
      "  METHOD fdr_bh:\n",
      "     223 GO terms found significant (< 0.05=alpha) (223 enriched +   0 purified): statsmodels fdr_bh\n",
      "      25 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n"
     ]
    }
   ],
   "source": [
    "with open(\"./lasso_models/10xKfold_lasso_output/rutin_nXcv.json\", 'r') as file:\n",
    "    rutin_nXcv = json.load(file)\n",
    "#rutin_nXcv_matched, rutin_goids, rutin_term_count, rutin_goea = matchgff2(rutin_nXcv['common_features'], goea=True)\n",
    "rutin_nXcv_matched, rut_goids_biop, rut_term_count_biop, rut_goea_biop = matchgff2(rutin_nXcv['common_features'], namespace=['biological_process', 'molecular_function'], depth_threshold=2, goea=True)\n",
    "rutin_nXcv_matched, rut_goids_molf, rut_term_count_molf, rut_goea_molf = matchgff2(rutin_nXcv['common_features'], namespace=['molecular_function'], depth_threshold=2, goea=True)\n",
    "rutin_nXcv_matched, rut_goids_cellc, rut_term_count_cellc, rut_goea_cellc = matchgff2(rutin_nXcv['common_features'], namespace=['cellular_component'], depth_threshold=2, goea=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# <<< RUTIN >>>\n",
      "# 10x3FoldCV mean score: 0.7485866213502098 \n",
      "# count common features: 31\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "6\tGO:0005515\tprotein binding\tmolecular_function\n",
      "4\tGO:0042803\tprotein homodimerization activity\tmolecular_function\n",
      "4\tGO:0009733\tresponse to auxin\tbiological_process\n",
      "3\tGO:0045892\tnegative regulation of DNA-templated transcription\tbiological_process\n",
      "3\tGO:0009414\tresponse to water deprivation\tbiological_process\n",
      "3\tGO:0003700\tDNA-binding transcription factor activity\tmolecular_function\n",
      "3\tGO:0045893\tpositive regulation of DNA-templated transcription\tbiological_process\n",
      "3\tGO:0009862\tsystemic acquired resistance, salicylic acid mediated signaling pathway\tbiological_process\n",
      "2\tGO:0003729\tmRNA binding\tmolecular_function\n",
      "2\tGO:0048481\tplant ovule development\tbiological_process\n",
      "2\tGO:0006096\tglycolytic process\tbiological_process\n",
      "2\tGO:0009734\tauxin-activated signaling pathway\tbiological_process\n",
      "2\tGO:0042542\tresponse to hydrogen peroxide\tbiological_process\n",
      "2\tGO:0009723\tresponse to ethylene\tbiological_process\n",
      "2\tGO:0000987\tcis-regulatory region sequence-specific DNA binding\tmolecular_function\n",
      "2\tGO:0009753\tresponse to jasmonic acid\tbiological_process\n",
      "2\tGO:0009410\tresponse to xenobiotic stimulus\tbiological_process\n",
      "2\tGO:1905623\tpositive regulation of leaf development\tbiological_process\n",
      "2\tGO:0042742\tdefense response to bacterium\tbiological_process\n",
      "2\tGO:0002237\tresponse to molecule of bacterial origin\tbiological_process\n",
      "2\tGO:0009909\tregulation of flower development\tbiological_process\n",
      "2\tGO:0048653\tanther development\tbiological_process\n",
      "2\tGO:0071588\thydrogen peroxide mediated signaling pathway\tbiological_process\n",
      "1\tGO:0010496\tintercellular transport\tbiological_process\n",
      "1\tGO:0046907\tintracellular transport\tbiological_process\n",
      "1\tGO:0004842\tubiquitin-protein transferase activity\tmolecular_function\n",
      "1\tGO:0000956\tnuclear-transcribed mRNA catabolic process\tbiological_process\n",
      "1\tGO:0010556\tregulation of macromolecule biosynthetic process\tbiological_process\n",
      "1\tGO:0016567\tprotein ubiquitination\tbiological_process\n",
      "1\tGO:0031326\tregulation of cellular biosynthetic process\tbiological_process\n",
      "1\tGO:0051254\tpositive regulation of RNA metabolic process\tbiological_process\n",
      "1\tGO:0003735\tstructural constituent of ribosome\tmolecular_function\n",
      "1\tGO:0019843\trRNA binding\tmolecular_function\n",
      "1\tGO:0030599\tpectinesterase activity\tmolecular_function\n",
      "1\tGO:0045330\taspartyl esterase activity\tmolecular_function\n",
      "1\tGO:0000462\tmaturation of SSU-rRNA from tricistronic rRNA transcript (SSU-rRNA, 5.8S rRNA, LSU-rRNA)\tbiological_process\n",
      "1\tGO:0002181\tcytoplasmic translation\tbiological_process\n",
      "1\tGO:0042545\tcell wall modification\tbiological_process\n",
      "1\tGO:0045490\tpectin catabolic process\tbiological_process\n",
      "1\tGO:0016757\tglycosyltransferase activity\tmolecular_function\n",
      "1\tGO:0008460\tdTDP-glucose 4,6-dehydratase activity\tmolecular_function\n",
      "1\tGO:0048040\tUDP-glucuronate decarboxylase activity\tmolecular_function\n",
      "1\tGO:0070403\tNAD+ binding\tmolecular_function\n",
      "1\tGO:0001503\tossification\tbiological_process\n",
      "1\tGO:0015012\theparan sulfate proteoglycan biosynthetic process\tbiological_process\n",
      "1\tGO:0019305\tdTDP-rhamnose biosynthetic process\tbiological_process\n",
      "1\tGO:0030198\textracellular matrix organization\tbiological_process\n",
      "1\tGO:0030206\tchondroitin sulfate biosynthetic process\tbiological_process\n",
      "1\tGO:0033320\tUDP-D-xylose biosynthetic process\tbiological_process\n",
      "1\tGO:0042732\tD-xylose metabolic process\tbiological_process\n",
      "1\tGO:0045492\txylan biosynthetic process\tbiological_process\n",
      "1\tGO:0051216\tcartilage development\tbiological_process\n",
      "1\tGO:0070979\tprotein K11-linked ubiquitination\tbiological_process\n",
      "1\tGO:0009793\tembryo development ending in seed dormancy\tbiological_process\n",
      "1\tGO:0031145\tanaphase-promoting complex-dependent catabolic process\tbiological_process\n",
      "1\tGO:0045842\tpositive regulation of mitotic metaphase/anaphase transition\tbiological_process\n",
      "1\tGO:0016709\toxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, NAD(P)H as one donor, and incorporation of one atom of oxygen\tmolecular_function\n",
      "1\tGO:0046872\tmetal ion binding\tmolecular_function\n",
      "1\tGO:0009813\tflavonoid biosynthetic process\tbiological_process\n",
      "1\tGO:0004742\tdihydrolipoyllysine-residue acetyltransferase activity\tmolecular_function\n",
      "1\tGO:0006086\tacetyl-CoA biosynthetic process from pyruvate\tbiological_process\n",
      "1\tGO:0030431\tsleep\tbiological_process\n",
      "1\tGO:0051240\tpositive regulation of multicellular organismal process\tbiological_process\n",
      "1\tGO:0004332\tfructose-bisphosphate aldolase activity\tmolecular_function\n",
      "1\tGO:0043621\tprotein self-association\tmolecular_function\n",
      "1\tGO:0005986\tsucrose biosynthetic process\tbiological_process\n",
      "1\tGO:0006000\tfructose metabolic process\tbiological_process\n",
      "1\tGO:0006094\tgluconeogenesis\tbiological_process\n",
      "1\tGO:0030388\tfructose 1,6-bisphosphate metabolic process\tbiological_process\n",
      "1\tGO:0031930\tmitochondria-nucleus signaling pathway\tbiological_process\n",
      "1\tGO:0034059\tresponse to anoxia\tbiological_process\n",
      "1\tGO:0048364\troot development\tbiological_process\n",
      "1\tGO:0008964\tphosphoenolpyruvate carboxylase activity\tmolecular_function\n",
      "1\tGO:0006099\ttricarboxylic acid cycle\tbiological_process\n",
      "1\tGO:0009760\tC4 photosynthesis\tbiological_process\n",
      "1\tGO:0009915\tphloem sucrose loading\tbiological_process\n",
      "1\tGO:0048366\tleaf development\tbiological_process\n",
      "1\tGO:0051262\tprotein tetramerization\tbiological_process\n",
      "1\tGO:0004144\tdiacylglycerol O-acyltransferase activity\tmolecular_function\n",
      "1\tGO:0006071\tglycerol metabolic process\tbiological_process\n",
      "1\tGO:0019216\tregulation of lipid metabolic process\tbiological_process\n",
      "1\tGO:0019432\ttriglyceride biosynthetic process\tbiological_process\n",
      "1\tGO:0019915\tlipid storage\tbiological_process\n",
      "1\tGO:0031323\tregulation of cellular metabolic process\tbiological_process\n",
      "1\tGO:0034389\tlipid droplet organization\tbiological_process\n",
      "1\tGO:0055088\tlipid homeostasis\tbiological_process\n",
      "1\tGO:0062012\tregulation of small molecule metabolic process\tbiological_process\n",
      "1\tGO:1900140\tregulation of seedling development\tbiological_process\n",
      "1\tGO:1902074\tresponse to salt\tbiological_process\n",
      "1\tGO:0009651\tresponse to salt stress\tbiological_process\n",
      "1\tGO:0090058\tmetaxylem development\tbiological_process\n",
      "1\tGO:0043068\tpositive regulation of programmed cell death\tbiological_process\n",
      "1\tGO:0000976\ttranscription cis-regulatory region binding\tmolecular_function\n",
      "1\tGO:1900057\tpositive regulation of leaf senescence\tbiological_process\n",
      "1\tGO:0051091\tpositive regulation of DNA-binding transcription factor activity\tbiological_process\n",
      "1\tGO:1901348\tpositive regulation of secondary cell wall biogenesis\tbiological_process\n",
      "1\tGO:0009735\tresponse to cytokinin\tbiological_process\n",
      "1\tGO:0048759\txylem vessel member cell differentiation\tbiological_process\n",
      "1\tGO:1903648\tpositive regulation of chlorophyll catabolic process\tbiological_process\n",
      "1\tGO:1990110\tcallus formation\tbiological_process\n",
      "1\tGO:0045491\txylan metabolic process\tbiological_process\n",
      "1\tGO:0051782\tnegative regulation of cell division\tbiological_process\n",
      "1\tGO:2000014\tregulation of endosperm development\tbiological_process\n",
      "1\tGO:1904250\tpositive regulation of age-related resistance\tbiological_process\n",
      "1\tGO:0010199\torgan boundary specification between lateral organs and the meristem\tbiological_process\n",
      "1\tGO:0048829\troot cap development\tbiological_process\n",
      "1\tGO:0090400\tstress-induced premature senescence\tbiological_process\n",
      "1\tGO:0080187\tfloral organ senescence\tbiological_process\n",
      "1\tGO:0010072\tprimary shoot apical meristem specification\tbiological_process\n",
      "1\tGO:0009620\tresponse to fungus\tbiological_process\n",
      "1\tGO:0009741\tresponse to brassinosteroid\tbiological_process\n",
      "1\tGO:0048527\tlateral root development\tbiological_process\n",
      "1\tGO:0009737\tresponse to abscisic acid\tbiological_process\n",
      "1\tGO:0010628\tpositive regulation of gene expression\tbiological_process\n",
      "1\tGO:0090709\tregulation of timing of plant organ formation\tbiological_process\n",
      "1\tGO:0010029\tregulation of seed germination\tbiological_process\n",
      "1\tGO:0010981\tregulation of cell wall macromolecule metabolic process\tbiological_process\n",
      "1\tGO:0010223\tsecondary shoot formation\tbiological_process\n",
      "1\tGO:0004707\tMAP kinase activity\tmolecular_function\n",
      "1\tGO:0004712\tprotein serine/threonine/tyrosine kinase activity\tmolecular_function\n",
      "1\tGO:0004713\tprotein tyrosine kinase activity\tmolecular_function\n",
      "1\tGO:0005524\tATP binding\tmolecular_function\n",
      "1\tGO:0019902\tphosphatase binding\tmolecular_function\n",
      "1\tGO:0106310\tprotein serine kinase activity\tmolecular_function\n",
      "1\tGO:0000165\tMAPK cascade\tbiological_process\n",
      "1\tGO:0002221\tpattern recognition receptor signaling pathway\tbiological_process\n",
      "1\tGO:0007049\tcell cycle\tbiological_process\n",
      "1\tGO:0009555\tpollen development\tbiological_process\n",
      "1\tGO:0009611\tresponse to wounding\tbiological_process\n",
      "1\tGO:0009626\tplant-type hypersensitive response\tbiological_process\n",
      "1\tGO:0009738\tabscisic acid-activated signaling pathway\tbiological_process\n",
      "1\tGO:0009864\tinduced systemic resistance, jasmonic acid mediated signaling pathway\tbiological_process\n",
      "1\tGO:0009875\tpollen-pistil interaction\tbiological_process\n",
      "1\tGO:0010051\txylem and phloem pattern formation\tbiological_process\n",
      "1\tGO:0010082\tregulation of root meristem growth\tbiological_process\n",
      "1\tGO:0010120\tcamalexin biosynthetic process\tbiological_process\n",
      "1\tGO:0010150\tleaf senescence\tbiological_process\n",
      "1\tGO:0010183\tpollen tube guidance\tbiological_process\n",
      "1\tGO:0010200\tresponse to chitin\tbiological_process\n",
      "1\tGO:0010224\tresponse to UV-B\tbiological_process\n",
      "1\tGO:0010229\tinflorescence development\tbiological_process\n",
      "1\tGO:0010311\tlateral root formation\tbiological_process\n",
      "1\tGO:0010468\tregulation of gene expression\tbiological_process\n",
      "1\tGO:0018108\tpeptidyl-tyrosine phosphorylation\tbiological_process\n",
      "1\tGO:0042539\thypotonic salinity response\tbiological_process\n",
      "1\tGO:0043622\tcortical microtubule organization\tbiological_process\n",
      "1\tGO:0046777\tprotein autophosphorylation\tbiological_process\n",
      "1\tGO:0050826\tresponse to freezing\tbiological_process\n",
      "1\tGO:0050832\tdefense response to fungus\tbiological_process\n",
      "1\tGO:0051301\tcell division\tbiological_process\n",
      "1\tGO:0051510\tregulation of unidimensional cell growth\tbiological_process\n",
      "1\tGO:0060918\tauxin transport\tbiological_process\n",
      "1\tGO:0071244\tcellular response to carbon dioxide\tbiological_process\n",
      "1\tGO:0080026\tresponse to indolebutyric acid\tbiological_process\n",
      "1\tGO:0080136\tpriming of cellular response to stress\tbiological_process\n",
      "1\tGO:0090333\tregulation of stomatal closure\tbiological_process\n",
      "1\tGO:1901002\tpositive regulation of response to salt stress\tbiological_process\n",
      "1\tGO:1902065\tresponse to L-glutamate\tbiological_process\n",
      "1\tGO:0004324\tferredoxin-NADP+ reductase activity\tmolecular_function\n",
      "1\tGO:0015979\tphotosynthesis\tbiological_process\n",
      "1\tGO:0001046\tcore promoter sequence-specific DNA binding\tmolecular_function\n",
      "1\tGO:0006355\tregulation of DNA-templated transcription\tbiological_process\n",
      "1\tGO:0010118\tstomatal movement\tbiological_process\n",
      "1\tGO:0010167\tresponse to nitrate\tbiological_process\n",
      "1\tGO:0042128\tnitrate assimilation\tbiological_process\n",
      "1\tGO:0003989\tacetyl-CoA carboxylase activity\tmolecular_function\n",
      "1\tGO:0006633\tfatty acid biosynthetic process\tbiological_process\n",
      "1\tGO:0004851\turoporphyrin-III C-methyltransferase activity\tmolecular_function\n",
      "1\tGO:0009416\tresponse to light stimulus\tbiological_process\n",
      "1\tGO:0019354\tsiroheme biosynthetic process\tbiological_process\n",
      "1\tGO:0090352\tregulation of nitrate assimilation\tbiological_process\n",
      "1\tGO:1900058\tregulation of sulfate assimilation\tbiological_process\n",
      "1\tGO:1902326\tpositive regulation of chlorophyll biosynthetic process\tbiological_process\n",
      "1\tGO:0004252\tserine-type endopeptidase activity\tmolecular_function\n",
      "1\tGO:0006508\tproteolysis\tbiological_process\n",
      "1\tGO:0009608\tresponse to symbiont\tbiological_process\n",
      "1\tGO:0009682\tinduced systemic resistance\tbiological_process\n",
      "1\tGO:0009860\tpollen tube growth\tbiological_process\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "6\tGO:0005515\tprotein binding\tmolecular_function\n",
      "4\tGO:0042803\tprotein homodimerization activity\tmolecular_function\n",
      "3\tGO:0003700\tDNA-binding transcription factor activity\tmolecular_function\n",
      "2\tGO:0003729\tmRNA binding\tmolecular_function\n",
      "2\tGO:0000987\tcis-regulatory region sequence-specific DNA binding\tmolecular_function\n",
      "1\tGO:0004842\tubiquitin-protein transferase activity\tmolecular_function\n",
      "1\tGO:0003735\tstructural constituent of ribosome\tmolecular_function\n",
      "1\tGO:0019843\trRNA binding\tmolecular_function\n",
      "1\tGO:0030599\tpectinesterase activity\tmolecular_function\n",
      "1\tGO:0045330\taspartyl esterase activity\tmolecular_function\n",
      "1\tGO:0016757\tglycosyltransferase activity\tmolecular_function\n",
      "1\tGO:0008460\tdTDP-glucose 4,6-dehydratase activity\tmolecular_function\n",
      "1\tGO:0048040\tUDP-glucuronate decarboxylase activity\tmolecular_function\n",
      "1\tGO:0070403\tNAD+ binding\tmolecular_function\n",
      "1\tGO:0016709\toxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, NAD(P)H as one donor, and incorporation of one atom of oxygen\tmolecular_function\n",
      "1\tGO:0046872\tmetal ion binding\tmolecular_function\n",
      "1\tGO:0004742\tdihydrolipoyllysine-residue acetyltransferase activity\tmolecular_function\n",
      "1\tGO:0004332\tfructose-bisphosphate aldolase activity\tmolecular_function\n",
      "1\tGO:0043621\tprotein self-association\tmolecular_function\n",
      "1\tGO:0008964\tphosphoenolpyruvate carboxylase activity\tmolecular_function\n",
      "1\tGO:0004144\tdiacylglycerol O-acyltransferase activity\tmolecular_function\n",
      "1\tGO:0000976\ttranscription cis-regulatory region binding\tmolecular_function\n",
      "1\tGO:0004707\tMAP kinase activity\tmolecular_function\n",
      "1\tGO:0004712\tprotein serine/threonine/tyrosine kinase activity\tmolecular_function\n",
      "1\tGO:0004713\tprotein tyrosine kinase activity\tmolecular_function\n",
      "1\tGO:0005524\tATP binding\tmolecular_function\n",
      "1\tGO:0019902\tphosphatase binding\tmolecular_function\n",
      "1\tGO:0106310\tprotein serine kinase activity\tmolecular_function\n",
      "1\tGO:0004324\tferredoxin-NADP+ reductase activity\tmolecular_function\n",
      "1\tGO:0001046\tcore promoter sequence-specific DNA binding\tmolecular_function\n",
      "1\tGO:0003989\tacetyl-CoA carboxylase activity\tmolecular_function\n",
      "1\tGO:0004851\turoporphyrin-III C-methyltransferase activity\tmolecular_function\n",
      "1\tGO:0004252\tserine-type endopeptidase activity\tmolecular_function\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "4\tGO:0005634\tnucleus\tcellular_component\n",
      "3\tGO:0005783\tendoplasmic reticulum\tcellular_component\n",
      "3\tGO:0016021\tmembrane\tcellular_component\n",
      "3\tGO:0016020\tmembrane\tcellular_component\n",
      "3\tGO:0005737\tcytoplasm\tcellular_component\n",
      "3\tGO:0009507\tchloroplast\tcellular_component\n",
      "2\tGO:0005576\textracellular region\tcellular_component\n",
      "2\tGO:0005829\tcytosol\tcellular_component\n",
      "2\tGO:0005886\tplasma membrane\tcellular_component\n",
      "2\tGO:0090575\tRNA polymerase II transcription regulator complex\tcellular_component\n",
      "2\tGO:0090571\tRNA polymerase II transcription repressor complex\tcellular_component\n",
      "1\tGO:0009506\tplasmodesma\tcellular_component\n",
      "1\tGO:0005622\tintracellular anatomical structure\tcellular_component\n",
      "1\tGO:0030014\tCCR4-NOT complex\tcellular_component\n",
      "1\tGO:0000228\tnuclear chromosome\tcellular_component\n",
      "1\tGO:0005730\tnucleolus\tcellular_component\n",
      "1\tGO:0022627\tcytosolic small ribosomal subunit\tcellular_component\n",
      "1\tGO:0032580\tGolgi cisterna membrane\tcellular_component\n",
      "1\tGO:1902494\tcatalytic complex\tcellular_component\n",
      "1\tGO:0005680\tanaphase-promoting complex\tcellular_component\n",
      "1\tGO:0034399\tnuclear periphery\tcellular_component\n",
      "1\tGO:0005759\tmitochondrial matrix\tcellular_component\n",
      "1\tGO:0009570\tchloroplast stroma\tcellular_component\n",
      "1\tGO:0045254\tpyruvate dehydrogenase complex\tcellular_component\n",
      "1\tGO:0098798\tmitochondrial protein-containing complex\tcellular_component\n",
      "1\tGO:0120025\tplasma membrane bounded cell projection\tcellular_component\n",
      "1\tGO:0005739\tmitochondrion\tcellular_component\n",
      "1\tGO:0010287\tplastoglobule\tcellular_component\n",
      "1\tGO:0005794\tGolgi apparatus\tcellular_component\n",
      "1\tGO:0005789\tendoplasmic reticulum membrane\tcellular_component\n",
      "1\tGO:0005811\tlipid droplet\tcellular_component\n",
      "1\tGO:0097038\tperinuclear endoplasmic reticulum\tcellular_component\n",
      "1\tGO:0009531\tsecondary cell wall\tcellular_component\n",
      "1\tGO:0005802\ttrans-Golgi network\tcellular_component\n",
      "1\tGO:0005938\tcell cortex\tcellular_component\n",
      "1\tGO:0009524\tphragmoplast\tcellular_component\n",
      "1\tGO:0009574\tpreprophase band\tcellular_component\n",
      "1\tGO:0009579\tthylakoid\tcellular_component\n",
      "1\tGO:0005618\tcell wall\tcellular_component\n",
      "1\tGO:0031012\textracellular matrix\tcellular_component\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#plot_go_terms(rutin_term_count, top_n=10)\n",
    "print(\"# <<< RUTIN >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(rutin_nXcv['mean_scores'])} \\n# count common features: {len(rutin_nXcv['common_features'])}\")\n",
    "tabulate(rut_term_count_biop)\n",
    "tabulate(rut_term_count_molf)\n",
    "tabulate(rut_term_count_cellc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    196 of 427,308 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 14 IDs.\n",
      "100%     14 of     14 study items found in association\n",
      "100%     14 of     14 study items found in population(427308)\n",
      "Calculating 139 uncorrected p-values using fisher_scipy_stats\n",
      "     139 terms are associated with     10 of 427,308 population items\n",
      "     139 terms are associated with     10 of     14 study items\n",
      "  METHOD fdr_bh:\n",
      "     139 GO terms found significant (< 0.05=alpha) (139 enriched +   0 purified): statsmodels fdr_bh\n",
      "      10 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    196 of 427,308 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 14 IDs.\n",
      "100%     14 of     14 study items found in association\n",
      "100%     14 of     14 study items found in population(427308)\n",
      "Calculating 139 uncorrected p-values using fisher_scipy_stats\n",
      "     139 terms are associated with     10 of 427,308 population items\n",
      "     139 terms are associated with     10 of     14 study items\n",
      "  METHOD fdr_bh:\n",
      "     139 GO terms found significant (< 0.05=alpha) (139 enriched +   0 purified): statsmodels fdr_bh\n",
      "      10 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%    196 of 427,308 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 14 IDs.\n",
      "100%     14 of     14 study items found in association\n",
      "100%     14 of     14 study items found in population(427308)\n",
      "Calculating 139 uncorrected p-values using fisher_scipy_stats\n",
      "     139 terms are associated with     10 of 427,308 population items\n",
      "     139 terms are associated with     10 of     14 study items\n",
      "  METHOD fdr_bh:\n",
      "     139 GO terms found significant (< 0.05=alpha) (139 enriched +   0 purified): statsmodels fdr_bh\n",
      "      10 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n"
     ]
    }
   ],
   "source": [
    "with open(\"./lasso_models/10xKfold_lasso_output/trehalose_nXcv.json\", 'r') as file:\n",
    "    trehalose_nXcv = json.load(file)\n",
    "#trehalose_nXcv_matched, trehalose_goids, trehalose_term_count, trehalose_goea = matchgff2(trehalose_nXcv['common_features'], goea=True)\n",
    "trehalose_nXcv_matched, tre_goids_biop, tre_term_count_biop, tre_goea_biop = matchgff2(trehalose_nXcv['common_features'], namespace=['biological_process', 'molecular_function'], depth_threshold=2, goea=True)\n",
    "trehalose_nXcv_matched, tre_goids_molf, tre_term_count_molf, tre_goea_molf = matchgff2(trehalose_nXcv['common_features'], namespace=['molecular_function'], depth_threshold=2, goea=True)\n",
    "trehalose_nXcv_matched, tre_goids_cellc, tre_term_count_cellc, tre_goea_cellc = matchgff2(trehalose_nXcv['common_features'], namespace=['cellular_component'], depth_threshold=2, goea=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# <<< TREHALOSE >>>\n",
      "# 10x3FoldCV mean score: 0.7162380753151226 \n",
      "# count common features: 14\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "3\tGO:0005515\tprotein binding\tmolecular_function\n",
      "2\tGO:0042127\tregulation of cell population proliferation\tbiological_process\n",
      "1\tGO:0006807\tnitrogen compound metabolic process\tbiological_process\n",
      "1\tGO:0006974\tDNA damage response\tbiological_process\n",
      "1\tGO:0043170\tmacromolecule metabolic process\tbiological_process\n",
      "1\tGO:0044238\tprimary metabolic process\tbiological_process\n",
      "1\tGO:0050794\tregulation of cellular process\tbiological_process\n",
      "1\tGO:0003676\tnucleic acid binding\tmolecular_function\n",
      "1\tGO:0034051\tnegative regulation of plant-type hypersensitive response\tbiological_process\n",
      "1\tGO:0050832\tdefense response to fungus\tbiological_process\n",
      "1\tGO:0061057\tpeptidoglycan recognition protein signaling pathway\tbiological_process\n",
      "1\tGO:0061630\tubiquitin protein ligase activity\tmolecular_function\n",
      "1\tGO:0070628\tproteasome binding\tmolecular_function\n",
      "1\tGO:0061631\tubiquitin conjugating enzyme activity\tmolecular_function\n",
      "1\tGO:0016925\tprotein sumoylation\tbiological_process\n",
      "1\tGO:0044314\tprotein K27-linked ubiquitination\tbiological_process\n",
      "1\tGO:0000278\tmitotic cell cycle\tbiological_process\n",
      "1\tGO:0005524\tATP binding\tmolecular_function\n",
      "1\tGO:0043130\tubiquitin binding\tmolecular_function\n",
      "1\tGO:0072344\trescue of stalled ribosome\tbiological_process\n",
      "1\tGO:0035519\tprotein K29-linked ubiquitination\tbiological_process\n",
      "1\tGO:0051865\tprotein autoubiquitination\tbiological_process\n",
      "1\tGO:0070979\tprotein K11-linked ubiquitination\tbiological_process\n",
      "1\tGO:0070936\tprotein K48-linked ubiquitination\tbiological_process\n",
      "1\tGO:0071629\tcytoplasm protein quality control by the ubiquitin-proteasome system\tbiological_process\n",
      "1\tGO:0007140\tmale meiotic nuclear division\tbiological_process\n",
      "1\tGO:0045676\tregulation of R7 cell differentiation\tbiological_process\n",
      "1\tGO:0009960\tendosperm development\tbiological_process\n",
      "1\tGO:0006513\tprotein monoubiquitination\tbiological_process\n",
      "1\tGO:0045879\tnegative regulation of smoothened signaling pathway\tbiological_process\n",
      "1\tGO:0072671\tmitochondria-associated ubiquitin-dependent protein catabolic process\tbiological_process\n",
      "1\tGO:0048132\tfemale germ-line stem cell asymmetric division\tbiological_process\n",
      "1\tGO:0032933\tSREBP signaling pathway\tbiological_process\n",
      "1\tGO:0031625\tubiquitin protein ligase binding\tmolecular_function\n",
      "1\tGO:0031647\tregulation of protein stability\tbiological_process\n",
      "1\tGO:1990116\tribosome-associated ubiquitin-dependent protein catabolic process\tbiological_process\n",
      "1\tGO:0051276\tchromosome organization\tbiological_process\n",
      "1\tGO:0043162\tubiquitin-dependent protein catabolic process via the multivesicular body sorting pathway\tbiological_process\n",
      "1\tGO:0030718\tgerm-line stem cell population maintenance\tbiological_process\n",
      "1\tGO:0097039\tprotein linear polyubiquitination\tbiological_process\n",
      "1\tGO:0031145\tanaphase-promoting complex-dependent catabolic process\tbiological_process\n",
      "1\tGO:0070534\tprotein K63-linked ubiquitination\tbiological_process\n",
      "1\tGO:0010623\tprogrammed cell death involved in cell development\tbiological_process\n",
      "1\tGO:0016322\tneuron remodeling\tbiological_process\n",
      "1\tGO:0085020\tprotein K6-linked ubiquitination\tbiological_process\n",
      "1\tGO:0030674\tprotein-macromolecule adaptor activity\tmolecular_function\n",
      "1\tGO:0005315\tinorganic phosphate transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0015114\tphosphate ion transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0015293\tsymporter activity\tmolecular_function\n",
      "1\tGO:1901683\tarsenate ion transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0009610\tresponse to symbiotic fungus\tbiological_process\n",
      "1\tGO:0009873\tethylene-activated signaling pathway\tbiological_process\n",
      "1\tGO:0016036\tcellular response to phosphate starvation\tbiological_process\n",
      "1\tGO:0035435\tphosphate ion transmembrane transport\tbiological_process\n",
      "1\tGO:0080187\tfloral organ senescence\tbiological_process\n",
      "1\tGO:1901684\tarsenate ion transmembrane transport\tbiological_process\n",
      "1\tGO:0000978\tRNA polymerase II cis-regulatory region sequence-specific DNA binding\tmolecular_function\n",
      "1\tGO:0048767\troot hair elongation\tbiological_process\n",
      "1\tGO:0046983\tprotein dimerization activity\tmolecular_function\n",
      "1\tGO:0048446\tpetal morphogenesis\tbiological_process\n",
      "1\tGO:0009637\tresponse to blue light\tbiological_process\n",
      "1\tGO:0009911\tpositive regulation of flower development\tbiological_process\n",
      "1\tGO:0045824\tnegative regulation of innate immune response\tbiological_process\n",
      "1\tGO:1902448\tpositive regulation of shade avoidance\tbiological_process\n",
      "1\tGO:0000981\tDNA-binding transcription factor activity, RNA polymerase II-specific\tmolecular_function\n",
      "1\tGO:0006357\tregulation of transcription by RNA polymerase II\tbiological_process\n",
      "1\tGO:0009739\tresponse to gibberellin\tbiological_process\n",
      "1\tGO:0048235\tpollen sperm cell differentiation\tbiological_process\n",
      "1\tGO:2000488\tpositive regulation of brassinosteroid biosynthetic process\tbiological_process\n",
      "1\tGO:0071368\tcellular response to cytokinin stimulus\tbiological_process\n",
      "1\tGO:2000762\tregulation of phenylpropanoid metabolic process\tbiological_process\n",
      "1\tGO:0008198\tferrous iron binding\tmolecular_function\n",
      "1\tGO:0008270\tzinc ion binding\tmolecular_function\n",
      "1\tGO:0046566\tDOPA dioxygenase activity\tmolecular_function\n",
      "1\tGO:0050297\tstizolobate synthase activity\tmolecular_function\n",
      "1\tGO:0006725\tcellular aromatic compound metabolic process\tbiological_process\n",
      "1\tGO:0046148\tpigment biosynthetic process\tbiological_process\n",
      "1\tGO:0001671\tATPase activator activity\tmolecular_function\n",
      "1\tGO:0003677\tDNA binding\tmolecular_function\n",
      "1\tGO:0051082\tunfolded protein binding\tmolecular_function\n",
      "1\tGO:0051087\tprotein-folding chaperone binding\tmolecular_function\n",
      "1\tGO:0000122\tnegative regulation of transcription by RNA polymerase II\tbiological_process\n",
      "1\tGO:0006986\tresponse to unfolded protein\tbiological_process\n",
      "1\tGO:0009408\tresponse to heat\tbiological_process\n",
      "1\tGO:0010322\tregulation of isopentenyl diphosphate biosynthetic process, methylerythritol 4-phosphate pathway\tbiological_process\n",
      "1\tGO:0030036\tactin cytoskeleton organization\tbiological_process\n",
      "1\tGO:0030198\textracellular matrix organization\tbiological_process\n",
      "1\tGO:0034504\tprotein localization to nucleus\tbiological_process\n",
      "1\tGO:0042407\tcristae formation\tbiological_process\n",
      "1\tGO:0060710\tchorio-allantoic fusion\tbiological_process\n",
      "1\tGO:0060715\tsyncytiotrophoblast cell differentiation involved in labyrinthine layer development\tbiological_process\n",
      "1\tGO:0060717\tchorion development\tbiological_process\n",
      "1\tGO:0061077\tchaperone-mediated protein folding\tbiological_process\n",
      "1\tGO:1902395\tregulation of 1-deoxy-D-xylulose-5-phosphate synthase activity\tbiological_process\n",
      "1\tGO:0010581\tregulation of starch biosynthetic process\tbiological_process\n",
      "1\tGO:0080050\tregulation of seed development\tbiological_process\n",
      "1\tGO:0060860\tregulation of floral organ abscission\tbiological_process\n",
      "1\tGO:0000976\ttranscription cis-regulatory region binding\tmolecular_function\n",
      "1\tGO:0080037\tnegative regulation of cytokinin-activated signaling pathway\tbiological_process\n",
      "1\tGO:0040019\tpositive regulation of embryonic development\tbiological_process\n",
      "1\tGO:1990110\tcallus formation\tbiological_process\n",
      "1\tGO:0051301\tcell division\tbiological_process\n",
      "1\tGO:0009409\tresponse to cold\tbiological_process\n",
      "1\tGO:0003700\tDNA-binding transcription factor activity\tmolecular_function\n",
      "1\tGO:0045892\tnegative regulation of DNA-templated transcription\tbiological_process\n",
      "1\tGO:0009909\tregulation of flower development\tbiological_process\n",
      "1\tGO:0019760\tglucosinolate metabolic process\tbiological_process\n",
      "1\tGO:2000014\tregulation of endosperm development\tbiological_process\n",
      "1\tGO:0097548\tseed abscission\tbiological_process\n",
      "1\tGO:0048481\tplant ovule development\tbiological_process\n",
      "1\tGO:0010311\tlateral root formation\tbiological_process\n",
      "1\tGO:0010492\tmaintenance of shoot apical meristem identity\tbiological_process\n",
      "1\tGO:0060772\tleaf phyllotactic patterning\tbiological_process\n",
      "1\tGO:0010093\tspecification of floral organ identity\tbiological_process\n",
      "1\tGO:0000723\ttelomere maintenance\tbiological_process\n",
      "1\tGO:0060774\tauxin mediated signaling pathway involved in phyllotactic patterning\tbiological_process\n",
      "1\tGO:0009845\tseed germination\tbiological_process\n",
      "1\tGO:0048830\tadventitious root development\tbiological_process\n",
      "1\tGO:0010449\troot meristem growth\tbiological_process\n",
      "1\tGO:0030154\tcell differentiation\tbiological_process\n",
      "1\tGO:0010228\tvegetative to reproductive phase transition of meristem\tbiological_process\n",
      "1\tGO:0010601\tpositive regulation of auxin biosynthetic process\tbiological_process\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "3\tGO:0005515\tprotein binding\tmolecular_function\n",
      "1\tGO:0003676\tnucleic acid binding\tmolecular_function\n",
      "1\tGO:0061630\tubiquitin protein ligase activity\tmolecular_function\n",
      "1\tGO:0070628\tproteasome binding\tmolecular_function\n",
      "1\tGO:0061631\tubiquitin conjugating enzyme activity\tmolecular_function\n",
      "1\tGO:0005524\tATP binding\tmolecular_function\n",
      "1\tGO:0043130\tubiquitin binding\tmolecular_function\n",
      "1\tGO:0031625\tubiquitin protein ligase binding\tmolecular_function\n",
      "1\tGO:0030674\tprotein-macromolecule adaptor activity\tmolecular_function\n",
      "1\tGO:0005315\tinorganic phosphate transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0015114\tphosphate ion transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0015293\tsymporter activity\tmolecular_function\n",
      "1\tGO:1901683\tarsenate ion transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0000978\tRNA polymerase II cis-regulatory region sequence-specific DNA binding\tmolecular_function\n",
      "1\tGO:0046983\tprotein dimerization activity\tmolecular_function\n",
      "1\tGO:0000981\tDNA-binding transcription factor activity, RNA polymerase II-specific\tmolecular_function\n",
      "1\tGO:0008198\tferrous iron binding\tmolecular_function\n",
      "1\tGO:0008270\tzinc ion binding\tmolecular_function\n",
      "1\tGO:0046566\tDOPA dioxygenase activity\tmolecular_function\n",
      "1\tGO:0050297\tstizolobate synthase activity\tmolecular_function\n",
      "1\tGO:0001671\tATPase activator activity\tmolecular_function\n",
      "1\tGO:0003677\tDNA binding\tmolecular_function\n",
      "1\tGO:0051082\tunfolded protein binding\tmolecular_function\n",
      "1\tGO:0051087\tprotein-folding chaperone binding\tmolecular_function\n",
      "1\tGO:0000976\ttranscription cis-regulatory region binding\tmolecular_function\n",
      "1\tGO:0003700\tDNA-binding transcription factor activity\tmolecular_function\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "3\tGO:0005634\tnucleus\tcellular_component\n",
      "3\tGO:0005886\tplasma membrane\tcellular_component\n",
      "3\tGO:0016021\tmembrane\tcellular_component\n",
      "2\tGO:0005829\tcytosol\tcellular_component\n",
      "2\tGO:0009507\tchloroplast\tcellular_component\n",
      "1\tGO:0005737\tcytoplasm\tcellular_component\n",
      "1\tGO:0005694\tchromosome\tcellular_component\n",
      "1\tGO:0005654\tnucleoplasm\tcellular_component\n",
      "1\tGO:0000151\tubiquitin ligase complex\tcellular_component\n",
      "1\tGO:0005794\tGolgi apparatus\tcellular_component\n",
      "1\tGO:0005622\tintracellular anatomical structure\tcellular_component\n",
      "1\tGO:0001401\tSAM complex\tcellular_component\n",
      "1\tGO:0005788\tendoplasmic reticulum lumen\tcellular_component\n",
      "1\tGO:0016607\tnuclear speck\tcellular_component\n",
      "1\tGO:0061617\tMICOS complex\tcellular_component\n",
      "1\tGO:0005667\ttranscription regulator complex\tcellular_component\n"
     ]
    }
   ],
   "source": [
    "#plot_go_terms(trehalose_term_count, top_n=10)\n",
    "print(\"# <<< TREHALOSE >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(trehalose_nXcv['mean_scores'])} \\n# count common features: {len(trehalose_nXcv['common_features'])}\")\n",
    "\n",
    "tabulate(tre_term_count_biop)\n",
    "tabulate(tre_term_count_molf)\n",
    "tabulate(tre_term_count_cellc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%  1,024 of 976,704 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 32 IDs.\n",
      "100%     32 of     32 study items found in association\n",
      "100%     32 of     32 study items found in population(976704)\n",
      "Calculating 244 uncorrected p-values using fisher_scipy_stats\n",
      "     244 terms are associated with     23 of 976,704 population items\n",
      "     244 terms are associated with     23 of     32 study items\n",
      "  METHOD fdr_bh:\n",
      "     244 GO terms found significant (< 0.05=alpha) (244 enriched +   0 purified): statsmodels fdr_bh\n",
      "      23 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%  1,024 of 976,704 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 32 IDs.\n",
      "100%     32 of     32 study items found in association\n",
      "100%     32 of     32 study items found in population(976704)\n",
      "Calculating 244 uncorrected p-values using fisher_scipy_stats\n",
      "     244 terms are associated with     23 of 976,704 population items\n",
      "     244 terms are associated with     23 of     32 study items\n",
      "  METHOD fdr_bh:\n",
      "     244 GO terms found significant (< 0.05=alpha) (244 enriched +   0 purified): statsmodels fdr_bh\n",
      "      23 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "GO Enrichment Analysis >>\n",
      "\n",
      "Load  Ontology Enrichment Analysis ...\n",
      "  0%  1,024 of 976,704 population items found in association\n",
      "\n",
      "Runing  Ontology Analysis: current study set of 32 IDs.\n",
      "100%     32 of     32 study items found in association\n",
      "100%     32 of     32 study items found in population(976704)\n",
      "Calculating 244 uncorrected p-values using fisher_scipy_stats\n",
      "     244 terms are associated with     23 of 976,704 population items\n",
      "     244 terms are associated with     23 of     32 study items\n",
      "  METHOD fdr_bh:\n",
      "     244 GO terms found significant (< 0.05=alpha) (244 enriched +   0 purified): statsmodels fdr_bh\n",
      "      23 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n"
     ]
    }
   ],
   "source": [
    "with open(\"./lasso_models/10xKfold_lasso_output/oxGlut_nXcv.json\", 'r') as file:\n",
    "    oxGlut_nXcv = json.load(file)\n",
    "#oxGlut_nXcv_matched, oxGlut_goids, oxGlut_term_count, oxGlut_goea = matchgff2(oxGlut_nXcv['common_features'], goea=True)\n",
    "oxGlut_nXcv_matched, oxGlut_goids_biop, oxGlut_term_count_biop, oxGlut_goea_biop = matchgff2(oxGlut_nXcv['common_features'], namespace=['biological_process', 'molecular_function'], depth_threshold=2, goea=True)\n",
    "oxGlut_nXcv_matched, oxGlut_goids_molf, oxGlut_term_count_molf, oxGlut_goea_molf = matchgff2(oxGlut_nXcv['common_features'], namespace=['molecular_function'], depth_threshold=2, goea=True)\n",
    "oxGlut_nXcv_matched, oxGlut_goids_cellc, oxGlut_term_count_cellc, oxGlut_goea_cellc = matchgff2(oxGlut_nXcv['common_features'], namespace=['cellular_component'], depth_threshold=2, goea=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# <<< OXIDIZED GLUTATHIONE >>>\n",
      "# 10x3FoldCV mean score: 0.657258809998042 \n",
      "# count common features: 32\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "10\tGO:0005515\tprotein binding\tmolecular_function\n",
      "4\tGO:0009414\tresponse to water deprivation\tbiological_process\n",
      "3\tGO:0000166\tnucleotide binding\tmolecular_function\n",
      "3\tGO:0003729\tmRNA binding\tmolecular_function\n",
      "2\tGO:0010970\ttransport along microtubule\tbiological_process\n",
      "2\tGO:0050794\tregulation of cellular process\tbiological_process\n",
      "2\tGO:0009409\tresponse to cold\tbiological_process\n",
      "2\tGO:0009651\tresponse to salt stress\tbiological_process\n",
      "2\tGO:0009737\tresponse to abscisic acid\tbiological_process\n",
      "2\tGO:0000978\tRNA polymerase II cis-regulatory region sequence-specific DNA binding\tmolecular_function\n",
      "2\tGO:0010228\tvegetative to reproductive phase transition of meristem\tbiological_process\n",
      "2\tGO:0016310\tphosphorylation\tbiological_process\n",
      "2\tGO:0008270\tzinc ion binding\tmolecular_function\n",
      "2\tGO:0031425\tchloroplast RNA processing\tbiological_process\n",
      "2\tGO:0080156\tmitochondrial mRNA modification\tbiological_process\n",
      "2\tGO:1900865\tchloroplast RNA modification\tbiological_process\n",
      "2\tGO:0015031\tprotein transport\tbiological_process\n",
      "2\tGO:0005524\tATP binding\tmolecular_function\n",
      "2\tGO:0016887\tATP hydrolysis activity\tmolecular_function\n",
      "1\tGO:0008092\tcytoskeletal protein binding\tmolecular_function\n",
      "1\tGO:0002009\tmorphogenesis of an epithelium\tbiological_process\n",
      "1\tGO:0006996\torganelle organization\tbiological_process\n",
      "1\tGO:0007368\tdetermination of left/right symmetry\tbiological_process\n",
      "1\tGO:0008104\tprotein localization\tbiological_process\n",
      "1\tGO:0009860\tpollen tube growth\tbiological_process\n",
      "1\tGO:0031347\tregulation of defense response\tbiological_process\n",
      "1\tGO:0035295\ttube development\tbiological_process\n",
      "1\tGO:0048513\tanimal organ development\tbiological_process\n",
      "1\tGO:0051656\testablishment of organelle localization\tbiological_process\n",
      "1\tGO:0072001\trenal system development\tbiological_process\n",
      "1\tGO:0120036\tplasma membrane bounded cell projection organization\tbiological_process\n",
      "1\tGO:1903046\tmeiotic cell cycle process\tbiological_process\n",
      "1\tGO:0000254\tC-4 methylsterol oxidase activity\tmolecular_function\n",
      "1\tGO:0005506\tiron ion binding\tmolecular_function\n",
      "1\tGO:0009924\toctadecanal decarbonylase activity\tmolecular_function\n",
      "1\tGO:1990465\taldehyde oxygenase (deformylating) activity\tmolecular_function\n",
      "1\tGO:0006723\tcuticle hydrocarbon biosynthetic process\tbiological_process\n",
      "1\tGO:0010025\twax biosynthetic process\tbiological_process\n",
      "1\tGO:0016126\tsterol biosynthetic process\tbiological_process\n",
      "1\tGO:0043447\talkane biosynthetic process\tbiological_process\n",
      "1\tGO:0046184\taldehyde biosynthetic process\tbiological_process\n",
      "1\tGO:0048235\tpollen sperm cell differentiation\tbiological_process\n",
      "1\tGO:0048653\tanther development\tbiological_process\n",
      "1\tGO:0098542\tdefense response to other organism\tbiological_process\n",
      "1\tGO:0003713\ttranscription coactivator activity\tmolecular_function\n",
      "1\tGO:0031490\tchromatin DNA binding\tmolecular_function\n",
      "1\tGO:0009751\tresponse to salicylic acid\tbiological_process\n",
      "1\tGO:0045723\tpositive regulation of fatty acid biosynthetic process\tbiological_process\n",
      "1\tGO:0042659\tregulation of cell fate specification\tbiological_process\n",
      "1\tGO:0001228\tDNA-binding transcription activator activity, RNA polymerase II-specific\tmolecular_function\n",
      "1\tGO:0008542\tvisual learning\tbiological_process\n",
      "1\tGO:0060323\thead morphogenesis\tbiological_process\n",
      "1\tGO:0009954\tproximal/distal pattern formation\tbiological_process\n",
      "1\tGO:0048363\tmucilage pectin metabolic process\tbiological_process\n",
      "1\tGO:0110024\tpositive regulation of cardiac muscle myoblast proliferation\tbiological_process\n",
      "1\tGO:0001742\toenocyte differentiation\tbiological_process\n",
      "1\tGO:0034504\tprotein localization to nucleus\tbiological_process\n",
      "1\tGO:0007380\tspecification of segmental identity, head\tbiological_process\n",
      "1\tGO:0010033\tresponse to organic substance\tbiological_process\n",
      "1\tGO:0010154\tfruit development\tbiological_process\n",
      "1\tGO:0010077\tmaintenance of inflorescence meristem identity\tbiological_process\n",
      "1\tGO:0010076\tmaintenance of floral meristem identity\tbiological_process\n",
      "1\tGO:0009640\tphotomorphogenesis\tbiological_process\n",
      "1\tGO:0007422\tperipheral nervous system development\tbiological_process\n",
      "1\tGO:0046983\tprotein dimerization activity\tmolecular_function\n",
      "1\tGO:0007420\tbrain development\tbiological_process\n",
      "1\tGO:0009965\tleaf morphogenesis\tbiological_process\n",
      "1\tGO:0045931\tpositive regulation of mitotic cell cycle\tbiological_process\n",
      "1\tGO:0003785\tactin monomer binding\tmolecular_function\n",
      "1\tGO:0051015\tactin filament binding\tmolecular_function\n",
      "1\tGO:0010229\tinflorescence development\tbiological_process\n",
      "1\tGO:0010227\tfloral organ abscission\tbiological_process\n",
      "1\tGO:0010223\tsecondary shoot formation\tbiological_process\n",
      "1\tGO:0008285\tnegative regulation of cell population proliferation\tbiological_process\n",
      "1\tGO:0007476\timaginal disc-derived wing morphogenesis\tbiological_process\n",
      "1\tGO:0048735\thaltere morphogenesis\tbiological_process\n",
      "1\tGO:0090470\tshoot organ boundary specification\tbiological_process\n",
      "1\tGO:0045664\tregulation of neuron differentiation\tbiological_process\n",
      "1\tGO:0048457\tfloral whorl morphogenesis\tbiological_process\n",
      "1\tGO:0045944\tpositive regulation of transcription by RNA polymerase II\tbiological_process\n",
      "1\tGO:0000122\tnegative regulation of transcription by RNA polymerase II\tbiological_process\n",
      "1\tGO:0009605\tresponse to external stimulus\tbiological_process\n",
      "1\tGO:0072002\tMalpighian tubule development\tbiological_process\n",
      "1\tGO:2000497\tpositive regulation of cell proliferation involved in compound eye morphogenesis\tbiological_process\n",
      "1\tGO:0007480\timaginal disc-derived leg morphogenesis\tbiological_process\n",
      "1\tGO:0010371\tregulation of gibberellin biosynthetic process\tbiological_process\n",
      "1\tGO:0045638\tnegative regulation of myeloid cell differentiation\tbiological_process\n",
      "1\tGO:0010051\txylem and phloem pattern formation\tbiological_process\n",
      "1\tGO:0008134\ttranscription factor binding\tmolecular_function\n",
      "1\tGO:0080006\tinternode patterning\tbiological_process\n",
      "1\tGO:0048510\tregulation of timing of transition from vegetative to reproductive phase\tbiological_process\n",
      "1\tGO:0010089\txylem development\tbiological_process\n",
      "1\tGO:0043069\tnegative regulation of programmed cell death\tbiological_process\n",
      "1\tGO:0043168\tanion binding\tmolecular_function\n",
      "1\tGO:0006672\tceramide metabolic process\tbiological_process\n",
      "1\tGO:0005509\tcalcium ion binding\tmolecular_function\n",
      "1\tGO:0001729\tceramide kinase activity\tmolecular_function\n",
      "1\tGO:0016773\tphosphotransferase activity, alcohol group as acceptor\tmolecular_function\n",
      "1\tGO:0030148\tsphingolipid biosynthetic process\tbiological_process\n",
      "1\tGO:0052793\tpectin acetylesterase activity\tmolecular_function\n",
      "1\tGO:0003002\tregionalization\tbiological_process\n",
      "1\tGO:0009966\tregulation of signal transduction\tbiological_process\n",
      "1\tGO:0035220\twing disc development\tbiological_process\n",
      "1\tGO:0048731\tsystem development\tbiological_process\n",
      "1\tGO:0071555\tcell wall organization\tbiological_process\n",
      "1\tGO:0004519\tendonuclease activity\tmolecular_function\n",
      "1\tGO:0010214\tseed coat development\tbiological_process\n",
      "1\tGO:0140547\tacquisition of seed longevity\tbiological_process\n",
      "1\tGO:0070979\tprotein K11-linked ubiquitination\tbiological_process\n",
      "1\tGO:0009793\tembryo development ending in seed dormancy\tbiological_process\n",
      "1\tGO:0031145\tanaphase-promoting complex-dependent catabolic process\tbiological_process\n",
      "1\tGO:0045842\tpositive regulation of mitotic metaphase/anaphase transition\tbiological_process\n",
      "1\tGO:0048481\tplant ovule development\tbiological_process\n",
      "1\tGO:0016301\tkinase activity\tmolecular_function\n",
      "1\tGO:0016787\thydrolase activity\tmolecular_function\n",
      "1\tGO:0009416\tresponse to light stimulus\tbiological_process\n",
      "1\tGO:0042742\tdefense response to bacterium\tbiological_process\n",
      "1\tGO:0051716\tcellular response to stimulus\tbiological_process\n",
      "1\tGO:0008195\tphosphatidate phosphatase activity\tmolecular_function\n",
      "1\tGO:0006651\tdiacylglycerol biosynthetic process\tbiological_process\n",
      "1\tGO:0000822\tinositol hexakisphosphate binding\tmolecular_function\n",
      "1\tGO:0003724\tRNA helicase activity\tmolecular_function\n",
      "1\tGO:0006409\ttRNA export from nucleus\tbiological_process\n",
      "1\tGO:0006415\ttranslational termination\tbiological_process\n",
      "1\tGO:0006417\tregulation of translation\tbiological_process\n",
      "1\tGO:0007286\tspermatid development\tbiological_process\n",
      "1\tGO:0009408\tresponse to heat\tbiological_process\n",
      "1\tGO:0016973\tpoly(A)+ mRNA export from nucleus\tbiological_process\n",
      "1\tGO:0000981\tDNA-binding transcription factor activity, RNA polymerase II-specific\tmolecular_function\n",
      "1\tGO:0006357\tregulation of transcription by RNA polymerase II\tbiological_process\n",
      "1\tGO:0009739\tresponse to gibberellin\tbiological_process\n",
      "1\tGO:0042631\tcellular response to water deprivation\tbiological_process\n",
      "1\tGO:0045893\tpositive regulation of DNA-templated transcription\tbiological_process\n",
      "1\tGO:0071215\tcellular response to abscisic acid stimulus\tbiological_process\n",
      "1\tGO:0071472\tcellular response to salt stress\tbiological_process\n",
      "1\tGO:2000214\tregulation of proline metabolic process\tbiological_process\n",
      "1\tGO:2000377\tregulation of reactive oxygen species metabolic process\tbiological_process\n",
      "1\tGO:0033862\tUMP kinase activity\tmolecular_function\n",
      "1\tGO:0006225\tUDP biosynthetic process\tbiological_process\n",
      "1\tGO:0044210\t'de novo' CTP biosynthetic process\tbiological_process\n",
      "1\tGO:0046940\tnucleoside monophosphate phosphorylation\tbiological_process\n",
      "1\tGO:0016920\tpyroglutamyl-peptidase activity\tmolecular_function\n",
      "1\tGO:0006508\tproteolysis\tbiological_process\n",
      "1\tGO:0008017\tmicrotubule binding\tmolecular_function\n",
      "1\tGO:0008574\tplus-end-directed microtubule motor activity\tmolecular_function\n",
      "1\tGO:0019903\tprotein phosphatase binding\tmolecular_function\n",
      "1\tGO:0030507\tspectrin binding\tmolecular_function\n",
      "1\tGO:0031267\tsmall GTPase binding\tmolecular_function\n",
      "1\tGO:0000281\tmitotic cytokinesis\tbiological_process\n",
      "1\tGO:0000911\tcytokinesis by cell plate formation\tbiological_process\n",
      "1\tGO:0007112\tmale meiosis cytokinesis\tbiological_process\n",
      "1\tGO:0051299\tcentrosome separation\tbiological_process\n",
      "1\tGO:0055046\tmicrogametogenesis\tbiological_process\n",
      "1\tGO:0080175\tphragmoplast microtubule organization\tbiological_process\n",
      "1\tGO:0090307\tmitotic spindle assembly\tbiological_process\n",
      "1\tGO:1902414\tprotein localization to cell junction\tbiological_process\n",
      "1\tGO:0015171\tamino acid transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0003333\tamino acid transmembrane transport\tbiological_process\n",
      "1\tGO:0009834\tplant-type secondary cell wall biogenesis\tbiological_process\n",
      "1\tGO:0010315\tauxin export across the plasma membrane\tbiological_process\n",
      "1\tGO:0000302\tresponse to reactive oxygen species\tbiological_process\n",
      "1\tGO:0009658\tchloroplast organization\tbiological_process\n",
      "1\tGO:0048366\tleaf development\tbiological_process\n",
      "1\tGO:0043565\tsequence-specific DNA binding\tmolecular_function\n",
      "1\tGO:0005543\tphospholipid binding\tmolecular_function\n",
      "1\tGO:0097001\tceramide binding\tmolecular_function\n",
      "1\tGO:0006869\tlipid transport\tbiological_process\n",
      "1\tGO:0009958\tpositive gravitropism\tbiological_process\n",
      "1\tGO:0045892\tnegative regulation of DNA-templated transcription\tbiological_process\n",
      "1\tGO:0046872\tmetal ion binding\tmolecular_function\n",
      "1\tGO:0031227\tN/A\tN/A\n",
      "1\tGO:0008380\tRNA splicing\tbiological_process\n",
      "1\tGO:1903047\tmitotic cell cycle process\tbiological_process\n",
      "1\tGO:0051707\tresponse to other organism\tbiological_process\n",
      "1\tGO:0033043\tregulation of organelle organization\tbiological_process\n",
      "1\tGO:0043167\tion binding\tmolecular_function\n",
      "1\tGO:0008289\tlipid binding\tmolecular_function\n",
      "1\tGO:0032981\tmitochondrial respiratory chain complex I assembly\tbiological_process\n",
      "1\tGO:0032446\tprotein modification by small protein conjugation\tbiological_process\n",
      "1\tGO:0072702\tresponse to methyl methanesulfonate\tbiological_process\n",
      "1\tGO:2001020\tN/A\tN/A\n",
      "1\tGO:0009788\tnegative regulation of abscisic acid-activated signaling pathway\tbiological_process\n",
      "1\tGO:0006511\tubiquitin-dependent protein catabolic process\tbiological_process\n",
      "1\tGO:0004842\tubiquitin-protein transferase activity\tmolecular_function\n",
      "1\tGO:0030154\tcell differentiation\tbiological_process\n",
      "1\tGO:0010564\tregulation of cell cycle process\tbiological_process\n",
      "1\tGO:0047268\tgalactinol-raffinose galactosyltransferase activity\tmolecular_function\n",
      "1\tGO:0047274\tgalactinol-sucrose galactosyltransferase activity\tmolecular_function\n",
      "1\tGO:0052692\traffinose alpha-galactosidase activity\tmolecular_function\n",
      "1\tGO:0006979\tresponse to oxidative stress\tbiological_process\n",
      "1\tGO:0034484\traffinose catabolic process\tbiological_process\n",
      "1\tGO:0010182\tsugar mediated signaling pathway\tbiological_process\n",
      "1\tGO:0016554\tcytidine to uridine editing\tbiological_process\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "10\tGO:0005515\tprotein binding\tmolecular_function\n",
      "3\tGO:0000166\tnucleotide binding\tmolecular_function\n",
      "3\tGO:0003729\tmRNA binding\tmolecular_function\n",
      "2\tGO:0000978\tRNA polymerase II cis-regulatory region sequence-specific DNA binding\tmolecular_function\n",
      "2\tGO:0008270\tzinc ion binding\tmolecular_function\n",
      "2\tGO:0005524\tATP binding\tmolecular_function\n",
      "2\tGO:0016887\tATP hydrolysis activity\tmolecular_function\n",
      "1\tGO:0008092\tcytoskeletal protein binding\tmolecular_function\n",
      "1\tGO:0000254\tC-4 methylsterol oxidase activity\tmolecular_function\n",
      "1\tGO:0005506\tiron ion binding\tmolecular_function\n",
      "1\tGO:0009924\toctadecanal decarbonylase activity\tmolecular_function\n",
      "1\tGO:1990465\taldehyde oxygenase (deformylating) activity\tmolecular_function\n",
      "1\tGO:0003713\ttranscription coactivator activity\tmolecular_function\n",
      "1\tGO:0031490\tchromatin DNA binding\tmolecular_function\n",
      "1\tGO:0001228\tDNA-binding transcription activator activity, RNA polymerase II-specific\tmolecular_function\n",
      "1\tGO:0046983\tprotein dimerization activity\tmolecular_function\n",
      "1\tGO:0003785\tactin monomer binding\tmolecular_function\n",
      "1\tGO:0051015\tactin filament binding\tmolecular_function\n",
      "1\tGO:0008134\ttranscription factor binding\tmolecular_function\n",
      "1\tGO:0043168\tanion binding\tmolecular_function\n",
      "1\tGO:0005509\tcalcium ion binding\tmolecular_function\n",
      "1\tGO:0001729\tceramide kinase activity\tmolecular_function\n",
      "1\tGO:0016773\tphosphotransferase activity, alcohol group as acceptor\tmolecular_function\n",
      "1\tGO:0052793\tpectin acetylesterase activity\tmolecular_function\n",
      "1\tGO:0004519\tendonuclease activity\tmolecular_function\n",
      "1\tGO:0016301\tkinase activity\tmolecular_function\n",
      "1\tGO:0016787\thydrolase activity\tmolecular_function\n",
      "1\tGO:0008195\tphosphatidate phosphatase activity\tmolecular_function\n",
      "1\tGO:0000822\tinositol hexakisphosphate binding\tmolecular_function\n",
      "1\tGO:0003724\tRNA helicase activity\tmolecular_function\n",
      "1\tGO:0000981\tDNA-binding transcription factor activity, RNA polymerase II-specific\tmolecular_function\n",
      "1\tGO:0033862\tUMP kinase activity\tmolecular_function\n",
      "1\tGO:0016920\tpyroglutamyl-peptidase activity\tmolecular_function\n",
      "1\tGO:0008017\tmicrotubule binding\tmolecular_function\n",
      "1\tGO:0008574\tplus-end-directed microtubule motor activity\tmolecular_function\n",
      "1\tGO:0019903\tprotein phosphatase binding\tmolecular_function\n",
      "1\tGO:0030507\tspectrin binding\tmolecular_function\n",
      "1\tGO:0031267\tsmall GTPase binding\tmolecular_function\n",
      "1\tGO:0015171\tamino acid transmembrane transporter activity\tmolecular_function\n",
      "1\tGO:0043565\tsequence-specific DNA binding\tmolecular_function\n",
      "1\tGO:0005543\tphospholipid binding\tmolecular_function\n",
      "1\tGO:0097001\tceramide binding\tmolecular_function\n",
      "1\tGO:0046872\tmetal ion binding\tmolecular_function\n",
      "1\tGO:0031227\tN/A\tN/A\n",
      "1\tGO:0043167\tion binding\tmolecular_function\n",
      "1\tGO:0008289\tlipid binding\tmolecular_function\n",
      "1\tGO:2001020\tN/A\tN/A\n",
      "1\tGO:0004842\tubiquitin-protein transferase activity\tmolecular_function\n",
      "1\tGO:0047268\tgalactinol-raffinose galactosyltransferase activity\tmolecular_function\n",
      "1\tGO:0047274\tgalactinol-sucrose galactosyltransferase activity\tmolecular_function\n",
      "1\tGO:0052692\traffinose alpha-galactosidase activity\tmolecular_function\n",
      "count\tGO ID\tGO Term\tnamespace\n",
      "6\tGO:0016021\tmembrane\tcellular_component\n",
      "4\tGO:0005739\tmitochondrion\tcellular_component\n",
      "3\tGO:0005829\tcytosol\tcellular_component\n",
      "3\tGO:0009507\tchloroplast\tcellular_component\n",
      "2\tGO:0005634\tnucleus\tcellular_component\n",
      "2\tGO:0005929\tcilium\tcellular_component\n",
      "2\tGO:0016020\tmembrane\tcellular_component\n",
      "2\tGO:0012505\tendomembrane system\tcellular_component\n",
      "2\tGO:0005737\tcytoplasm\tcellular_component\n",
      "2\tGO:0005886\tplasma membrane\tcellular_component\n",
      "2\tGO:0005794\tGolgi apparatus\tcellular_component\n",
      "1\tGO:0005871\tkinesin complex\tcellular_component\n",
      "1\tGO:0030424\taxon\tcellular_component\n",
      "1\tGO:0031982\tvesicle\tcellular_component\n",
      "1\tGO:0055028\tcortical microtubule\tcellular_component\n",
      "1\tGO:0005789\tendoplasmic reticulum membrane\tcellular_component\n",
      "1\tGO:0071944\tcell periphery\tcellular_component\n",
      "1\tGO:0016592\tmediator complex\tcellular_component\n",
      "1\tGO:0015630\tmicrotubule cytoskeleton\tcellular_component\n",
      "1\tGO:0090575\tRNA polymerase II transcription regulator complex\tcellular_component\n",
      "1\tGO:0045171\tintercellular bridge\tcellular_component\n",
      "1\tGO:0015629\tactin cytoskeleton\tcellular_component\n",
      "1\tGO:0000785\tchromatin\tcellular_component\n",
      "1\tGO:0005654\tnucleoplasm\tcellular_component\n",
      "1\tGO:0005773\tvacuole\tcellular_component\n",
      "1\tGO:0098588\tbounding membrane of organelle\tcellular_component\n",
      "1\tGO:0005576\textracellular region\tcellular_component\n",
      "1\tGO:0005680\tanaphase-promoting complex\tcellular_component\n",
      "1\tGO:0034399\tnuclear periphery\tcellular_component\n",
      "1\tGO:0070013\tintracellular organelle lumen\tcellular_component\n",
      "1\tGO:0031090\torganelle membrane\tcellular_component\n",
      "1\tGO:0005844\tpolysome\tcellular_component\n",
      "1\tGO:0005934\tcellular bud tip\tcellular_component\n",
      "1\tGO:0010494\tcytoplasmic stress granule\tcellular_component\n",
      "1\tGO:0033391\tchromatoid body\tcellular_component\n",
      "1\tGO:0044614\tnuclear pore cytoplasmic filaments\tcellular_component\n",
      "1\tGO:0000922\tspindle pole\tcellular_component\n",
      "1\tGO:0005641\tnuclear envelope lumen\tcellular_component\n",
      "1\tGO:0005813\tcentrosome\tcellular_component\n",
      "1\tGO:0005814\tcentriole\tcellular_component\n",
      "1\tGO:0005876\tspindle microtubule\tcellular_component\n",
      "1\tGO:0009524\tphragmoplast\tcellular_component\n",
      "1\tGO:0016939\tkinesin II complex\tcellular_component\n",
      "1\tGO:0032153\tcell division site\tcellular_component\n",
      "1\tGO:0031969\tchloroplast membrane\tcellular_component\n",
      "1\tGO:0031965\tnuclear membrane\tcellular_component\n",
      "1\tGO:0044232\torganelle membrane contact site\tcellular_component\n",
      "1\tGO:0031227\tN/A\tN/A\n",
      "1\tGO:0005694\tchromosome\tcellular_component\n",
      "1\tGO:0031410\tcytoplasmic vesicle\tcellular_component\n",
      "1\tGO:2001020\tN/A\tN/A\n",
      "1\tGO:0005856\tcytoskeleton\tcellular_component\n",
      "1\tGO:0031981\tnuclear lumen\tcellular_component\n"
     ]
    }
   ],
   "source": [
    "#plot_go_terms(oxGlut_term_count, top_n=10)\n",
    "print(\"# <<< OXIDIZED GLUTATHIONE >>>\")\n",
    "print(f\"# 10x3FoldCV mean score: {np.mean(oxGlut_nXcv['mean_scores'])} \\n# count common features: {len(oxGlut_nXcv['common_features'])}\")\n",
    "tabulate(oxGlut_term_count_biop)\n",
    "tabulate(oxGlut_term_count_molf)\n",
    "tabulate(oxGlut_term_count_cellc)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GO Enrichment Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `GOEnrichmentStudy` object in `goatools`, key arguments:\n",
    "\n",
    "1. **`pop`**:\n",
    "   - The population parameter represents all possible items that could be sampled. In the context of GO enrichment, it's typically the list of all gene IDs from your background set.\n",
    "\n",
    "2. **`assoc`**:\n",
    "   - The association between your items (genes) and the groups (GO terms). This is a dictionary where keys are gene IDs and values are sets of GO IDs.\n",
    "\n",
    "3. **`obo_dag`**:\n",
    "   - An object representing the GO terms and their hierarchical relationships, usually obtained by parsing the GO OBO file with `obo_parser.GODag()`.\n",
    "\n",
    "4. **`propagate_counts`** (optional, default `True`):\n",
    "   - If `True`, the gene counts will be propagated up the GO DAG, so that each term includes counts for all descendant terms.\n",
    "\n",
    "5. **`alpha`** (optional, default `0.05`):\n",
    "   - The significance level for the statistical tests.\n",
    "\n",
    "6. **`methods`** (optional):\n",
    "   - The methods for multiple test correction, such as `['fdr_bh']` for Benjamini-Hochberg.\n",
    "\n",
    "7. **`pvalcalc`** (optional):\n",
    "   - The object for calculating p-values. By default, `goatools` uses Fisher's exact test.\n",
    "\n",
    "8. **`min_overlap`** (optional, default `0.7`):\n",
    "   - The minimum overlap between the study group and the GO term.\n",
    "\n",
    "9. **`prunetree`** (optional, default `True`):\n",
    "   - If `True`, GO terms without any associated genes in the population will be pruned from the analysis.\n",
    "\n",
    "10. **`go2geneids`** (optional):\n",
    "    - A dictionary mapping GO IDs to gene IDs. This can be used instead of `assoc` to provide associations.\n",
    "\n",
    "11. **`geneid2symbol`** (optional):\n",
    "    - A dictionary mapping gene IDs to gene symbols, which can be useful for labeling output.\n",
    "\n",
    "12. **`evidence_set`** (optional):\n",
    "    - A set of evidence codes to use. If not specified, all evidence codes will be included.\n",
    "\n",
    "13. **`relationships`** (optional):\n",
    "    - A set of extra relationship types to include when building the GO DAG, such as `'part_of'` or `'regulates'`.\n",
    "\n",
    "14. **`keep_if_no_id`** (optional, default `False`):\n",
    "    - If `True`, genes without GO IDs in the `assoc` will be kept in the population.\n",
    "\n",
    "15. **`study`**:\n",
    "    - A list of items (genes) for which you want to check enrichment. This is typically the list of genes corresponding to your non-zero features.\n",
    "\n",
    "\n",
    "The `GOEnrichmentStudy` class in the `goatools` library key attributes:\n",
    "\n",
    "1. **`goterm`**:\n",
    "   - This attribute is an object representing the GO term itself. It contains further details about the term, like its name, namespace, and description.\n",
    "\n",
    "2. **`p_uncorrected`**:\n",
    "   - The raw, uncorrected p-value resulting from the enrichment test (e.g., Fisher's exact test).\n",
    "\n",
    "3. **`p_fdr_bh`**:\n",
    "   - The p-value adjusted for multiple testing using the Benjamini-Hochberg false discovery rate method.\n",
    "\n",
    "4. **`enrichment`**:\n",
    "   - This indicates whether the GO term is overrepresented (`'e'` for enriched) or underrepresented (`'p'` for purified) in the study set compared to the background.\n",
    "\n",
    "5. **`ratio_in_study`**:\n",
    "   - A tuple showing the number of genes in the study set that are annotated with this GO term and the total number of genes in the study set.\n",
    "\n",
    "6. **`ratio_in_pop`**:\n",
    "   - A tuple showing the number of genes in the background population that are annotated with this GO term and the total number of genes in the background population.\n",
    "\n",
    "7. **`study_items`**:\n",
    "   - A list of genes from your study set that are annotated with this GO term.\n",
    "\n",
    "8. **`study_count`**:\n",
    "   - The number of genes in your study set that are annotated with this GO term.\n",
    "\n",
    "9. **`pop_count`**:\n",
    "   - The number of genes in the background population that are annotated with this GO term.\n",
    "\n",
    "10. **`study_n`**:\n",
    "    - The total number of genes in your study set.\n",
    "\n",
    "11. **`pop_n`**:\n",
    "    - The total number of genes in the background population.\n",
    "\n",
    "12. **`name`** (if available):\n",
    "    - The name of the GO term.\n",
    "\n",
    "13. **`namespace`** (if available):\n",
    "    - The namespace/category of the GO term, such as biological_process, Molecular Function, or Cellular Component.\n",
    "\n",
    "14. **`depth`** (if available):\n",
    "    - The depth of the GO term in the GO hierarchy.\n",
    "\n",
    "15. **`is_obsolete`** (if available):\n",
    "    - Indicates whether the GO term is obsolete.\n",
    "\n",
    "Remember, the availability of some of these attributes might depend on the specifics of how `GOEnrichmentStudy` is set up and run, and the version of `goatools` you are using. These attributes provide a comprehensive view of each GO term's enrichment analysis results, allowing for detailed interpretation and further analysis.\n",
    "\n",
    "https://github.com/tanghaibao/goatools/blob/main/notebooks/goea_nbt3102.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(236, 190, 70)\n"
     ]
    }
   ],
   "source": [
    "glucose_goea_sig = [r for r in glucose_goea if r.p_fdr_bh < 0.05]\n",
    "citricAcid_goea_sig = [r for r in citricAcid_goea if r.p_fdr_bh < 0.05]\n",
    "sucrose_goea_sig = [r for r in sucrose_goea if r.p_fdr_bh < 0.05]\n",
    "print(f\"{len(glucose_goea_sig), len(citricAcid_goea_sig), len(sucrose_goea_sig)}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py3-6",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
