{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use kernel py3-6\n",
    "import goatools\n",
    "from goatools import obo_parser\n",
    "import re\n",
    "import json\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Function to match non-zero Features in GFF and Map GO-Term to GO-ID using GOATools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def matchgff2(feature, gff_file='/home/t44p/PW_rawdata/Transciptome_GenomeAnnotation/Xele_annotated2_gff_export2.gff', obo_path=\"/home/t44p/PW_rawdata/go_obo/go.obo\"):\n",
    "    \"\"\"\n",
    "    Searches a GFF (General Feature Format) file for specific features and retrieves associated Gene Ontology (GO) terms.\n",
    "\n",
    "    Parameters:\n",
    "    ----------\n",
    "    feature : list or iterable\n",
    "        An iterable of strings representing the features to search for in the GFF file.\n",
    "\n",
    "    gff_file : str, optional\n",
    "        The file path to the GFF file. Defaults to '/home/t44p/PW_rawdata/Transciptome_GenomeAnnotation/Xele_annotated2_gff_export2.gff'.\n",
    "\n",
    "    obo_path : str, optional\n",
    "        The file path to the Gene Ontology .obo file. Defaults to \"/home/t44p/PW_rawdata/go_obo/combined.obo\".\n",
    "\n",
    "    Returns:\n",
    "    -------\n",
    "    tuple of dict\n",
    "        A tuple containing two dictionaries:\n",
    "        1. A dictionary where keys are the features and values are lists of lines (as strings) from the GFF file where these features are found.\n",
    "        2. A dictionary where keys are the features and values are dictionaries. Each inner dictionary has GO IDs as keys and a dictionary with 'name' and 'namespace' of the GO term as values.\n",
    "\n",
    "    Example:\n",
    "    --------\n",
    "    #>>> features = ['gene1', 'gene2']\n",
    "    #>>> matchgff2(features)\n",
    "    ({'gene1': ['line content from GFF file'], 'gene2': ['line content from GFF file']},\n",
    "     {'gene1': {'GO:0000001': {'name': 'term name', 'namespace': 'biological_process'}}, ... })\n",
    "\n",
    "    Notes:\n",
    "    -----\n",
    "    The function uses regular expressions for precise matching of features. It expects the feature followed by a tab character in the GFF file.\n",
    "    Additionally, it extracts GO IDs from matched lines and retrieves their corresponding names and namespaces from the provided Gene Ontology .obo file.\n",
    "    \"\"\"\n",
    "    with open(gff_file, 'r') as file:\n",
    "        go_ontology = obo_parser.GODag(obo_path)\n",
    "        lines_where_feat_found = {}\n",
    "        go_ids = {}\n",
    "\n",
    "        for feat in feature:\n",
    "            file.seek(0)  # reset file pointer to the beginning for each feature\n",
    "            lines_where_feat_found[feat] = []\n",
    "            go_ids[feat] = {}\n",
    "            pattern = re.compile(re.escape(feat) + r'\\t')  # exact match followed by a tab\n",
    "            for line in file:\n",
    "                if pattern.search(line):\n",
    "                    lines_where_feat_found[feat].append(line.strip())  # Store the line (as a string) if feature is found\n",
    "                    # Extract GO id\n",
    "                    match = re.search(r\"Ontology_id=([GO:\\d,]+)\", line.strip())\n",
    "                    if match:\n",
    "                        ids = match.group(1).split(',')\n",
    "                        # Map Terms to Ids\n",
    "                        for id in ids:\n",
    "                            term = go_ontology.get(id)\n",
    "                            if term is not None:\n",
    "                                go_ids[feat][id] = {'name': term.name, 'namespace': term.namespace}\n",
    "                                #print(id, \": \", term.name, term.namespace)\n",
    "                            else:\n",
    "                                go_ids[feat][id] = {'name': None, 'namespace': None}\n",
    "                                #print(id, \":\", \"None\")\n",
    "    \n",
    "        return lines_where_feat_found, go_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<<< GLUCOSE >>>\n",
      "mapping GO terms to GO id's\n",
      "to common non-zero features over 10x 3fold crossvalidation \n",
      "count common features:  26\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "['Xele.ptg000012l.772\\tBlast2GO\\tCDS\\t1\\t424\\t.\\t.\\t.\\t\"ID=Xele.ptg000012l.772_1;Description=RecName: Full=Probable purple acid phosphatase 20; Flags: Precursor;Gene=PAP18;Gene=PAP19;Gene=PAP25;Gene=P80366;Gene=PAP26;Gene=PAP15;Gene=PAP12;Gene=PAP23;Gene=PAP13;Gene=PAP6;Gene=PAP5;Gene=PAP2;Gene=ACPEPP;Gene=PAP1;Gene=PAP10;Gene=PAP21;Gene=PAP11;Gene=PAP22;Gene=PAP20;Gene=PAP;Ontology_id=GO:0005615,GO:0005773,GO:0003993,GO:0004721,GO:0008199,GO:0008270,GO:0030145,GO:0016036,GO:0016311,GO:0032501,GO:0050829,GO:0050832,GO:0051174,GO:0055062;Enzyme_code=EC:3.1.3.2;Enzyme_name=Acid phosphatase\"']\n",
      "GO:0005615 :  {'name': 'extracellular space', 'namespace': 'cellular_component'}\n",
      "GO:0005773 :  {'name': 'vacuole', 'namespace': 'cellular_component'}\n",
      "GO:0003993 :  {'name': 'acid phosphatase activity', 'namespace': 'molecular_function'}\n",
      "GO:0004721 :  {'name': 'phosphoprotein phosphatase activity', 'namespace': 'molecular_function'}\n",
      "GO:0008199 :  {'name': 'ferric iron binding', 'namespace': 'molecular_function'}\n",
      "GO:0008270 :  {'name': 'zinc ion binding', 'namespace': 'molecular_function'}\n",
      "GO:0030145 :  {'name': 'manganese ion binding', 'namespace': 'molecular_function'}\n",
      "GO:0016036 :  {'name': 'cellular response to phosphate starvation', 'namespace': 'biological_process'}\n",
      "GO:0016311 :  {'name': 'dephosphorylation', 'namespace': 'biological_process'}\n",
      "GO:0032501 :  {'name': 'multicellular organismal process', 'namespace': 'biological_process'}\n",
      "GO:0050829 :  {'name': 'defense response to Gram-negative bacterium', 'namespace': 'biological_process'}\n",
      "GO:0050832 :  {'name': 'defense response to fungus', 'namespace': 'biological_process'}\n",
      "GO:0051174 :  {'name': 'regulation of phosphorus metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0055062 :  {'name': 'phosphate ion homeostasis', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000010l.105\\tBlast2GO\\tCDS\\t1\\t147\\t.\\t.\\t.\\tID=Xele.ptg000010l.105_1;Description=UniRef90_A0A6I9RB04uncharacterized protein LOC105046447 n=1 Tax=Elaeis guineensis var. tenera TaxID=51953 RepID=A0A6I9RB04_ELAGV']\n",
      "['Xele.ptg000006l.630\\tBlast2GO\\tCDS\\t1\\t554\\t.\\t.\\t.\\t\"ID=Xele.ptg000006l.630_1;Description=RecName: Full=Lipase; AltName: Full=Triacylglycerol lipase; Flags: Precursor;Gene=P19515;Gene=LIP;Gene=P61871;Gene=OsI_03088;Gene=At1g76200;Gene=LOC_Os01g46240;Gene=faeA;Gene=LOC_Os01g46290;Gene=OsI_03083;Gene=P61872;Gene=OBL1;Ontology_id=GO:0005747,GO:0005811,GO:0090406,GO:0004806,GO:0033878,GO:0047372,GO:0005975,GO:0009057;Enzyme_code=EC:3.1.1.79,EC:3.1.1.23,EC:3.1.1.1,EC:3.1.1.3;Enzyme_name=Hormone-sensitive lipase,Acylglycerol lipase,Carboxylesterase,Triacylglycerol lipase\"']\n",
      "GO:0005747 :  {'name': 'mitochondrial respiratory chain complex I', 'namespace': 'cellular_component'}\n",
      "GO:0005811 :  {'name': 'lipid droplet', 'namespace': 'cellular_component'}\n",
      "GO:0090406 :  {'name': 'pollen tube', 'namespace': 'cellular_component'}\n",
      "GO:0004806 :  {'name': 'triglyceride lipase activity', 'namespace': 'molecular_function'}\n",
      "GO:0033878 :  {'name': 'hormone-sensitive lipase activity', 'namespace': 'molecular_function'}\n",
      "GO:0047372 :  {'name': 'acylglycerol lipase activity', 'namespace': 'molecular_function'}\n",
      "GO:0005975 :  {'name': 'carbohydrate metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0009057 :  {'name': 'macromolecule catabolic process', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000016l.332\\tBlast2GO\\tCDS\\t1\\t520\\t.\\t.\\t.\\tID=Xele.ptg000016l.332_1;Description=RecName: Full=Putative F-box/LRR-repeat protein At1g56400;Gene=At2g04230;Gene=At3g58940;Gene=At5g54820;Gene=At3g26922;Gene=At5g02930']\n",
      "['Xele.ptg000013l.459\\tBlast2GO\\tCDS\\t1\\t329\\t.\\t.\\t.\\t\"ID=Xele.ptg000013l.459_1;Description=RecName: Full=Glucan endo-1,3-beta-glucosidase, basic vacuolar isoform; AltName: Full=(1->3)-beta-glucan endohydrolase; Short=(1->3)-beta-glucanase; AltName: Full=Beta-1,3-endoglucanase, basic; Flags: Precursor;Gene=HGN1;Gene=GNS1;Gene=GN2;Gene=GN1;Gene=Q03773;Gene=GLUB2;Gene=BG2;Gene=BG1;Gene=P15797;Gene=P12257;Gene=BG3;Gene=BANGLUC;Gene=Q01413;Gene=Q02438;Gene=GLUB1;Gene=VIT_06s0061g00120;Gene=At4g16260;Gene=P23546;Gene=P49237;Gene=P27666;Ontology_id=GO:0005576,GO:0005737,GO:0043231,GO:0046658,GO:0008810,GO:0042973,GO:0008152,GO:0009617,GO:0009627;Enzyme_code=EC:3.2.1.21,EC:3.2.1.39,EC:3.2.1.4;Enzyme_name=Beta-glucosidase,Glucan endo-1,3-beta-D-glucosidase,Cellulase\"']\n",
      "GO:0005576 :  {'name': 'extracellular region', 'namespace': 'cellular_component'}\n",
      "GO:0005737 :  {'name': 'cytoplasm', 'namespace': 'cellular_component'}\n",
      "GO:0043231 :  {'name': 'intracellular membrane-bounded organelle', 'namespace': 'cellular_component'}\n",
      "GO:0046658 :  {'name': None, 'namespace': None}\n",
      "GO:0008810 :  {'name': 'cellulase activity', 'namespace': 'molecular_function'}\n",
      "GO:0042973 :  {'name': 'glucan endo-1,3-beta-D-glucosidase activity', 'namespace': 'molecular_function'}\n",
      "GO:0008152 :  {'name': 'metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0009617 :  {'name': 'response to bacterium', 'namespace': 'biological_process'}\n",
      "GO:0009627 :  {'name': 'systemic acquired resistance', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000068l.74\\tBlast2GO\\tCDS\\t1\\t346\\t.\\t.\\t.\\t\"ID=Xele.ptg000068l.74_1;Description=RecName: Full=Calcium uniporter protein, mitochondrial; Flags: Precursor;Gene=At5g66650;Gene=Mcub;Gene=MCUB;Gene=DDB0186972;Gene=At1g57610;Gene=mcu-1;Gene=At5g42610;Gene=At4g36820;Gene=At1g09575;Gene=Mcu;Gene=mcu;Gene=MCU;Gene=At2g23790;Ontology_id=GO:0005654,GO:0009941,GO:0031305,GO:1990246,GO:0005262,GO:0015292,GO:0019855,GO:0042802,GO:0006979,GO:0007615,GO:0008016,GO:0019722,GO:0032024,GO:0036444,GO:0042593,GO:0043009,GO:0051259,GO:0051561,GO:0060027,GO:0070509,GO:0090023,GO:0090141,GO:0090527;Enzyme_code=EC:7;Enzyme_name=Translocases\"']\n",
      "GO:0005654 :  {'name': 'nucleoplasm', 'namespace': 'cellular_component'}\n",
      "GO:0009941 :  {'name': 'chloroplast envelope', 'namespace': 'cellular_component'}\n",
      "GO:0031305 :  {'name': None, 'namespace': None}\n",
      "GO:1990246 :  {'name': 'uniplex complex', 'namespace': 'cellular_component'}\n",
      "GO:0005262 :  {'name': 'calcium channel activity', 'namespace': 'molecular_function'}\n",
      "GO:0015292 :  {'name': 'uniporter activity', 'namespace': 'molecular_function'}\n",
      "GO:0019855 :  {'name': 'calcium channel inhibitor activity', 'namespace': 'molecular_function'}\n",
      "GO:0042802 :  {'name': 'identical protein binding', 'namespace': 'molecular_function'}\n",
      "GO:0006979 :  {'name': 'response to oxidative stress', 'namespace': 'biological_process'}\n",
      "GO:0007615 :  {'name': 'anesthesia-resistant memory', 'namespace': 'biological_process'}\n",
      "GO:0008016 :  {'name': 'regulation of heart contraction', 'namespace': 'biological_process'}\n",
      "GO:0019722 :  {'name': 'calcium-mediated signaling', 'namespace': 'biological_process'}\n",
      "GO:0032024 :  {'name': 'positive regulation of insulin secretion', 'namespace': 'biological_process'}\n",
      "GO:0036444 :  {'name': 'calcium import into the mitochondrion', 'namespace': 'biological_process'}\n",
      "GO:0042593 :  {'name': 'glucose homeostasis', 'namespace': 'biological_process'}\n",
      "GO:0043009 :  {'name': 'chordate embryonic development', 'namespace': 'biological_process'}\n",
      "GO:0051259 :  {'name': 'protein complex oligomerization', 'namespace': 'biological_process'}\n",
      "GO:0051561 :  {'name': 'positive regulation of mitochondrial calcium ion concentration', 'namespace': 'biological_process'}\n",
      "GO:0060027 :  {'name': 'convergent extension involved in gastrulation', 'namespace': 'biological_process'}\n",
      "GO:0070509 :  {'name': 'calcium ion import', 'namespace': 'biological_process'}\n",
      "GO:0090023 :  {'name': 'positive regulation of neutrophil chemotaxis', 'namespace': 'biological_process'}\n",
      "GO:0090141 :  {'name': 'positive regulation of mitochondrial fission', 'namespace': 'biological_process'}\n",
      "GO:0090527 :  {'name': None, 'namespace': None}\n",
      "['Xele.ptg000001l.144\\tBlast2GO\\tCDS\\t1\\t1052\\t.\\t.\\t.\\t\"ID=Xele.ptg000001l.144_1;Description=RecName: Full=Alpha,alpha-trehalose-phosphate synthase [UDP-forming]; AltName: Full=Trehalose-6-phosphate synthase; AltName: Full=UDP-glucose-glucosephosphate glucosyltransferase;Gene=TPS5;Gene=TPS4;Gene=TPS3;Gene=tpsA;Gene=tps1;Gene=TPS2;Gene=tpsB;Gene=TPS1;Ontology_id=GO:0005576,GO:0005634,GO:0005773,GO:0005946,GO:0003825,GO:0004805,GO:0005515,GO:0102986,GO:0005992,GO:0005993,GO:0009793,GO:0009832,GO:0010182,GO:0030437,GO:0034605,GO:0036168,GO:0036180,GO:0048364,GO:0048574,GO:0051301,GO:0070301,GO:0070413,GO:0080186;Enzyme_code=EC:2.4.1.15,EC:3.1.3.12,EC:2.4.1.245;Enzyme_name=Alpha,alpha-trehalose-phosphate synthase (UDP-forming),Trehalose-phosphatase,Alpha,alpha-trehalose synthase\"']\n",
      "GO:0005576 :  {'name': 'extracellular region', 'namespace': 'cellular_component'}\n",
      "GO:0005634 :  {'name': 'nucleus', 'namespace': 'cellular_component'}\n",
      "GO:0005773 :  {'name': 'vacuole', 'namespace': 'cellular_component'}\n",
      "GO:0005946 :  {'name': 'alpha,alpha-trehalose-phosphate synthase complex (UDP-forming)', 'namespace': 'cellular_component'}\n",
      "GO:0003825 :  {'name': 'alpha,alpha-trehalose-phosphate synthase (UDP-forming) activity', 'namespace': 'molecular_function'}\n",
      "GO:0004805 :  {'name': 'trehalose-phosphatase activity', 'namespace': 'molecular_function'}\n",
      "GO:0005515 :  {'name': 'protein binding', 'namespace': 'molecular_function'}\n",
      "GO:0102986 :  {'name': 'trehalose synthase activity', 'namespace': 'molecular_function'}\n",
      "GO:0005992 :  {'name': 'trehalose biosynthetic process', 'namespace': 'biological_process'}\n",
      "GO:0005993 :  {'name': 'trehalose catabolic process', 'namespace': 'biological_process'}\n",
      "GO:0009793 :  {'name': 'embryo development ending in seed dormancy', 'namespace': 'biological_process'}\n",
      "GO:0009832 :  {'name': 'plant-type cell wall biogenesis', 'namespace': 'biological_process'}\n",
      "GO:0010182 :  {'name': 'sugar mediated signaling pathway', 'namespace': 'biological_process'}\n",
      "GO:0030437 :  {'name': 'ascospore formation', 'namespace': 'biological_process'}\n",
      "GO:0034605 :  {'name': 'cellular response to heat', 'namespace': 'biological_process'}\n",
      "GO:0036168 :  {'name': 'filamentous growth of a population of unicellular organisms in response to heat', 'namespace': 'biological_process'}\n",
      "GO:0036180 :  {'name': 'filamentous growth of a population of unicellular organisms in response to biotic stimulus', 'namespace': 'biological_process'}\n",
      "GO:0048364 :  {'name': 'root development', 'namespace': 'biological_process'}\n",
      "GO:0048574 :  {'name': 'long-day photoperiodism, flowering', 'namespace': 'biological_process'}\n",
      "GO:0051301 :  {'name': 'cell division', 'namespace': 'biological_process'}\n",
      "GO:0070301 :  {'name': 'cellular response to hydrogen peroxide', 'namespace': 'biological_process'}\n",
      "GO:0070413 :  {'name': 'trehalose metabolism in response to stress', 'namespace': 'biological_process'}\n",
      "GO:0080186 :  {'name': 'developmental vegetative growth', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000003l.525\\tBlast2GO\\tCDS\\t1\\t87\\t.\\t.\\t.\\tID=Xele.ptg000003l.525_1;Description=KAH9711115.1DnaJ']\n",
      "['Xele.ptg000022l.1586\\tBlast2GO\\tCDS\\t1\\t580\\t.\\t.\\t.\\t\"ID=Xele.ptg000022l.1586_1;Description=RecName: Full=Protein kinase G11A;Gene=PID2;Gene=OsI_021818;Gene=PID;Gene=AGC1-7;Gene=AGC1-5;Gene=RHS3;Gene=PHOT1B;Gene=PHOT1A;Gene=D6PK;Gene=P15792;Gene=LOC_Os06g18830;Gene=KIPK2;Gene=D6PKL1;Gene=KIPK1;Gene=D6PKL2;Gene=D6PKL3;Gene=PHOT2;Ontology_id=GO:0005634,GO:0005794,GO:0009925,GO:0009986,GO:0004674,GO:0004712,GO:0005524,GO:0005546,GO:0005547,GO:0009882,GO:0010314,GO:0019901,GO:0032266,GO:0070273,GO:0070300,GO:0080025,GO:0106310,GO:0007623,GO:0009638,GO:0009734,GO:0009860,GO:0009908,GO:0009958,GO:0010362,GO:0010540,GO:0040008,GO:0046777,GO:0048766,GO:0048767,GO:0048825,GO:0062075,GO:2000114;Enzyme_code=EC:2.7.12.1;Enzyme_name=Dual-specificity kinase\"']\n",
      "GO:0005634 :  {'name': 'nucleus', 'namespace': 'cellular_component'}\n",
      "GO:0005794 :  {'name': 'Golgi apparatus', 'namespace': 'cellular_component'}\n",
      "GO:0009925 :  {'name': 'basal plasma membrane', 'namespace': 'cellular_component'}\n",
      "GO:0009986 :  {'name': 'cell surface', 'namespace': 'cellular_component'}\n",
      "GO:0004674 :  {'name': 'protein serine/threonine kinase activity', 'namespace': 'molecular_function'}\n",
      "GO:0004712 :  {'name': 'protein serine/threonine/tyrosine kinase activity', 'namespace': 'molecular_function'}\n",
      "GO:0005524 :  {'name': 'ATP binding', 'namespace': 'molecular_function'}\n",
      "GO:0005546 :  {'name': 'phosphatidylinositol-4,5-bisphosphate binding', 'namespace': 'molecular_function'}\n",
      "GO:0005547 :  {'name': 'phosphatidylinositol-3,4,5-trisphosphate binding', 'namespace': 'molecular_function'}\n",
      "GO:0009882 :  {'name': 'blue light photoreceptor activity', 'namespace': 'molecular_function'}\n",
      "GO:0010314 :  {'name': 'phosphatidylinositol-5-phosphate binding', 'namespace': 'molecular_function'}\n",
      "GO:0019901 :  {'name': 'protein kinase binding', 'namespace': 'molecular_function'}\n",
      "GO:0032266 :  {'name': 'phosphatidylinositol-3-phosphate binding', 'namespace': 'molecular_function'}\n",
      "GO:0070273 :  {'name': 'phosphatidylinositol-4-phosphate binding', 'namespace': 'molecular_function'}\n",
      "GO:0070300 :  {'name': 'phosphatidic acid binding', 'namespace': 'molecular_function'}\n",
      "GO:0080025 :  {'name': 'phosphatidylinositol-3,5-bisphosphate binding', 'namespace': 'molecular_function'}\n",
      "GO:0106310 :  {'name': 'protein serine kinase activity', 'namespace': 'molecular_function'}\n",
      "GO:0007623 :  {'name': 'circadian rhythm', 'namespace': 'biological_process'}\n",
      "GO:0009638 :  {'name': 'phototropism', 'namespace': 'biological_process'}\n",
      "GO:0009734 :  {'name': 'auxin-activated signaling pathway', 'namespace': 'biological_process'}\n",
      "GO:0009860 :  {'name': 'pollen tube growth', 'namespace': 'biological_process'}\n",
      "GO:0009908 :  {'name': 'flower development', 'namespace': 'biological_process'}\n",
      "GO:0009958 :  {'name': 'positive gravitropism', 'namespace': 'biological_process'}\n",
      "GO:0010362 :  {'name': 'negative regulation of anion channel activity by blue light', 'namespace': 'biological_process'}\n",
      "GO:0010540 :  {'name': 'basipetal auxin transport', 'namespace': 'biological_process'}\n",
      "GO:0040008 :  {'name': 'regulation of growth', 'namespace': 'biological_process'}\n",
      "GO:0046777 :  {'name': 'protein autophosphorylation', 'namespace': 'biological_process'}\n",
      "GO:0048766 :  {'name': 'root hair initiation', 'namespace': 'biological_process'}\n",
      "GO:0048767 :  {'name': 'root hair elongation', 'namespace': 'biological_process'}\n",
      "GO:0048825 :  {'name': 'cotyledon development', 'namespace': 'biological_process'}\n",
      "GO:0062075 :  {'name': 'pollen aperture formation', 'namespace': 'biological_process'}\n",
      "GO:2000114 :  {'name': 'regulation of establishment of cell polarity', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000067l.472\\tBlast2GO\\tCDS\\t1\\t255\\t.\\t.\\t.\\t\"ID=Xele.ptg000067l.472_1;Description=RecName: Full=Xenotropic and polytropic retrovirus receptor 1 homolog;Gene=SPX1;Gene=SPX2;Gene=SPX5;Gene=SPX6;Gene=XPR1;Gene=Xpr1;Gene=SPX3;Gene=xpr1;Gene=SPX4;Ontology_id=GO:0005634,GO:0055062,GO:0080040,GO:0005515,GO:0016020,GO:0005829,GO:0070417\"']\n",
      "GO:0005634 :  {'name': 'nucleus', 'namespace': 'cellular_component'}\n",
      "GO:0055062 :  {'name': 'phosphate ion homeostasis', 'namespace': 'biological_process'}\n",
      "GO:0080040 :  {'name': 'positive regulation of cellular response to phosphate starvation', 'namespace': 'biological_process'}\n",
      "GO:0005515 :  {'name': 'protein binding', 'namespace': 'molecular_function'}\n",
      "GO:0016020 :  {'name': 'membrane', 'namespace': 'cellular_component'}\n",
      "GO:0005829 :  {'name': 'cytosol', 'namespace': 'cellular_component'}\n",
      "GO:0070417 :  {'name': 'cellular response to cold', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000006l.592\\tBlast2GO\\tCDS\\t1\\t538\\t.\\t.\\t.\\t\"ID=Xele.ptg000006l.592_1;Description=RecName: Full=Extended synaptotagmin-3; Short=E-Syt3;Gene=SYT5;Gene=SYT4;Gene=SYT3;Gene=SYT2;Gene=SYT1;Gene=SPAPYUK71.03c;Gene=Esyt3;Gene=Esyt1;Gene=Esyt2;Gene=esyt2-a;Gene=SYT7;Gene=esyt2-b;Gene=esyt3;Gene=CLB;Gene=ESYT3;Gene=TCB1;Gene=ESYT2;Gene=ESYT1;Ontology_id=GO:0000139,GO:0005886,GO:0010008,GO:0016021,GO:0045202,GO:0098827,GO:0005515,GO:0005543,GO:0046872,GO:0006869,GO:0006897,GO:0009306,GO:0016043,GO:1903530\"']\n",
      "GO:0000139 :  {'name': 'Golgi membrane', 'namespace': 'cellular_component'}\n",
      "GO:0005886 :  {'name': 'plasma membrane', 'namespace': 'cellular_component'}\n",
      "GO:0010008 :  {'name': 'endosome membrane', 'namespace': 'cellular_component'}\n",
      "GO:0016021 :  {'name': 'membrane', 'namespace': 'cellular_component'}\n",
      "GO:0045202 :  {'name': 'synapse', 'namespace': 'cellular_component'}\n",
      "GO:0098827 :  {'name': 'endoplasmic reticulum subcompartment', 'namespace': 'cellular_component'}\n",
      "GO:0005515 :  {'name': 'protein binding', 'namespace': 'molecular_function'}\n",
      "GO:0005543 :  {'name': 'phospholipid binding', 'namespace': 'molecular_function'}\n",
      "GO:0046872 :  {'name': 'metal ion binding', 'namespace': 'molecular_function'}\n",
      "GO:0006869 :  {'name': 'lipid transport', 'namespace': 'biological_process'}\n",
      "GO:0006897 :  {'name': 'endocytosis', 'namespace': 'biological_process'}\n",
      "GO:0009306 :  {'name': 'protein secretion', 'namespace': 'biological_process'}\n",
      "GO:0016043 :  {'name': 'cellular component organization', 'namespace': 'biological_process'}\n",
      "GO:1903530 :  {'name': 'regulation of secretion by cell', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000331l.2\\tBlast2GO\\tCDS\\t1\\t479\\t.\\t.\\t.\\tID=Xele.ptg000331l.2_1;Description=UniRef90_A0A5P1FBC4Uncharacterized protein n=2 Tax=Asparagus officinalis TaxID=4686 RepID=A0A5P1FBC4_ASPOF;Gene=MA16_Dca018125;Ontology_id=GO:0009507']\n",
      "GO:0009507 :  {'name': 'chloroplast', 'namespace': 'cellular_component'}\n",
      "['Xele.ptg000063l.174\\tBlast2GO\\tCDS\\t1\\t285\\t.\\t.\\t.\\tID=Xele.ptg000063l.174_1;Description=RecName: Full=Probable membrane-associated kinase regulator 4;Gene=MAKR4;Gene=MAKR3;Gene=MAKR1']\n",
      "['Xele.ptg000025l.274\\tBlast2GO\\tCDS\\t1\\t779\\t.\\t.\\t.\\t\"ID=Xele.ptg000025l.274_1;Description=RecName: Full=Cystinosin homolog;Gene=At3g06410;Gene=LOC_Os12g21700;Gene=At3g48440;Gene=ZFN1;Gene=At5g18550;Gene=LOC_Os01g68860;Gene=At5g40670;Gene=CTNS;Gene=LOC_Os12g18120;Gene=ERS1;Gene=At5g63260;Gene=LOC_Os11g28270;Gene=Ctns;Gene=ctns;Gene=CG17119;Gene=At1g04990;Gene=ctns-1;Gene=LOC_Os01g42970;Ontology_id=GO:0005765,GO:0005770,GO:0005886,GO:0016021,GO:0042470,GO:0003723,GO:0015184,GO:0015295,GO:0034639,GO:0002088,GO:0006749,GO:0006909,GO:0007040,GO:0007420,GO:0007616,GO:0007625,GO:0007628,GO:0008542,GO:0015811,GO:0031326,GO:0046034,GO:0055085,GO:1903427,GO:1904263;Enzyme_code=EC:7;Enzyme_name=Translocases\"']\n",
      "GO:0005765 :  {'name': 'lysosomal membrane', 'namespace': 'cellular_component'}\n",
      "GO:0005770 :  {'name': 'late endosome', 'namespace': 'cellular_component'}\n",
      "GO:0005886 :  {'name': 'plasma membrane', 'namespace': 'cellular_component'}\n",
      "GO:0016021 :  {'name': 'membrane', 'namespace': 'cellular_component'}\n",
      "GO:0042470 :  {'name': 'melanosome', 'namespace': 'cellular_component'}\n",
      "GO:0003723 :  {'name': 'RNA binding', 'namespace': 'molecular_function'}\n",
      "GO:0015184 :  {'name': 'L-cystine transmembrane transporter activity', 'namespace': 'molecular_function'}\n",
      "GO:0015295 :  {'name': 'solute:proton symporter activity', 'namespace': 'molecular_function'}\n",
      "GO:0034639 :  {'name': 'L-amino acid efflux transmembrane transporter activity', 'namespace': 'molecular_function'}\n",
      "GO:0002088 :  {'name': 'lens development in camera-type eye', 'namespace': 'biological_process'}\n",
      "GO:0006749 :  {'name': 'glutathione metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0006909 :  {'name': 'phagocytosis', 'namespace': 'biological_process'}\n",
      "GO:0007040 :  {'name': 'lysosome organization', 'namespace': 'biological_process'}\n",
      "GO:0007420 :  {'name': 'brain development', 'namespace': 'biological_process'}\n",
      "GO:0007616 :  {'name': 'long-term memory', 'namespace': 'biological_process'}\n",
      "GO:0007625 :  {'name': 'grooming behavior', 'namespace': 'biological_process'}\n",
      "GO:0007628 :  {'name': 'adult walking behavior', 'namespace': 'biological_process'}\n",
      "GO:0008542 :  {'name': 'visual learning', 'namespace': 'biological_process'}\n",
      "GO:0015811 :  {'name': 'L-cystine transport', 'namespace': 'biological_process'}\n",
      "GO:0031326 :  {'name': 'regulation of cellular biosynthetic process', 'namespace': 'biological_process'}\n",
      "GO:0046034 :  {'name': 'ATP metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0055085 :  {'name': 'transmembrane transport', 'namespace': 'biological_process'}\n",
      "GO:1903427 :  {'name': 'negative regulation of reactive oxygen species biosynthetic process', 'namespace': 'biological_process'}\n",
      "GO:1904263 :  {'name': 'positive regulation of TORC1 signaling', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000006l.423\\tBlast2GO\\tCDS\\t1\\t328\\t.\\t.\\t.\\t\"ID=Xele.ptg000006l.423_1;Description=UniRef90_A0A1D1YPF7RanBP2-type zinc finger protein At1g67325 n=1 Tax=Anthurium amnicola TaxID=1678845 RepID=A0A1D1YPF7_9ARAE;Gene=L484_026694;Gene=LOC105055042;Gene=IFM89_014740;Gene=A4U43_C01F14510;Gene=Taro_011375;Gene=ZOSMA_37G00930;Gene=LOC104588005;Gene=HHK36_023013;Gene=LOC105059604;Gene=SI8410_02002669;Gene=VIT_04s0023g03280;Gene=LOC109714208;Gene=CKAN_01044000;Gene=At1g67325_13;Ontology_id=GO:0005737,GO:0003729\"']\n",
      "GO:0005737 :  {'name': 'cytoplasm', 'namespace': 'cellular_component'}\n",
      "GO:0003729 :  {'name': 'mRNA binding', 'namespace': 'molecular_function'}\n",
      "['Xele.ptg000007l.418\\tBlast2GO\\tCDS\\t1\\t110\\t.\\t.\\t.\\tID=Xele.ptg000007l.418_1;Description=---NA---']\n",
      "['Xele.ptg000023l.344\\tBlast2GO\\tCDS\\t1\\t237\\t.\\t.\\t.\\tID=Xele.ptg000023l.344_1;Description=RecName: Full=Uncharacterized protein At1g66480;Gene=At1g66480;Ontology_id=GO:0009620']\n",
      "GO:0009620 :  {'name': 'response to fungus', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000025l.124\\tBlast2GO\\tCDS\\t1\\t315\\t.\\t.\\t.\\t\"ID=Xele.ptg000025l.124_1;Description=RecName: Full=Protein PHOSPHATE-INDUCED 1; Flags: Precursor;Gene=EXL1;Gene=EXL2;Gene=EXL3;Gene=EXL4;Gene=PHI-1;Gene=EXL5;Gene=EXL6;Gene=EXO;Gene=EXL7;Ontology_id=GO:0005576,GO:0009505,GO:0001666,GO:0002239,GO:0009741\"']\n",
      "GO:0005576 :  {'name': 'extracellular region', 'namespace': 'cellular_component'}\n",
      "GO:0009505 :  {'name': 'plant-type cell wall', 'namespace': 'cellular_component'}\n",
      "GO:0001666 :  {'name': 'response to hypoxia', 'namespace': 'biological_process'}\n",
      "GO:0002239 :  {'name': 'response to oomycetes', 'namespace': 'biological_process'}\n",
      "GO:0009741 :  {'name': 'response to brassinosteroid', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000011l.430\\tBlast2GO\\tCDS\\t1\\t71\\t.\\t.\\t.\\tID=Xele.ptg000011l.430_1;Description=---NA---']\n",
      "['Xele.ptg000028l.519\\tBlast2GO\\tCDS\\t1\\t817\\t.\\t.\\t.\\t\"ID=Xele.ptg000028l.519_1;Description=RecName: Full=SWI/SNF complex subunit SMARCC2; AltName: Full=BRG1-associated factor 170; Short=BAF170; AltName: Full=SWI/SNF complex 170 kDa subunit; AltName: Full=SWI/SNF-related matrix-associated actin-dependent regulator of chromatin subfamily C member 2;Gene=SMARCC1;Gene=SMARCC2;Gene=ssr1;Gene=SWI3;Gene=SWI3A;Gene=ADA2;Gene=SWI3D;Gene=SWI3B;Gene=SWI3C;Gene=RSC8;Gene=mybX;Gene=Smarcc2;Gene=ssr2;Gene=Smarcc1;Gene=ADA2B;Gene=ADA2A;Ontology_id=GO:0016514,GO:2000070,GO:1901990,GO:0008286,GO:0045597,GO:0003682,GO:0045944,GO:0006337,GO:0042127,GO:0005654,GO:0009887,GO:0000122,GO:0001741,GO:0005515,GO:0071564,GO:0051240,GO:0071565,GO:2000781,GO:0032435\"']\n",
      "GO:0016514 :  {'name': 'SWI/SNF complex', 'namespace': 'cellular_component'}\n",
      "GO:2000070 :  {'name': 'regulation of response to water deprivation', 'namespace': 'biological_process'}\n",
      "GO:1901990 :  {'name': 'regulation of mitotic cell cycle phase transition', 'namespace': 'biological_process'}\n",
      "GO:0008286 :  {'name': 'insulin receptor signaling pathway', 'namespace': 'biological_process'}\n",
      "GO:0045597 :  {'name': 'positive regulation of cell differentiation', 'namespace': 'biological_process'}\n",
      "GO:0003682 :  {'name': 'chromatin binding', 'namespace': 'molecular_function'}\n",
      "GO:0045944 :  {'name': 'positive regulation of transcription by RNA polymerase II', 'namespace': 'biological_process'}\n",
      "GO:0006337 :  {'name': 'nucleosome disassembly', 'namespace': 'biological_process'}\n",
      "GO:0042127 :  {'name': 'regulation of cell population proliferation', 'namespace': 'biological_process'}\n",
      "GO:0005654 :  {'name': 'nucleoplasm', 'namespace': 'cellular_component'}\n",
      "GO:0009887 :  {'name': 'animal organ morphogenesis', 'namespace': 'biological_process'}\n",
      "GO:0000122 :  {'name': 'negative regulation of transcription by RNA polymerase II', 'namespace': 'biological_process'}\n",
      "GO:0001741 :  {'name': 'XY body', 'namespace': 'cellular_component'}\n",
      "GO:0005515 :  {'name': 'protein binding', 'namespace': 'molecular_function'}\n",
      "GO:0071564 :  {'name': 'npBAF complex', 'namespace': 'cellular_component'}\n",
      "GO:0051240 :  {'name': 'positive regulation of multicellular organismal process', 'namespace': 'biological_process'}\n",
      "GO:0071565 :  {'name': 'nBAF complex', 'namespace': 'cellular_component'}\n",
      "GO:2000781 :  {'name': 'positive regulation of double-strand break repair', 'namespace': 'biological_process'}\n",
      "GO:0032435 :  {'name': 'negative regulation of proteasomal ubiquitin-dependent protein catabolic process', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000056l.25\\tBlast2GO\\tCDS\\t1\\t249\\t.\\t.\\t.\\t\"ID=Xele.ptg000056l.25_1;Description=RecName: Full=Bidirectional sugar transporter SWEET4; Short=OsSWEET4;Gene=SWEET8;Gene=SWEET1B;Gene=SWEET1A;Gene=SWEET4;Gene=SWEET5;Gene=SWEET6;Gene=SWEET7;Gene=SWEET1;Gene=SWEET7E;Gene=SWEET7D;Gene=SWEET7C;Gene=SWEET6B;Gene=SWEET7B;Gene=SWEET6A;Ontology_id=GO:0005789,GO:0005887,GO:0005515,GO:0051119,GO:0010584,GO:0010930,GO:0015757,GO:0051260,GO:0052543,GO:0060918;Enzyme_code=EC:7;Enzyme_name=Translocases\"']\n",
      "GO:0005789 :  {'name': 'endoplasmic reticulum membrane', 'namespace': 'cellular_component'}\n",
      "GO:0005887 :  {'name': 'plasma membrane', 'namespace': 'cellular_component'}\n",
      "GO:0005515 :  {'name': 'protein binding', 'namespace': 'molecular_function'}\n",
      "GO:0051119 :  {'name': 'sugar transmembrane transporter activity', 'namespace': 'molecular_function'}\n",
      "GO:0010584 :  {'name': 'pollen exine formation', 'namespace': 'biological_process'}\n",
      "GO:0010930 :  {'name': 'negative regulation of auxin mediated signaling pathway', 'namespace': 'biological_process'}\n",
      "GO:0015757 :  {'name': 'galactose transmembrane transport', 'namespace': 'biological_process'}\n",
      "GO:0051260 :  {'name': 'protein homooligomerization', 'namespace': 'biological_process'}\n",
      "GO:0052543 :  {'name': 'callose deposition in cell wall', 'namespace': 'biological_process'}\n",
      "GO:0060918 :  {'name': 'auxin transport', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000007l.71\\tBlast2GO\\tCDS\\t1\\t410\\t.\\t.\\t.\\t\"ID=Xele.ptg000007l.71_1;Description=RecName: Full=Synaptojanin-1; AltName: Full=Synaptic inositol 1,4,5-trisphosphate 5-phosphatase 1;Gene=INP52;Gene=unc-26;Gene=Synj1;Gene=Inpp5b;Gene=IP5P1;Gene=OCRL;Gene=Ocrl;Gene=INPP5B;Gene=IP5P10;Gene=SYNJ1;Gene=IP5P3;Gene=IP5P2;Gene=IP5P5;Gene=IP5P4;Gene=IP5P7;Gene=IP5P6;Gene=IP5P9;Gene=IP5P8;Ontology_id=GO:0005634,GO:0005768,GO:0005794,GO:0005856,GO:0005886,GO:0043005,GO:0045202,GO:0004439,GO:0004445,GO:0005515,GO:0034485,GO:0009737,GO:0009845,GO:0009932,GO:0010067,GO:0010305,GO:0010588,GO:0016050,GO:0032957,GO:0043933,GO:0046855,GO:0046856,GO:0048016,GO:0048488,GO:0048522,GO:0048766,GO:0051240,GO:0065008,GO:0071472,GO:0071702,GO:0071705,GO:2000369,GO:2000377;Enzyme_code=EC:3.1.3.36,EC:3.1.3.56;Enzyme_name=Phosphoinositide 5-phosphatase,Inositol-polyphosphate 5-phosphatase\"']\n",
      "GO:0005634 :  {'name': 'nucleus', 'namespace': 'cellular_component'}\n",
      "GO:0005768 :  {'name': 'endosome', 'namespace': 'cellular_component'}\n",
      "GO:0005794 :  {'name': 'Golgi apparatus', 'namespace': 'cellular_component'}\n",
      "GO:0005856 :  {'name': 'cytoskeleton', 'namespace': 'cellular_component'}\n",
      "GO:0005886 :  {'name': 'plasma membrane', 'namespace': 'cellular_component'}\n",
      "GO:0043005 :  {'name': 'neuron projection', 'namespace': 'cellular_component'}\n",
      "GO:0045202 :  {'name': 'synapse', 'namespace': 'cellular_component'}\n",
      "GO:0004439 :  {'name': 'phosphatidylinositol-4,5-bisphosphate 5-phosphatase activity', 'namespace': 'molecular_function'}\n",
      "GO:0004445 :  {'name': 'inositol-polyphosphate 5-phosphatase activity', 'namespace': 'molecular_function'}\n",
      "GO:0005515 :  {'name': 'protein binding', 'namespace': 'molecular_function'}\n",
      "GO:0034485 :  {'name': 'phosphatidylinositol-3,4,5-trisphosphate 5-phosphatase activity', 'namespace': 'molecular_function'}\n",
      "GO:0009737 :  {'name': 'response to abscisic acid', 'namespace': 'biological_process'}\n",
      "GO:0009845 :  {'name': 'seed germination', 'namespace': 'biological_process'}\n",
      "GO:0009932 :  {'name': 'cell tip growth', 'namespace': 'biological_process'}\n",
      "GO:0010067 :  {'name': 'procambium histogenesis', 'namespace': 'biological_process'}\n",
      "GO:0010305 :  {'name': 'leaf vascular tissue pattern formation', 'namespace': 'biological_process'}\n",
      "GO:0010588 :  {'name': 'cotyledon vascular tissue pattern formation', 'namespace': 'biological_process'}\n",
      "GO:0016050 :  {'name': 'vesicle organization', 'namespace': 'biological_process'}\n",
      "GO:0032957 :  {'name': 'inositol trisphosphate metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0043933 :  {'name': 'protein-containing complex organization', 'namespace': 'biological_process'}\n",
      "GO:0046855 :  {'name': None, 'namespace': None}\n",
      "GO:0046856 :  {'name': 'phosphatidylinositol dephosphorylation', 'namespace': 'biological_process'}\n",
      "GO:0048016 :  {'name': None, 'namespace': None}\n",
      "GO:0048488 :  {'name': 'synaptic vesicle endocytosis', 'namespace': 'biological_process'}\n",
      "GO:0048522 :  {'name': 'positive regulation of cellular process', 'namespace': 'biological_process'}\n",
      "GO:0048766 :  {'name': 'root hair initiation', 'namespace': 'biological_process'}\n",
      "GO:0051240 :  {'name': 'positive regulation of multicellular organismal process', 'namespace': 'biological_process'}\n",
      "GO:0065008 :  {'name': 'regulation of biological quality', 'namespace': 'biological_process'}\n",
      "GO:0071472 :  {'name': 'cellular response to salt stress', 'namespace': 'biological_process'}\n",
      "GO:0071702 :  {'name': 'organic substance transport', 'namespace': 'biological_process'}\n",
      "GO:0071705 :  {'name': 'nitrogen compound transport', 'namespace': 'biological_process'}\n",
      "GO:2000369 :  {'name': 'regulation of clathrin-dependent endocytosis', 'namespace': 'biological_process'}\n",
      "GO:2000377 :  {'name': 'regulation of reactive oxygen species metabolic process', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000008l.432\\tBlast2GO\\tCDS\\t1\\t561\\t.\\t.\\t.\\t\"ID=Xele.ptg000008l.432_1;Description=RecName: Full=Phosphoacetylglucosamine mutase; Short=PAGM; AltName: Full=Acetylglucosamine phosphomutase; AltName: Full=N-acetylglucosamine-phosphate mutase;Gene=PCM1;Gene=SPAC1296.01c;Gene=LOC_Os07g09720;Gene=AGM1;Gene=PGM3;Gene=SPAC13C5.05c;Gene=Pgm3;Gene=DRT101;Ontology_id=GO:0005737,GO:0000287,GO:0004610,GO:0004614,GO:0005975,GO:0006048,GO:0006281,GO:0006487,GO:0006493,GO:0007283,GO:0019255,GO:0030097,GO:0034221;Enzyme_code=EC:5.4.2.2,EC:5.4.2.3;Enzyme_name=Phosphoglucomutase (alpha-D-glucose-1,6-bisphosphate-dependent),Phosphoacetylglucosamine mutase\"']\n",
      "GO:0005737 :  {'name': 'cytoplasm', 'namespace': 'cellular_component'}\n",
      "GO:0000287 :  {'name': 'magnesium ion binding', 'namespace': 'molecular_function'}\n",
      "GO:0004610 :  {'name': 'phosphoacetylglucosamine mutase activity', 'namespace': 'molecular_function'}\n",
      "GO:0004614 :  {'name': 'phosphoglucomutase activity', 'namespace': 'molecular_function'}\n",
      "GO:0005975 :  {'name': 'carbohydrate metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0006048 :  {'name': 'UDP-N-acetylglucosamine biosynthetic process', 'namespace': 'biological_process'}\n",
      "GO:0006281 :  {'name': 'DNA repair', 'namespace': 'biological_process'}\n",
      "GO:0006487 :  {'name': 'protein N-linked glycosylation', 'namespace': 'biological_process'}\n",
      "GO:0006493 :  {'name': 'protein O-linked glycosylation', 'namespace': 'biological_process'}\n",
      "GO:0007283 :  {'name': 'spermatogenesis', 'namespace': 'biological_process'}\n",
      "GO:0019255 :  {'name': 'glucose 1-phosphate metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0030097 :  {'name': 'hemopoiesis', 'namespace': 'biological_process'}\n",
      "GO:0034221 :  {'name': 'fungal-type cell wall chitin biosynthetic process', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000018l.1563\\tBlast2GO\\tCDS\\t1\\t334\\t.\\t.\\t.\\t\"ID=Xele.ptg000018l.1563_1;Description=RecName: Full=UDP-galactose transporter 1; Short=At-UDP-GalT1;Gene=Slc35e1;Gene=At5g25400;Gene=GPT1;Gene=GONST5;Gene=UUAT1;Gene=At1g12500;Gene=UDP-GALT1;Gene=At4g32390;Gene=At3g17430;Gene=At3g10290;Gene=At2g25520;Gene=At5g11230;Gene=At1g48230;Gene=PPT1;Gene=At3g11320;Gene=TPT;Gene=slc35e3;Gene=PPT3;Gene=SLC35E2B;Gene=At5g05820;Ontology_id=GO:0005783,GO:0005794,GO:0016021,GO:0005457,GO:0005459,GO:0005460,GO:0015315,GO:0071917,GO:0008643,GO:0015714,GO:0015783,GO:0015786,GO:0048359,GO:0072334;Enzyme_code=EC:7;Enzyme_name=Translocases\"']\n",
      "GO:0005783 :  {'name': 'endoplasmic reticulum', 'namespace': 'cellular_component'}\n",
      "GO:0005794 :  {'name': 'Golgi apparatus', 'namespace': 'cellular_component'}\n",
      "GO:0016021 :  {'name': 'membrane', 'namespace': 'cellular_component'}\n",
      "GO:0005457 :  {'name': 'GDP-fucose transmembrane transporter activity', 'namespace': 'molecular_function'}\n",
      "GO:0005459 :  {'name': 'UDP-galactose transmembrane transporter activity', 'namespace': 'molecular_function'}\n",
      "GO:0005460 :  {'name': 'UDP-glucose transmembrane transporter activity', 'namespace': 'molecular_function'}\n",
      "GO:0015315 :  {'name': 'organophosphate:inorganic phosphate antiporter activity', 'namespace': 'molecular_function'}\n",
      "GO:0071917 :  {'name': 'triose-phosphate transmembrane transporter activity', 'namespace': 'molecular_function'}\n",
      "GO:0008643 :  {'name': 'carbohydrate transport', 'namespace': 'biological_process'}\n",
      "GO:0015714 :  {'name': 'phosphoenolpyruvate transport', 'namespace': 'biological_process'}\n",
      "GO:0015783 :  {'name': 'GDP-fucose transmembrane transport', 'namespace': 'biological_process'}\n",
      "GO:0015786 :  {'name': 'UDP-glucose transmembrane transport', 'namespace': 'biological_process'}\n",
      "GO:0048359 :  {'name': 'mucilage metabolic process involved in seed coat development', 'namespace': 'biological_process'}\n",
      "GO:0072334 :  {'name': 'UDP-galactose transmembrane transport', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000013l.792\\tBlast2GO\\tCDS\\t1\\t337\\t.\\t.\\t.\\t\"ID=Xele.ptg000013l.792_1;Description=RecName: Full=Glyceraldehyde-3-phosphate dehydrogenase, cytosolic;Gene=GAPC3;Gene=Q39769;Gene=GAPC2;Gene=GAPC1;Gene=P34783;Gene=GAPC;Ontology_id=GO:0005634,GO:0005740,GO:0005829,GO:0032991,GO:0003677,GO:0004365,GO:0005515,GO:0042301,GO:0050661,GO:0070403,GO:0006094,GO:0006096,GO:0009408,GO:0009416,GO:0010154,GO:0034059,GO:0045893,GO:0046686;Enzyme_code=EC:1.2.1.59,EC:1.2.1.12;Enzyme_name=Glyceraldehyde-3-phosphate dehydrogenase (NAD(P)(+)) (phosphorylating),Glyceraldehyde-3-phosphate dehydrogenase (phosphorylating)\"']\n",
      "GO:0005634 :  {'name': 'nucleus', 'namespace': 'cellular_component'}\n",
      "GO:0005740 :  {'name': 'mitochondrial envelope', 'namespace': 'cellular_component'}\n",
      "GO:0005829 :  {'name': 'cytosol', 'namespace': 'cellular_component'}\n",
      "GO:0032991 :  {'name': 'protein-containing complex', 'namespace': 'cellular_component'}\n",
      "GO:0003677 :  {'name': 'DNA binding', 'namespace': 'molecular_function'}\n",
      "GO:0004365 :  {'name': 'glyceraldehyde-3-phosphate dehydrogenase (NAD+) (phosphorylating) activity', 'namespace': 'molecular_function'}\n",
      "GO:0005515 :  {'name': 'protein binding', 'namespace': 'molecular_function'}\n",
      "GO:0042301 :  {'name': 'phosphate ion binding', 'namespace': 'molecular_function'}\n",
      "GO:0050661 :  {'name': 'NADP binding', 'namespace': 'molecular_function'}\n",
      "GO:0070403 :  {'name': 'NAD+ binding', 'namespace': 'molecular_function'}\n",
      "GO:0006094 :  {'name': 'gluconeogenesis', 'namespace': 'biological_process'}\n",
      "GO:0006096 :  {'name': 'glycolytic process', 'namespace': 'biological_process'}\n",
      "GO:0009408 :  {'name': 'response to heat', 'namespace': 'biological_process'}\n",
      "GO:0009416 :  {'name': 'response to light stimulus', 'namespace': 'biological_process'}\n",
      "GO:0010154 :  {'name': 'fruit development', 'namespace': 'biological_process'}\n",
      "GO:0034059 :  {'name': 'response to anoxia', 'namespace': 'biological_process'}\n",
      "GO:0045893 :  {'name': 'positive regulation of DNA-templated transcription', 'namespace': 'biological_process'}\n",
      "GO:0046686 :  {'name': 'response to cadmium ion', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000021l.354\\tBlast2GO\\tCDS\\t1\\t170\\t.\\t.\\t.\\tID=Xele.ptg000021l.354_1;Description=UniRef90_A0A8B7CRI3uncharacterized N-acetyltransferase p20-like n=2 Tax=Phoenix dactylifera TaxID=42345 RepID=A0A8B7CRI3_PHODC;Gene=A0A7N2LUK5;Gene=LOC105041549;Gene=SVIM_LOCUS423800;Gene=B456_005G044300;Gene=FCM35_KLT11811;Gene=CISIN_1g047585mg;Gene=FCM35_KLT11812;Gene=LOC105033200;Gene=LOC103717924;Gene=FCM35_KLT11813;Gene=FCM35_KLT16508;Gene=F3Y22_tig00110195pilonHSYRG00243;Gene=Taro_049932;Gene=CCACVL1_12759;Gene=HS088_TW04G00034;Gene=FSB_LOCUS15128;Gene=C4D60_Mb07t27160;Gene=DKX38_015389;Ontology_id=GO:0008080;Enzyme_code=EC:2.3.1;Enzyme_name=Acyltransferases']\n",
      "GO:0008080 :  {'name': 'N-acetyltransferase activity', 'namespace': 'molecular_function'}\n"
     ]
    }
   ],
   "source": [
    "with open(\"./models/glucose_nXcv.json\", 'r') as file:\n",
    "    glucose_nXcv = json.load(file)\n",
    "\n",
    "print(\"<<< GLUCOSE >>>\")\n",
    "print(f\"mapping GO terms to GO id's\\nto common non-zero features over 10x 3fold crossvalidation \")\n",
    "print(\"count common features: \", len(glucose_nXcv['common_features']))\n",
    "glucose_nXcv_matched, glucose_goids = matchgff2(glucose_nXcv['common_features'])\n",
    "\n",
    "\n",
    "for ele in glucose_nXcv_matched:\n",
    "    print(f\"{glucose_nXcv_matched[ele]}\")\n",
    "    #print(glucose_goids[ele], \"\\n\")\n",
    "    for id in glucose_goids[ele]:\n",
    "        print(id, \": \", glucose_goids[ele][id])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<<< SUCROSE >>>\n",
      "mapping GO terms to GO id's\n",
      "to common non-zero features over 10x 3fold crossvalidation \n",
      "count common features:  8\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "['Xele.ptg000012l.303\\tBlast2GO\\tCDS\\t1\\t308\\t.\\t.\\t.\\t\"ID=Xele.ptg000012l.303_1;Description=RecName: Full=Phosphoglycerate mutase-like protein AT74; Short=At-74;Gene=DET1;Gene=At3g05170;Gene=At1g08940;Ontology_id=GO:0016791,GO:0005975,GO:0009987;Enzyme_code=EC:3.1.3;Enzyme_name=Acting on ester bonds\"']\n",
      "GO:0016791 :  {'name': 'phosphatase activity', 'namespace': 'molecular_function'}\n",
      "GO:0005975 :  {'name': 'carbohydrate metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0009987 :  {'name': 'cellular process', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000068l.7\\tBlast2GO\\tCDS\\t1\\t283\\t.\\t.\\t.\\t\"ID=Xele.ptg000068l.7_1;Description=RecName: Full=Lateral signaling target protein 2 homolog;Gene=CG6051;Gene=GI24295;Gene=GJ23073;Gene=GE10583;Gene=CPIJ004116;Gene=HGS;Gene=Hgs;Gene=AAEL005241;Gene=GF22946;Gene=GM10129;Gene=GK22512;Gene=VPS27;Gene=GG12136;Ontology_id=GO:0005764,GO:0005769,GO:0005774,GO:0005829,GO:0010008,GO:0033565,GO:0019904,GO:0032266,GO:0042802,GO:0043130,GO:0044389,GO:0046982,GO:0140036,GO:0006622,GO:0009306,GO:0010324,GO:0010642,GO:0016525,GO:0030948,GO:0043162,GO:0043405,GO:0045053,GO:0045324,GO:0046426,GO:0071985,GO:0072657,GO:0140504,GO:1903319,GO:1903543,GO:1904669\"']\n",
      "GO:0005764 :  {'name': 'lysosome', 'namespace': 'cellular_component'}\n",
      "GO:0005769 :  {'name': 'early endosome', 'namespace': 'cellular_component'}\n",
      "GO:0005774 :  {'name': 'vacuolar membrane', 'namespace': 'cellular_component'}\n",
      "GO:0005829 :  {'name': 'cytosol', 'namespace': 'cellular_component'}\n",
      "GO:0010008 :  {'name': 'endosome membrane', 'namespace': 'cellular_component'}\n",
      "GO:0033565 :  {'name': 'ESCRT-0 complex', 'namespace': 'cellular_component'}\n",
      "GO:0019904 :  {'name': 'protein domain specific binding', 'namespace': 'molecular_function'}\n",
      "GO:0032266 :  {'name': 'phosphatidylinositol-3-phosphate binding', 'namespace': 'molecular_function'}\n",
      "GO:0042802 :  {'name': 'identical protein binding', 'namespace': 'molecular_function'}\n",
      "GO:0043130 :  {'name': 'ubiquitin binding', 'namespace': 'molecular_function'}\n",
      "GO:0044389 :  {'name': 'ubiquitin-like protein ligase binding', 'namespace': 'molecular_function'}\n",
      "GO:0046982 :  {'name': 'protein heterodimerization activity', 'namespace': 'molecular_function'}\n",
      "GO:0140036 :  {'name': 'ubiquitin-dependent protein binding', 'namespace': 'molecular_function'}\n",
      "GO:0006622 :  {'name': 'protein targeting to lysosome', 'namespace': 'biological_process'}\n",
      "GO:0009306 :  {'name': 'protein secretion', 'namespace': 'biological_process'}\n",
      "GO:0010324 :  {'name': 'membrane invagination', 'namespace': 'biological_process'}\n",
      "GO:0010642 :  {'name': 'negative regulation of platelet-derived growth factor receptor signaling pathway', 'namespace': 'biological_process'}\n",
      "GO:0016525 :  {'name': 'negative regulation of angiogenesis', 'namespace': 'biological_process'}\n",
      "GO:0030948 :  {'name': 'negative regulation of vascular endothelial growth factor receptor signaling pathway', 'namespace': 'biological_process'}\n",
      "GO:0043162 :  {'name': 'ubiquitin-dependent protein catabolic process via the multivesicular body sorting pathway', 'namespace': 'biological_process'}\n",
      "GO:0043405 :  {'name': 'regulation of MAP kinase activity', 'namespace': 'biological_process'}\n",
      "GO:0045053 :  {'name': 'protein retention in Golgi apparatus', 'namespace': 'biological_process'}\n",
      "GO:0045324 :  {'name': 'late endosome to vacuole transport', 'namespace': 'biological_process'}\n",
      "GO:0046426 :  {'name': 'negative regulation of receptor signaling pathway via JAK-STAT', 'namespace': 'biological_process'}\n",
      "GO:0071985 :  {'name': 'multivesicular body sorting pathway', 'namespace': 'biological_process'}\n",
      "GO:0072657 :  {'name': 'protein localization to membrane', 'namespace': 'biological_process'}\n",
      "GO:0140504 :  {'name': 'microlipophagy', 'namespace': 'biological_process'}\n",
      "GO:1903319 :  {'name': 'positive regulation of protein maturation', 'namespace': 'biological_process'}\n",
      "GO:1903543 :  {'name': 'positive regulation of exosomal secretion', 'namespace': 'biological_process'}\n",
      "GO:1904669 :  {'name': 'ATP export', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000012l.930\\tBlast2GO\\tCDS\\t1\\t321\\t.\\t.\\t.\\t\"ID=Xele.ptg000012l.930_1;Description=RecName: Full=1-aminocyclopropane-1-carboxylate oxidase; Short=ACC oxidase; AltName: Full=Ethylene-forming enzyme; Short=EFE;Gene=Q00985;Gene=MAO1B;Gene=ACO3;Gene=ACO4;Gene=ACO1;Gene=At1g77330;Gene=ACO2;Gene=ACO;Ontology_id=GO:0005783,GO:0005794,GO:0009815,GO:0031418,GO:0046872,GO:0051213,GO:0009620,GO:0009693,GO:0009727,GO:0009835,GO:0010030,GO:0071398,GO:0071732;Enzyme_code=EC:1.14.17.4;Enzyme_name=Aminocyclopropanecarboxylate oxidase\"']\n",
      "GO:0005783 :  {'name': 'endoplasmic reticulum', 'namespace': 'cellular_component'}\n",
      "GO:0005794 :  {'name': 'Golgi apparatus', 'namespace': 'cellular_component'}\n",
      "GO:0009815 :  {'name': '1-aminocyclopropane-1-carboxylate oxidase activity', 'namespace': 'molecular_function'}\n",
      "GO:0031418 :  {'name': 'L-ascorbic acid binding', 'namespace': 'molecular_function'}\n",
      "GO:0046872 :  {'name': 'metal ion binding', 'namespace': 'molecular_function'}\n",
      "GO:0051213 :  {'name': 'dioxygenase activity', 'namespace': 'molecular_function'}\n",
      "GO:0009620 :  {'name': 'response to fungus', 'namespace': 'biological_process'}\n",
      "GO:0009693 :  {'name': 'ethylene biosynthetic process', 'namespace': 'biological_process'}\n",
      "GO:0009727 :  {'name': 'detection of ethylene stimulus', 'namespace': 'biological_process'}\n",
      "GO:0009835 :  {'name': 'fruit ripening', 'namespace': 'biological_process'}\n",
      "GO:0010030 :  {'name': 'positive regulation of seed germination', 'namespace': 'biological_process'}\n",
      "GO:0071398 :  {'name': 'cellular response to fatty acid', 'namespace': 'biological_process'}\n",
      "GO:0071732 :  {'name': 'cellular response to nitric oxide', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000014l.41\\tBlast2GO\\tCDS\\t1\\t190\\t.\\t.\\t.\\t\"ID=Xele.ptg000014l.41_1;Description=RecName: Full=NADH dehydrogenase [ubiquinone] 1 alpha subcomplex subunit 5; AltName: Full=Complex I subunit B13; AltName: Full=Complex I-13kD-B; Short=CI-13kD-B; AltName: Full=NADH-ubiquinone oxidoreductase 13 kDa-B subunit;Gene=NDUFA5;Gene=P80266;Gene=ndufa5;Gene=At5g52840;Gene=Ndufa5;Gene=C33A12.1;Ontology_id=GO:0005747,GO:0006119,GO:0022904\"']\n",
      "GO:0005747 :  {'name': 'mitochondrial respiratory chain complex I', 'namespace': 'cellular_component'}\n",
      "GO:0006119 :  {'name': 'oxidative phosphorylation', 'namespace': 'biological_process'}\n",
      "GO:0022904 :  {'name': 'respiratory electron transport chain', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000063l.174\\tBlast2GO\\tCDS\\t1\\t285\\t.\\t.\\t.\\tID=Xele.ptg000063l.174_1;Description=RecName: Full=Probable membrane-associated kinase regulator 4;Gene=MAKR4;Gene=MAKR3;Gene=MAKR1']\n",
      "['Xele.ptg000008l.151\\tBlast2GO\\tCDS\\t1\\t742\\t.\\t.\\t.\\t\"ID=Xele.ptg000008l.151_1;Description=RecName: Full=Subtilisin-like protease SBT3.8; AltName: Full=Subtilase subfamily 3 member 8; Short=AtSBT3.8; Flags: Precursor;Gene=SBT3.10;Gene=SbtS;Gene=SBT3.9;Gene=SBT3.8;Gene=AIR3;Gene=SBT3.3;Gene=SBT3.15;Gene=SBT3.16;Gene=SBT3.2;Gene=SBT3.1;Gene=SBT3.17;Gene=SBT3.18;Gene=SBT3.7;Gene=SBT3.11;Gene=SBT3.12;Gene=SBT3.6;Gene=SBT3.13;Gene=SBT3.5;Gene=SBT3.4;Gene=SBT3.14;Ontology_id=GO:0005576,GO:0005618,GO:0031012,GO:0004252,GO:0006508,GO:0009567,GO:0009609,GO:0009610,GO:0009682,GO:0009733,GO:0009860,GO:0010102;Enzyme_code=EC:3.4.21;Enzyme_name=Acting on peptide bonds (peptidases)\"']\n",
      "GO:0005576 :  {'name': 'extracellular region', 'namespace': 'cellular_component'}\n",
      "GO:0005618 :  {'name': 'cell wall', 'namespace': 'cellular_component'}\n",
      "GO:0031012 :  {'name': 'extracellular matrix', 'namespace': 'cellular_component'}\n",
      "GO:0004252 :  {'name': 'serine-type endopeptidase activity', 'namespace': 'molecular_function'}\n",
      "GO:0006508 :  {'name': 'proteolysis', 'namespace': 'biological_process'}\n",
      "GO:0009567 :  {'name': 'double fertilization forming a zygote and endosperm', 'namespace': 'biological_process'}\n",
      "GO:0009609 :  {'name': 'response to symbiotic bacterium', 'namespace': 'biological_process'}\n",
      "GO:0009610 :  {'name': 'response to symbiotic fungus', 'namespace': 'biological_process'}\n",
      "GO:0009682 :  {'name': 'induced systemic resistance', 'namespace': 'biological_process'}\n",
      "GO:0009733 :  {'name': 'response to auxin', 'namespace': 'biological_process'}\n",
      "GO:0009860 :  {'name': 'pollen tube growth', 'namespace': 'biological_process'}\n",
      "GO:0010102 :  {'name': 'lateral root morphogenesis', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000009l.79\\tBlast2GO\\tCDS\\t1\\t509\\t.\\t.\\t.\\t\"ID=Xele.ptg000009l.79_1;Description=RecName: Full=Transcription termination factor MTEF18, mitochondrial; AltName: Full=Mitochondrial transcription termination factor 18; AltName: Full=Protein SUPPRESSOR OF HOT1-4 1; Flags: Precursor;Gene=MTERF18;Ontology_id=GO:0005737,GO:0043231,GO:0009987\"']\n",
      "GO:0005737 :  {'name': 'cytoplasm', 'namespace': 'cellular_component'}\n",
      "GO:0043231 :  {'name': 'intracellular membrane-bounded organelle', 'namespace': 'cellular_component'}\n",
      "GO:0009987 :  {'name': 'cellular process', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000012l.442\\tBlast2GO\\tCDS\\t1\\t375\\t.\\t.\\t.\\t\"ID=Xele.ptg000012l.442_1;Description=RecName: Full=BRISC and BRCA1-A complex member 2; AltName: Full=BRCA1-A complex subunit BRE; AltName: Full=BRCA1/BRCA2-containing complex subunit 45; AltName: Full=Brain and reproductive organ-expressed protein;Gene=BABAM2;Gene=Babam2;Gene=babam2;Ontology_id=GO:0005634,GO:0005488,GO:0006807,GO:0006974,GO:0043170,GO:0044238,GO:0050794\"']\n",
      "GO:0005634 :  {'name': 'nucleus', 'namespace': 'cellular_component'}\n",
      "GO:0005488 :  {'name': 'binding', 'namespace': 'molecular_function'}\n",
      "GO:0006807 :  {'name': 'nitrogen compound metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0006974 :  {'name': 'DNA damage response', 'namespace': 'biological_process'}\n",
      "GO:0043170 :  {'name': 'macromolecule metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0044238 :  {'name': 'primary metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0050794 :  {'name': 'regulation of cellular process', 'namespace': 'biological_process'}\n"
     ]
    }
   ],
   "source": [
    "with open(\"./models/sucrose_nXcv.json\", 'r') as file:\n",
    "    sucrose_nXcv = json.load(file)\n",
    "\n",
    "print(\"<<< SUCROSE >>>\")\n",
    "print(f\"mapping GO terms to GO id's\\nto common non-zero features over 10x 3fold crossvalidation \")\n",
    "print(\"count common features: \", len(sucrose_nXcv['common_features']))\n",
    "sucrose_nXcv_matched, sucrose_goids = matchgff2(sucrose_nXcv['common_features'])\n",
    "\n",
    "\n",
    "for ele in sucrose_nXcv_matched:\n",
    "    print(f\"{sucrose_nXcv_matched[ele]}\")\n",
    "    #print(glucose_goids[ele], \"\\n\")\n",
    "    for id in sucrose_goids[ele]:\n",
    "        print(id, \": \", sucrose_goids[ele][id])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<<< CITRIC ACID >>>\n",
      "mapping GO terms to GO id's\n",
      "to common non-zero features over 10x 3fold crossvalidation \n",
      "count common features:  18\n",
      "/home/t44p/PW_rawdata/go_obo/go.obo: fmt(1.2) rel(2023-11-15) 46,228 Terms\n",
      "['Xele.ptg000020l.540\\tBlast2GO\\tCDS\\t1\\t326\\t.\\t.\\t.\\t\"ID=Xele.ptg000020l.540_1;Description=UniRef90_A0A6V7NL77Uncharacterized protein n=1 Tax=Ananas comosus var. bracteatus TaxID=296719 RepID=A0A6V7NL77_ANACO;Gene=C2845_PM03G09980;Gene=LOC104605637;Gene=AMTR_s00003p00261050;Ontology_id=GO:0035267,GO:0016573\"']\n",
      "GO:0035267 :  {'name': 'NuA4 histone acetyltransferase complex', 'namespace': 'cellular_component'}\n",
      "GO:0016573 :  {'name': 'histone acetylation', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000055l.154\\tBlast2GO\\tCDS\\t1\\t129\\t.\\t.\\t.\\t\"ID=Xele.ptg000055l.154_1;Description=RecName: Full=Cysteine desulfurase 1, chloroplastic; AltName: Full=NIFS-like protein 1; Short=CpNifS1; AltName: Full=Plastid sufS-like protein; AltName: Full=Protein AtCpNifS; AltName: Full=Selenocysteine lyase; Flags: Precursor;Gene=NFS2;Ontology_id=GO:0009570,GO:0009000,GO:0030170,GO:0031071,GO:0006534,GO:0010269;Enzyme_code=EC:2.8.1.7,EC:4.4.1.16;Enzyme_name=Cysteine desulfurase,Selenocysteine lyase\"']\n",
      "GO:0009570 :  {'name': 'chloroplast stroma', 'namespace': 'cellular_component'}\n",
      "GO:0009000 :  {'name': 'selenocysteine lyase activity', 'namespace': 'molecular_function'}\n",
      "GO:0030170 :  {'name': 'pyridoxal phosphate binding', 'namespace': 'molecular_function'}\n",
      "GO:0031071 :  {'name': 'cysteine desulfurase activity', 'namespace': 'molecular_function'}\n",
      "GO:0006534 :  {'name': 'cysteine metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0010269 :  {'name': 'response to selenium ion', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000011l.328\\tBlast2GO\\tCDS\\t1\\t287\\t.\\t.\\t.\\tID=Xele.ptg000011l.328_1;Description=---NA---']\n",
      "['Xele.ptg000049l.171\\tBlast2GO\\tCDS\\t1\\t353\\t.\\t.\\t.\\t\"ID=Xele.ptg000049l.171_1;Description=RecName: Full=Inactive protein RESTRICTED TEV MOVEMENT 2; AltName: Full=Inactive restricted tobacco etch virus movement protein 2;Gene=RTM2;Gene=HSP17.9;Ontology_id=GO:0016020,GO:0044419,GO:0050896\"']\n",
      "GO:0016020 :  {'name': 'membrane', 'namespace': 'cellular_component'}\n",
      "GO:0044419 :  {'name': 'biological process involved in interspecies interaction between organisms', 'namespace': 'biological_process'}\n",
      "GO:0050896 :  {'name': 'response to stimulus', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000011l.21\\tBlast2GO\\tCDS\\t1\\t511\\t.\\t.\\t.\\t\"ID=Xele.ptg000011l.21_1;Description=RecName: Full=AFG1-like ATPase; AltName: Full=Lactation elevated protein 1;Gene=AFG1;Gene=SPBC115.02c;Gene=P46441;Gene=lace1b;Gene=AFG1L;Gene=Afg1l;Ontology_id=GO:0031966,GO:0005515,GO:0006123,GO:0035694\"']\n",
      "GO:0031966 :  {'name': 'mitochondrial membrane', 'namespace': 'cellular_component'}\n",
      "GO:0005515 :  {'name': 'protein binding', 'namespace': 'molecular_function'}\n",
      "GO:0006123 :  {'name': 'mitochondrial electron transport, cytochrome c to oxygen', 'namespace': 'biological_process'}\n",
      "GO:0035694 :  {'name': 'mitochondrial protein catabolic process', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000020l.342\\tBlast2GO\\tCDS\\t1\\t591\\t.\\t.\\t.\\t\"ID=Xele.ptg000020l.342_1;Description=RecName: Full=Ubiquitin carboxyl-terminal hydrolase 42; AltName: Full=Deubiquitinating enzyme 42; AltName: Full=Ubiquitin thioesterase 42; AltName: Full=Ubiquitin-specific-processing protease 42;Gene=USP36;Gene=USP17L15;Gene=USP17L18;Gene=USP17L17;Gene=USP42;Gene=USP17L19;Gene=USP17L21;Gene=USP17L20;Gene=UBP25;Gene=USP17L12;Gene=USP17L22;Gene=USP17L11;Gene=USP17L13;Gene=USP17L24;Gene=USP17L5;Gene=Usp42;Gene=Usp36;Gene=USP17L6P;Ontology_id=GO:0016579,GO:0031647,GO:0005737,GO:0010506,GO:0008234,GO:0005488,GO:0005654;Enzyme_code=EC:3.4;Enzyme_name=Acting on peptide bonds (peptidases)\"']\n",
      "GO:0016579 :  {'name': 'protein deubiquitination', 'namespace': 'biological_process'}\n",
      "GO:0031647 :  {'name': 'regulation of protein stability', 'namespace': 'biological_process'}\n",
      "GO:0005737 :  {'name': 'cytoplasm', 'namespace': 'cellular_component'}\n",
      "GO:0010506 :  {'name': 'regulation of autophagy', 'namespace': 'biological_process'}\n",
      "GO:0008234 :  {'name': 'cysteine-type peptidase activity', 'namespace': 'molecular_function'}\n",
      "GO:0005488 :  {'name': 'binding', 'namespace': 'molecular_function'}\n",
      "GO:0005654 :  {'name': 'nucleoplasm', 'namespace': 'cellular_component'}\n",
      "['Xele.ptg000047l.208\\tBlast2GO\\tCDS\\t1\\t142\\t.\\t.\\t.\\tID=Xele.ptg000047l.208_1;Description=---NA---']\n",
      "['Xele.ptg000028l.203\\tBlast2GO\\tCDS\\t1\\t176\\t.\\t.\\t.\\tID=Xele.ptg000028l.203_1;Description=UniRef90_A0A5A7PT37B-box type zinc finger family protein (Fragment) n=1 Tax=Striga asiatica TaxID=4170 RepID=A0A5A7PT37_STRAF']\n",
      "['Xele.ptg000018l.209\\tBlast2GO\\tCDS\\t1\\t579\\t.\\t.\\t.\\t\"ID=Xele.ptg000018l.209_1;Description=RecName: Full=Pentatricopeptide repeat-containing protein At2g33760;Gene=PCMP-H40;Gene=PCMP-H6;Gene=PCMP-H61;Gene=EMB2758;Gene=PCMP-H41;Gene=ELI1;Gene=PCMP-H33;Gene=CRR2;Gene=PCMP-H88;Gene=PCMP-H32;Gene=PCMP-H87;Gene=PCMP-H43;Gene=PCMP-H24;Gene=PCMP-H12;Gene=PCMP-H34;Gene=PCMP-H37;Gene=PCMP-H14;Gene=PCMP-H28;Gene=DYW9;Gene=PCMP-H38;Ontology_id=GO:0005739,GO:0009507,GO:0003729,GO:0004519,GO:0005515,GO:0008270,GO:0000963,GO:0010214,GO:0016554,GO:0031425,GO:0031426,GO:0080156,GO:0140547,GO:1900865;Enzyme_code=EC:3.1;Enzyme_name=Acting on ester bonds\"']\n",
      "GO:0005739 :  {'name': 'mitochondrion', 'namespace': 'cellular_component'}\n",
      "GO:0009507 :  {'name': 'chloroplast', 'namespace': 'cellular_component'}\n",
      "GO:0003729 :  {'name': 'mRNA binding', 'namespace': 'molecular_function'}\n",
      "GO:0004519 :  {'name': 'endonuclease activity', 'namespace': 'molecular_function'}\n",
      "GO:0005515 :  {'name': 'protein binding', 'namespace': 'molecular_function'}\n",
      "GO:0008270 :  {'name': 'zinc ion binding', 'namespace': 'molecular_function'}\n",
      "GO:0000963 :  {'name': 'mitochondrial RNA processing', 'namespace': 'biological_process'}\n",
      "GO:0010214 :  {'name': 'seed coat development', 'namespace': 'biological_process'}\n",
      "GO:0016554 :  {'name': 'cytidine to uridine editing', 'namespace': 'biological_process'}\n",
      "GO:0031425 :  {'name': 'chloroplast RNA processing', 'namespace': 'biological_process'}\n",
      "GO:0031426 :  {'name': 'polycistronic mRNA processing', 'namespace': 'biological_process'}\n",
      "GO:0080156 :  {'name': 'mitochondrial mRNA modification', 'namespace': 'biological_process'}\n",
      "GO:0140547 :  {'name': 'acquisition of seed longevity', 'namespace': 'biological_process'}\n",
      "GO:1900865 :  {'name': 'chloroplast RNA modification', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000042l.33\\tBlast2GO\\tCDS\\t1\\t347\\t.\\t.\\t.\\t\"ID=Xele.ptg000042l.33_1;Description=RecName: Full=Signal recognition particle receptor subunit alpha; Short=SR-alpha; AltName: Full=Docking protein alpha; Short=DP-alpha;Gene=Srpra;Gene=SRPRA;Gene=srpra;Gene=SrpRalpha;Ontology_id=GO:0005789,GO:0003924,GO:0005047,GO:0097159,GO:1901363,GO:0006613,GO:0007409,GO:0045047,GO:0050708;Enzyme_code=EC:3.6.1.15;Enzyme_name=Nucleoside-triphosphate phosphatase\"']\n",
      "GO:0005789 :  {'name': 'endoplasmic reticulum membrane', 'namespace': 'cellular_component'}\n",
      "GO:0003924 :  {'name': 'GTPase activity', 'namespace': 'molecular_function'}\n",
      "GO:0005047 :  {'name': 'signal recognition particle binding', 'namespace': 'molecular_function'}\n",
      "GO:0097159 :  {'name': 'organic cyclic compound binding', 'namespace': 'molecular_function'}\n",
      "GO:1901363 :  {'name': 'heterocyclic compound binding', 'namespace': 'molecular_function'}\n",
      "GO:0006613 :  {'name': 'cotranslational protein targeting to membrane', 'namespace': 'biological_process'}\n",
      "GO:0007409 :  {'name': 'axonogenesis', 'namespace': 'biological_process'}\n",
      "GO:0045047 :  {'name': 'protein targeting to ER', 'namespace': 'biological_process'}\n",
      "GO:0050708 :  {'name': 'regulation of protein secretion', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000042l.232\\tBlast2GO\\tCDS\\t1\\t827\\t.\\t.\\t.\\tID=Xele.ptg000042l.232_1;Description=UniRef90_A0A6I9QAZ7uncharacterized protein LOC105033060 n=3 Tax=Arecaceae TaxID=4710 RepID=A0A6I9QAZ7_ELAGV;Gene=LOC105033060;Gene=LOC105033061;Gene=HHK36_020675;Gene=CICLE_v10004326mg;Gene=LOC105045390;Gene=CK203_075270;Gene=GSMUA_280550.1;Gene=AQUCO_01700396v1;Gene=BVC80_1395g156;Gene=LOC107423229;Gene=LOC116193795;Gene=FH972_015561;Gene=C4D60_Mb01t08710;Ontology_id=GO:0016020']\n",
      "GO:0016020 :  {'name': 'membrane', 'namespace': 'cellular_component'}\n",
      "['Xele.ptg000005l.569\\tBlast2GO\\tCDS\\t1\\t161\\t.\\t.\\t.\\tID=Xele.ptg000005l.569_1;Description=---NA---']\n",
      "['Xele.ptg000025l.209\\tBlast2GO\\tCDS\\t1\\t966\\t.\\t.\\t.\\t\"ID=Xele.ptg000025l.209_1;Description=RecName: Full=Cholesterol transporter ABCA5; AltName: Full=ATP-binding cassette sub-family A member 5;Gene=ABCA1;Gene=ABCA2;Gene=ABCA5;Gene=ABCA6;Gene=ABCA3;Gene=ABCA9;Gene=ABCA7;Gene=ABCA8;Gene=ABCA12;Gene=ABCA10;Gene=abcA6;Gene=ABCA11;Gene=Abca5;Gene=abcA1;Gene=Abca3;Gene=abcA2;Gene=Abca17;Ontology_id=GO:0120013,GO:0005770,GO:0008202,GO:0048316,GO:1901564,GO:0005794,GO:0048519,GO:0008525,GO:0005524,GO:0051641,GO:0140359,GO:0097006,GO:0034204,GO:0010874,GO:0033993,GO:0097208,GO:0140352,GO:0031334,GO:2001140,GO:0046903,GO:0071705,GO:0015245,GO:0055085,GO:0071827,GO:0007165,GO:0055088,GO:0005783,GO:0006650,GO:0006996,GO:0005886,GO:0005764,GO:0097232,GO:0005515,GO:0140328,GO:0009719,GO:0016021,GO:0042908,GO:0032376,GO:0019216;Enzyme_code=EC:7.6.2.1;Enzyme_name=P-type phospholipid transporter\"']\n",
      "GO:0120013 :  {'name': 'lipid transfer activity', 'namespace': 'molecular_function'}\n",
      "GO:0005770 :  {'name': 'late endosome', 'namespace': 'cellular_component'}\n",
      "GO:0008202 :  {'name': 'steroid metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0048316 :  {'name': 'seed development', 'namespace': 'biological_process'}\n",
      "GO:1901564 :  {'name': 'organonitrogen compound metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0005794 :  {'name': 'Golgi apparatus', 'namespace': 'cellular_component'}\n",
      "GO:0048519 :  {'name': 'negative regulation of biological process', 'namespace': 'biological_process'}\n",
      "GO:0008525 :  {'name': 'phosphatidylcholine transporter activity', 'namespace': 'molecular_function'}\n",
      "GO:0005524 :  {'name': 'ATP binding', 'namespace': 'molecular_function'}\n",
      "GO:0051641 :  {'name': 'cellular localization', 'namespace': 'biological_process'}\n",
      "GO:0140359 :  {'name': 'ABC-type transporter activity', 'namespace': 'molecular_function'}\n",
      "GO:0097006 :  {'name': 'regulation of plasma lipoprotein particle levels', 'namespace': 'biological_process'}\n",
      "GO:0034204 :  {'name': 'lipid translocation', 'namespace': 'biological_process'}\n",
      "GO:0010874 :  {'name': 'regulation of cholesterol efflux', 'namespace': 'biological_process'}\n",
      "GO:0033993 :  {'name': 'response to lipid', 'namespace': 'biological_process'}\n",
      "GO:0097208 :  {'name': 'alveolar lamellar body', 'namespace': 'cellular_component'}\n",
      "GO:0140352 :  {'name': 'export from cell', 'namespace': 'biological_process'}\n",
      "GO:0031334 :  {'name': 'positive regulation of protein-containing complex assembly', 'namespace': 'biological_process'}\n",
      "GO:2001140 :  {'name': 'positive regulation of phospholipid transport', 'namespace': 'biological_process'}\n",
      "GO:0046903 :  {'name': 'secretion', 'namespace': 'biological_process'}\n",
      "GO:0071705 :  {'name': 'nitrogen compound transport', 'namespace': 'biological_process'}\n",
      "GO:0015245 :  {'name': 'fatty acid transmembrane transporter activity', 'namespace': 'molecular_function'}\n",
      "GO:0055085 :  {'name': 'transmembrane transport', 'namespace': 'biological_process'}\n",
      "GO:0071827 :  {'name': 'plasma lipoprotein particle organization', 'namespace': 'biological_process'}\n",
      "GO:0007165 :  {'name': 'signal transduction', 'namespace': 'biological_process'}\n",
      "GO:0055088 :  {'name': 'lipid homeostasis', 'namespace': 'biological_process'}\n",
      "GO:0005783 :  {'name': 'endoplasmic reticulum', 'namespace': 'cellular_component'}\n",
      "GO:0006650 :  {'name': 'glycerophospholipid metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0006996 :  {'name': 'organelle organization', 'namespace': 'biological_process'}\n",
      "GO:0005886 :  {'name': 'plasma membrane', 'namespace': 'cellular_component'}\n",
      "GO:0005764 :  {'name': 'lysosome', 'namespace': 'cellular_component'}\n",
      "GO:0097232 :  {'name': 'lamellar body membrane', 'namespace': 'cellular_component'}\n",
      "GO:0005515 :  {'name': 'protein binding', 'namespace': 'molecular_function'}\n",
      "GO:0140328 :  {'name': 'floppase activity', 'namespace': 'molecular_function'}\n",
      "GO:0009719 :  {'name': 'response to endogenous stimulus', 'namespace': 'biological_process'}\n",
      "GO:0016021 :  {'name': 'membrane', 'namespace': 'cellular_component'}\n",
      "GO:0042908 :  {'name': 'xenobiotic transport', 'namespace': 'biological_process'}\n",
      "GO:0032376 :  {'name': 'positive regulation of cholesterol transport', 'namespace': 'biological_process'}\n",
      "GO:0019216 :  {'name': 'regulation of lipid metabolic process', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000022l.1456\\tBlast2GO\\tCDS\\t1\\t291\\t.\\t.\\t.\\t\"ID=Xele.ptg000022l.1456_1;Description=RecName: Full=Transcription factor RHD6; AltName: Full=Basic helix-loop-helix protein 83; Short=AtbHLH83; Short=bHLH 83; AltName: Full=Protein ROOT HAIR DEFECTIVE 6; Short=AtRHD6; AltName: Full=Transcription factor EN 112; AltName: Full=Transcription factor bHLH83; AltName: Full=bHLH transcription factor bHLH083;Gene=RSL1;Gene=PIF1;Gene=RSL2;Gene=BHLH140;Gene=PIF3;Gene=BHLH84;Gene=BHLH133;Gene=LRL1;Gene=LAX1;Gene=LRL3;Gene=RHD6;Gene=LRL2;Gene=RSL4;Gene=APG;Gene=LF;Gene=BHLH87;Gene=RHL1;Gene=BHLH139;Gene=PIL15;Gene=IND;Ontology_id=GO:0005634,GO:0000978,GO:0000981,GO:0003697,GO:0003725,GO:0005524,GO:0010313,GO:0030983,GO:0033699,GO:0042802,GO:0046872,GO:0046983,GO:0047627,GO:1990165,GO:0000012,GO:0006260,GO:0006302,GO:0006357,GO:0006783,GO:0006790,GO:0009150,GO:0009585,GO:0009686,GO:0009704,GO:0009733,GO:0009740,GO:0009826,GO:0009959,GO:0010100,GO:0010161,GO:0010187,GO:0010197,GO:0015995,GO:0031539,GO:0042548,GO:0045893,GO:0048235,GO:0048586,GO:0048587,GO:0048766,GO:0048767,GO:0080113,GO:0090229,GO:0090305,GO:1905421,GO:1990785,GO:2000032;Enzyme_code=EC:3.6.2.1,EC:3.1.11,EC:3.1.15;Enzyme_name=Adenylylsulfatase,Acting on ester bonds,Acting on ester bonds\"']\n",
      "GO:0005634 :  {'name': 'nucleus', 'namespace': 'cellular_component'}\n",
      "GO:0000978 :  {'name': 'RNA polymerase II cis-regulatory region sequence-specific DNA binding', 'namespace': 'molecular_function'}\n",
      "GO:0000981 :  {'name': 'DNA-binding transcription factor activity, RNA polymerase II-specific', 'namespace': 'molecular_function'}\n",
      "GO:0003697 :  {'name': 'single-stranded DNA binding', 'namespace': 'molecular_function'}\n",
      "GO:0003725 :  {'name': 'double-stranded RNA binding', 'namespace': 'molecular_function'}\n",
      "GO:0005524 :  {'name': 'ATP binding', 'namespace': 'molecular_function'}\n",
      "GO:0010313 :  {'name': 'phytochrome binding', 'namespace': 'molecular_function'}\n",
      "GO:0030983 :  {'name': 'mismatched DNA binding', 'namespace': 'molecular_function'}\n",
      "GO:0033699 :  {'name': \"DNA 5'-adenosine monophosphate hydrolase activity\", 'namespace': 'molecular_function'}\n",
      "GO:0042802 :  {'name': 'identical protein binding', 'namespace': 'molecular_function'}\n",
      "GO:0046872 :  {'name': 'metal ion binding', 'namespace': 'molecular_function'}\n",
      "GO:0046983 :  {'name': 'protein dimerization activity', 'namespace': 'molecular_function'}\n",
      "GO:0047627 :  {'name': 'adenylylsulfatase activity', 'namespace': 'molecular_function'}\n",
      "GO:1990165 :  {'name': 'single-strand break-containing DNA binding', 'namespace': 'molecular_function'}\n",
      "GO:0000012 :  {'name': 'single strand break repair', 'namespace': 'biological_process'}\n",
      "GO:0006260 :  {'name': 'DNA replication', 'namespace': 'biological_process'}\n",
      "GO:0006302 :  {'name': 'double-strand break repair', 'namespace': 'biological_process'}\n",
      "GO:0006357 :  {'name': 'regulation of transcription by RNA polymerase II', 'namespace': 'biological_process'}\n",
      "GO:0006783 :  {'name': 'heme biosynthetic process', 'namespace': 'biological_process'}\n",
      "GO:0006790 :  {'name': 'sulfur compound metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0009150 :  {'name': 'purine ribonucleotide metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0009585 :  {'name': 'red, far-red light phototransduction', 'namespace': 'biological_process'}\n",
      "GO:0009686 :  {'name': 'gibberellin biosynthetic process', 'namespace': 'biological_process'}\n",
      "GO:0009704 :  {'name': 'de-etiolation', 'namespace': 'biological_process'}\n",
      "GO:0009733 :  {'name': 'response to auxin', 'namespace': 'biological_process'}\n",
      "GO:0009740 :  {'name': 'gibberellic acid mediated signaling pathway', 'namespace': 'biological_process'}\n",
      "GO:0009826 :  {'name': 'unidimensional cell growth', 'namespace': 'biological_process'}\n",
      "GO:0009959 :  {'name': 'negative gravitropism', 'namespace': 'biological_process'}\n",
      "GO:0010100 :  {'name': 'negative regulation of photomorphogenesis', 'namespace': 'biological_process'}\n",
      "GO:0010161 :  {'name': 'red light signaling pathway', 'namespace': 'biological_process'}\n",
      "GO:0010187 :  {'name': 'negative regulation of seed germination', 'namespace': 'biological_process'}\n",
      "GO:0010197 :  {'name': 'polar nucleus fusion', 'namespace': 'biological_process'}\n",
      "GO:0015995 :  {'name': 'chlorophyll biosynthetic process', 'namespace': 'biological_process'}\n",
      "GO:0031539 :  {'name': 'positive regulation of anthocyanin metabolic process', 'namespace': 'biological_process'}\n",
      "GO:0042548 :  {'name': 'regulation of photosynthesis, light reaction', 'namespace': 'biological_process'}\n",
      "GO:0045893 :  {'name': 'positive regulation of DNA-templated transcription', 'namespace': 'biological_process'}\n",
      "GO:0048235 :  {'name': 'pollen sperm cell differentiation', 'namespace': 'biological_process'}\n",
      "GO:0048586 :  {'name': 'regulation of long-day photoperiodism, flowering', 'namespace': 'biological_process'}\n",
      "GO:0048587 :  {'name': 'regulation of short-day photoperiodism, flowering', 'namespace': 'biological_process'}\n",
      "GO:0048766 :  {'name': 'root hair initiation', 'namespace': 'biological_process'}\n",
      "GO:0048767 :  {'name': 'root hair elongation', 'namespace': 'biological_process'}\n",
      "GO:0080113 :  {'name': 'regulation of seed growth', 'namespace': 'biological_process'}\n",
      "GO:0090229 :  {'name': 'negative regulation of red or far-red light signaling pathway', 'namespace': 'biological_process'}\n",
      "GO:0090305 :  {'name': None, 'namespace': None}\n",
      "GO:1905421 :  {'name': 'regulation of plant organ morphogenesis', 'namespace': 'biological_process'}\n",
      "GO:1990785 :  {'name': 'response to water-immersion restraint stress', 'namespace': 'biological_process'}\n",
      "GO:2000032 :  {'name': 'regulation of secondary shoot formation', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000059l.297\\tBlast2GO\\tCDS\\t1\\t249\\t.\\t.\\t.\\t\"ID=Xele.ptg000059l.297_1;Description=RecName: Full=Probable aquaporin TIP5-1; AltName: Full=Tonoplast intrinsic protein 5-1; Short=OsTIP5;1;Gene=TIP5;1;Gene=TIP1-1;Gene=DIP;Gene=TIP2-1;Gene=TIP1-2;Gene=TIP2-2;Gene=TIP1-3;Gene=TIP2-3;Gene=TIP4-1;Gene=P21653;Gene=TIP5-1;Gene=P24422;Ontology_id=GO:0005739,GO:0016021,GO:0032586,GO:0042807,GO:0090406,GO:0015200,GO:0015204,GO:0015250,GO:0042802,GO:0006833,GO:0007033,GO:0009826,GO:0015840,GO:0048235,GO:0048366,GO:0055085;Enzyme_code=EC:7;Enzyme_name=Translocases\"']\n",
      "GO:0005739 :  {'name': 'mitochondrion', 'namespace': 'cellular_component'}\n",
      "GO:0016021 :  {'name': 'membrane', 'namespace': 'cellular_component'}\n",
      "GO:0032586 :  {'name': 'protein storage vacuole membrane', 'namespace': 'cellular_component'}\n",
      "GO:0042807 :  {'name': 'central vacuole', 'namespace': 'cellular_component'}\n",
      "GO:0090406 :  {'name': 'pollen tube', 'namespace': 'cellular_component'}\n",
      "GO:0015200 :  {'name': 'methylammonium transmembrane transporter activity', 'namespace': 'molecular_function'}\n",
      "GO:0015204 :  {'name': 'urea transmembrane transporter activity', 'namespace': 'molecular_function'}\n",
      "GO:0015250 :  {'name': 'water channel activity', 'namespace': 'molecular_function'}\n",
      "GO:0042802 :  {'name': 'identical protein binding', 'namespace': 'molecular_function'}\n",
      "GO:0006833 :  {'name': 'water transport', 'namespace': 'biological_process'}\n",
      "GO:0007033 :  {'name': 'vacuole organization', 'namespace': 'biological_process'}\n",
      "GO:0009826 :  {'name': 'unidimensional cell growth', 'namespace': 'biological_process'}\n",
      "GO:0015840 :  {'name': 'urea transport', 'namespace': 'biological_process'}\n",
      "GO:0048235 :  {'name': 'pollen sperm cell differentiation', 'namespace': 'biological_process'}\n",
      "GO:0048366 :  {'name': 'leaf development', 'namespace': 'biological_process'}\n",
      "GO:0055085 :  {'name': 'transmembrane transport', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000025l.266\\tBlast2GO\\tCDS\\t1\\t669\\t.\\t.\\t.\\t\"ID=Xele.ptg000025l.266_1;Description=RecName: Full=AUGMIN subunit 6;Gene=AUG6;Ontology_id=GO:0005819,GO:0005874,GO:0009524,GO:0051011,GO:0000226,GO:0009561,GO:0055046\"']\n",
      "GO:0005819 :  {'name': 'spindle', 'namespace': 'cellular_component'}\n",
      "GO:0005874 :  {'name': 'microtubule', 'namespace': 'cellular_component'}\n",
      "GO:0009524 :  {'name': 'phragmoplast', 'namespace': 'cellular_component'}\n",
      "GO:0051011 :  {'name': 'microtubule minus-end binding', 'namespace': 'molecular_function'}\n",
      "GO:0000226 :  {'name': 'microtubule cytoskeleton organization', 'namespace': 'biological_process'}\n",
      "GO:0009561 :  {'name': 'megagametogenesis', 'namespace': 'biological_process'}\n",
      "GO:0055046 :  {'name': 'microgametogenesis', 'namespace': 'biological_process'}\n",
      "['Xele.ptg000022l.1164\\tBlast2GO\\tCDS\\t1\\t429\\t.\\t.\\t.\\tID=Xele.ptg000022l.1164_1;Description=UniRef90_A0A8B7D513uncharacterized protein LOC103724006 n=1 Tax=Phoenix dactylifera TaxID=42345 RepID=A0A8B7D513_PHODC']\n",
      "['Xele.ptg000009l.645\\tBlast2GO\\tCDS\\t1\\t804\\t.\\t.\\t.\\t\"ID=Xele.ptg000009l.645_1;Description=RecName: Full=Serine/threonine-protein kinase CTR1; AltName: Full=Protein CONSTITUTIVE TRIPLE RESPONSE1;Gene=STY46;Gene=HT1;Gene=STY17;Gene=splA;Gene=STY8;Gene=drkD;Gene=drkC;Gene=drkB;Gene=drkA;Gene=DDB_G0267514;Gene=CTR1;Gene=DDB_G0271682;Gene=MAP3K20;Gene=shkC;Gene=shkB;Gene=Map3k20;Gene=shkD;Gene=SIS8;Gene=EDR1;Gene=shkA;Ontology_id=GO:0009750,GO:0090333,GO:0032010,GO:0010182,GO:0007254,GO:0042733,GO:0005789,GO:0001666,GO:2000035,GO:0005829,GO:1904291,GO:0012510,GO:0036289,GO:0038066,GO:0009686,GO:1900424,GO:0005886,GO:0005634,GO:0071480,GO:0097696,GO:0071244,GO:0004706,GO:0010105,GO:0050765,GO:1900150,GO:0009651,GO:0009658,GO:0009414,GO:0000287,GO:0006935,GO:0005524,GO:0005769,GO:0005802,GO:0004713,GO:0004712,GO:0001845,GO:0050829,GO:2000031,GO:2000030,GO:0009744,GO:0009788,GO:0002229,GO:0048510,GO:0005938,GO:0097677,GO:0051593,GO:0043065,GO:2000146,GO:2000069,GO:0030670,GO:0030036,GO:0030435;Enzyme_code=EC:2.7.11.25,EC:2.7.12.1;Enzyme_name=Mitogen-activated protein kinase kinase kinase,Dual-specificity kinase\"']\n",
      "GO:0009750 :  {'name': 'response to fructose', 'namespace': 'biological_process'}\n",
      "GO:0090333 :  {'name': 'regulation of stomatal closure', 'namespace': 'biological_process'}\n",
      "GO:0032010 :  {'name': 'phagolysosome', 'namespace': 'cellular_component'}\n",
      "GO:0010182 :  {'name': 'sugar mediated signaling pathway', 'namespace': 'biological_process'}\n",
      "GO:0007254 :  {'name': 'JNK cascade', 'namespace': 'biological_process'}\n",
      "GO:0042733 :  {'name': 'embryonic digit morphogenesis', 'namespace': 'biological_process'}\n",
      "GO:0005789 :  {'name': 'endoplasmic reticulum membrane', 'namespace': 'cellular_component'}\n",
      "GO:0001666 :  {'name': 'response to hypoxia', 'namespace': 'biological_process'}\n",
      "GO:2000035 :  {'name': 'regulation of stem cell division', 'namespace': 'biological_process'}\n",
      "GO:0005829 :  {'name': 'cytosol', 'namespace': 'cellular_component'}\n",
      "GO:1904291 :  {'name': 'positive regulation of mitotic DNA damage checkpoint', 'namespace': 'biological_process'}\n",
      "GO:0012510 :  {'name': 'trans-Golgi network transport vesicle membrane', 'namespace': 'cellular_component'}\n",
      "GO:0036289 :  {'name': 'peptidyl-serine autophosphorylation', 'namespace': 'biological_process'}\n",
      "GO:0038066 :  {'name': 'p38MAPK cascade', 'namespace': 'biological_process'}\n",
      "GO:0009686 :  {'name': 'gibberellin biosynthetic process', 'namespace': 'biological_process'}\n",
      "GO:1900424 :  {'name': 'regulation of defense response to bacterium', 'namespace': 'biological_process'}\n",
      "GO:0005886 :  {'name': 'plasma membrane', 'namespace': 'cellular_component'}\n",
      "GO:0005634 :  {'name': 'nucleus', 'namespace': 'cellular_component'}\n",
      "GO:0071480 :  {'name': 'cellular response to gamma radiation', 'namespace': 'biological_process'}\n",
      "GO:0097696 :  {'name': 'receptor signaling pathway via STAT', 'namespace': 'biological_process'}\n",
      "GO:0071244 :  {'name': 'cellular response to carbon dioxide', 'namespace': 'biological_process'}\n",
      "GO:0004706 :  {'name': 'JUN kinase kinase kinase activity', 'namespace': 'molecular_function'}\n",
      "GO:0010105 :  {'name': 'negative regulation of ethylene-activated signaling pathway', 'namespace': 'biological_process'}\n",
      "GO:0050765 :  {'name': 'negative regulation of phagocytosis', 'namespace': 'biological_process'}\n",
      "GO:1900150 :  {'name': 'regulation of defense response to fungus', 'namespace': 'biological_process'}\n",
      "GO:0009651 :  {'name': 'response to salt stress', 'namespace': 'biological_process'}\n",
      "GO:0009658 :  {'name': 'chloroplast organization', 'namespace': 'biological_process'}\n",
      "GO:0009414 :  {'name': 'response to water deprivation', 'namespace': 'biological_process'}\n",
      "GO:0000287 :  {'name': 'magnesium ion binding', 'namespace': 'molecular_function'}\n",
      "GO:0006935 :  {'name': 'chemotaxis', 'namespace': 'biological_process'}\n",
      "GO:0005524 :  {'name': 'ATP binding', 'namespace': 'molecular_function'}\n",
      "GO:0005769 :  {'name': 'early endosome', 'namespace': 'cellular_component'}\n",
      "GO:0005802 :  {'name': 'trans-Golgi network', 'namespace': 'cellular_component'}\n",
      "GO:0004713 :  {'name': 'protein tyrosine kinase activity', 'namespace': 'molecular_function'}\n",
      "GO:0004712 :  {'name': 'protein serine/threonine/tyrosine kinase activity', 'namespace': 'molecular_function'}\n",
      "GO:0001845 :  {'name': 'phagolysosome assembly', 'namespace': 'biological_process'}\n",
      "GO:0050829 :  {'name': 'defense response to Gram-negative bacterium', 'namespace': 'biological_process'}\n",
      "GO:2000031 :  {'name': 'regulation of salicylic acid mediated signaling pathway', 'namespace': 'biological_process'}\n",
      "GO:2000030 :  {'name': 'regulation of response to red or far red light', 'namespace': 'biological_process'}\n",
      "GO:0009744 :  {'name': 'response to sucrose', 'namespace': 'biological_process'}\n",
      "GO:0009788 :  {'name': 'negative regulation of abscisic acid-activated signaling pathway', 'namespace': 'biological_process'}\n",
      "GO:0002229 :  {'name': 'defense response to oomycetes', 'namespace': 'biological_process'}\n",
      "GO:0048510 :  {'name': 'regulation of timing of transition from vegetative to reproductive phase', 'namespace': 'biological_process'}\n",
      "GO:0005938 :  {'name': 'cell cortex', 'namespace': 'cellular_component'}\n",
      "GO:0097677 :  {'name': 'STAT family protein binding', 'namespace': 'molecular_function'}\n",
      "GO:0051593 :  {'name': 'response to folic acid', 'namespace': 'biological_process'}\n",
      "GO:0043065 :  {'name': 'positive regulation of apoptotic process', 'namespace': 'biological_process'}\n",
      "GO:2000146 :  {'name': 'negative regulation of cell motility', 'namespace': 'biological_process'}\n",
      "GO:2000069 :  {'name': 'regulation of post-embryonic root development', 'namespace': 'biological_process'}\n",
      "GO:0030670 :  {'name': 'phagocytic vesicle membrane', 'namespace': 'cellular_component'}\n",
      "GO:0030036 :  {'name': 'actin cytoskeleton organization', 'namespace': 'biological_process'}\n",
      "GO:0030435 :  {'name': 'sporulation resulting in formation of a cellular spore', 'namespace': 'biological_process'}\n"
     ]
    }
   ],
   "source": [
    "with open(\"./models/citricAcid_nXcv.json\", 'r') as file:\n",
    "    citricAcid_nXcv = json.load(file)\n",
    "\n",
    "\n",
    "print(\"<<< CITRIC ACID >>>\")\n",
    "print(f\"mapping GO terms to GO id's\\nto common non-zero features over 10x 3fold crossvalidation \")\n",
    "print(\"count common features: \", len(citricAcid_nXcv['common_features']))\n",
    "citricAcid_nXcv_matched, citricAcid_goids = matchgff2(citricAcid_nXcv['common_features'])\n",
    "\n",
    "\n",
    "for ele in citricAcid_nXcv_matched:\n",
    "    print(f\"{citricAcid_nXcv_matched[ele]}\")\n",
    "    #print(glucose_goids[ele], \"\\n\")\n",
    "    for id in citricAcid_goids[ele]:\n",
    "        print(id, \": \", citricAcid_goids[ele][id])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['random_state', 'scores', 'mean_scores', 'common_features', 'model'])\n",
      "10x3FoldCV mean score >>> \n",
      "  glucose  0.37702307897847137 sucrose  0.31889166659943624 citricAcid  0.22358901669508696\n"
     ]
    }
   ],
   "source": [
    "print(glucose_nXcv.keys())\n",
    "print(\"10x3FoldCV mean score >>> \\n \",\"glucose \", np.mean(glucose_nXcv['mean_scores']), \"sucrose \", np.mean(sucrose_nXcv['mean_scores']) ,\"citricAcid \", np.mean(citricAcid_nXcv['mean_scores']))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ida_tf",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
